---
title: "Bayesian DEgs Improvement"
author: "Fl√°vio Lichtenstein - Carlos Eduardo Madureira Trufen"
date: "March 1, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

# This algorithm requires a a normalized expression table with genes in rows and conditions in columns, a differential expression results table (as obtained with edgeR, DESeq2, limma-voom or other tool), as well as an enrichment table result table
```{r, message = FALSE, warning=FALSE}
library(tidyverse)
library(edgeR)
library(clusterProfiler)
library(CEMiTool)
library(cowplot)
library(Rcpp)
library(inline)
library(profvis)
```

# some functions
```{r}
calculate_pgs_log10 <- function(df = df, gene_ids = gene_ids, column = column) {
  # pis <- as.numeric(paste(median_df[median_df[["rowname"]] %in% sample.symbols, "pis_2"]))
  # dfPis = median_df, gene_ids = sample.symbols, thisCol = pis_2
  pis <- calculate_median_output %>% filter(rowname %in% sample.symbols) %>% dplyr::select(column) %>% pull %>% as.numeric
  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8
  pis %>% log10() %>% sum() %>% return()
  #return(sum(log10(pis)))
}

write_txt <- function(x, filename, append = FALSE, sep = "\t", verbose = T) {
  filename <- paste(dir_result, filename, sep = "")
  ret <- tryCatch({
    write(x, file = filename, append = append, sep = sep)
    if (verbose) cat(sprintf("File saved at '%s'.\n", filename))
    TRUE
  }, warning = function(w) {
    cat(paste("Warning saving text -", w, " file =", filename, "\n"))
    TRUE
  }, error = function(e) {
    cat(paste("Error saving text -", e, " file =", filename, "\n"))
    FALSE
  })

  return(ret)
}

get_list <- function(...) {
  lista <- unlist(list(...))
  lista
}

write.log <- function(sep = " ", verbose = TRUE, ...) {
  x <- get_list(...)

  if (!"filename" %in% names(x)) {
    cat("'filename' is obrigatory")
    return("")
  }
  if (!"text" %in% names(x)) {
    cat("'text' is obrigatory")
    return("")
  }

  filename <- x[["filename"]]
  text <- x[["text"]]

  posFilename <- which(names(x) == "filename")
  posText <- which(names(x) == "text")
  x <- x[-c(posFilename, posText)]

  stri <- trim(paste(x, collapse = sep))
  stri <- paste(stri, "\n", sep = "")
  if (verbose) cat(stri)
  stri <- paste(text, paste(Sys.time()), "\n     ", stri, sep = "")

  write_txt(stri, filename, verbose = F)

  return(stri)
}
```

# Median df
```{r}

calculate_median <- function(countdata_input = countdata_input, control_columns = control_columns, compare_columns = compare_columns){
  library(edgeR)
  countdata_input <- filtered_countData %>% cpm() %>% as.data.frame %>% rownames_to_column()
  median_df <- data.frame(
  rowname = countdata_input %>% dplyr::select(rowname) %>% pull(),
  median_control = countdata_input %>% dplyr::select(c(control_columns)) %>% apply(MARGIN = 1, FUN = median),
  median_compare = countdata_input %>% dplyr::select(c(compare_columns)) %>% apply(MARGIN = 1, FUN = median) )
  median_df <- median_df %>% mutate(pis_control = median_control/(median_control %>% sum),
                                    pis_compare = median_compare/(median_compare %>% sum))
  return(median_df)
}

calculate_median_output <- calculate_median(countdata_input = filtered_countData, control_columns = dplyr::contains("CONTROL_TET"), compare_columns = dplyr::contains("LLAEVEN_TET"))

countdata_input <- filtered_countData %>% cpm() %>% as.data.frame %>% rownames_to_column()

# CONTROL_TET_median = median1
# LLAEVEN_TET_median = median2
median_df <- data.frame(
  rowname = countdata_input$rowname,
  median1 = countdata_input %>% dplyr::select(dplyr::contains("TET")) %>% dplyr::select(dplyr::contains("CONTROL")) %>% apply(MARGIN = 1, FUN = median),
  median2 = countdata_input %>% dplyr::select(dplyr::contains("TET")) %>% dplyr::select(dplyr::contains("LLAEVEN")) %>% apply(MARGIN = 1, FUN = median)
)

median_df <- median_df %>% mutate(pis_1 = median1/(median1 %>% sum),
                                  pis_2 = median2/(median2 %>% sum))

countdata_input <- countdata_input %>% full_join(median_df)
```

# Classify genes
```{r}
FDR_limbo <- 0.07
LFC_limbo <- log2(1.4)
# DE_results <- edgeR_results$LLAEVEN_TETxCONTROL_TET
DE_results <- edgeR_results$TNFxCTRL

classify_genes <- function(DE_results = DE_results, calculate_median_output = calculate_median_output, FDR_limbo = 0.07, LFC_limbo = log2(1.4)){
  HC_DEGs <- DE_results %>% dplyr::filter(FDR < THRESHOLD & abs(logFC) >= LFC) %>% dplyr::select(rowname) %>% pull()
  LowVar_df <- calculate_median_output %>% filter(median_control < 2 & median_compare < 2) %>% filter(median_control > 0 & median_compare > 0)
  #---- low Var cannot be DEGs ---------------------
  LowVar_df <- LowVar_df %>% dplyr::slice(-(LowVar_df$rowname %in% HC_DEGs %>% which()))
  midLayer <- LowVar_df %>% dplyr::select(rowname) %>% pull() %>% as.character()
  Limbo_df <- DE_results %>% dplyr::filter(FDR >= THRESHOLD & FDR < FDR_limbo & abs(logFC) > LFC_limbo)
  Limbo_genes <- dfLimbo %>% dplyr::select(rowname) %>% pull()
  cutoff_median <- LowVar_df %>% dplyr::select(median_compare) %>% pull() %>% median()
  MiddleHighExp_df <- dfLowVar %>% filter(median_compare > cutoff.median)
  No_DEGs <- MiddleHighExp_df %>% dplyr::select(rowname) %>% pull() %>% as.character()
  median_hist <- LowVar_df %>% dplyr::select(c(median_control, median_compare)) %>% reshape2::melt() %>% ggplot(aes(x = log10(value))) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", breaks = seq(from = -1.5, to = 1, by = .1)) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(median)") + ggtitle("Limbo frequencies") + facet_wrap(~variable)
  genes_class_list = list()
  genes_class_list[["HC_DEGs"]] <- HC_DEGs
  genes_class_list[["Limbo_genes"]] <- Limbo_genes
  genes_class_list[["No_DEGs"]] <- No_DEGs
  genes_class_list[["Histograms"]] <- median_hist
  genes_class_list[["BDIA_input_genes"]] <- c(HC_DEGs, Limbo_genes, No_DEGs) %>% unique()
  return(genes_class_list)
}

# genes_class <- classify_genes(DE_results = edgeR_results$LLAEVEN_TETxCONTROL_TET, calculate_median_output = calculate_median_output)
genes_class <- classify_genes(DE_results = edgeR_results$TNFxCTRL, calculate_median_output = calculate_median_output)

# Classification step: High Confidence (DEGs), Limbo, Lowvar, NO DEGS
# High Confidence
HC_DEGs <- DEGs_edgeR <- DE_results %>% dplyr::filter(FDR < THRESHOLD & abs(logFC) >= LFC) %>% dplyr::select(rowname) %>% pull()
# bayes.min.median <- 2
dfLowVar <- median_df %>% filter(median1 < 2 & median2 < 2) %>% filter(median1 > 0 & median2 > 0)


# HC_genes %in% dfLowVar$rowname %>% sum

#-- high confidence -------------------------------------------------
#-- parameter: fdr.limbo = 0.07,  cutoffLFC.limbo = log2(1.4) =  0.4854268
FDR_limbo <- 0.07
LFC_limbo <- log2(1.4)

dfLimbo <- DE_results %>% dplyr::filter(FDR >= THRESHOLD & FDR < FDR_limbo & abs(logFC) > LFC_limbo) # %>% dplyr::select(rowname) %>% pull()
dfLimbo %>% dim()
Limbo <- dfLimbo$rowname
#---- low Var cannot be DEGs ---------------------
dfLowVar$rowname %in% HC_DEGs %>% which()
#---- low Var cannot be DEGs ---------------------
dfLowVar %>% dplyr::slice(-(dfLowVar$rowname %in% HC_DEGs %>% which()))

midLayer <- dfLowVar$rowname %>% as.character()

# Obs: Change to ggplot
par(mfrow = c(1, 2))
hist(log10(dfLowVar$median1), main = "median1 - 'limbo' frequencies", breaks = 100) # median1
hist(log10(dfLowVar$median2), main = "median2 - 'limbo' frequencies", breaks = 100) # median2

cutoff.median <- median(dfLowVar$median2)
dfMiddleHighExp <- dfLowVar %>% filter(median2 > cutoff.median)
NO.DEGs <- dfMiddleHighExp$rowname %>% as.character()
```

# Adding ORA analysis
```{r}
# gmt_fname <- system.file("extdata", "pathways.gmt", package = "CEMiTool")
# gmt_fname <- "~/Dropbox/GMT/c2.cp.kegg.v6.1.symbols.gmt" # KEGG: KEGG gene sets
# gmt_fname <- "~/Dropbox/GMT/c1.all.v6.1.symbols.gmt" # positional gene sets
# gmt_fname <- "~/Dropbox/GMT/c2.all.v6.1.symbols.gmt" # curated gene sets
# gmt_fname <- "~/Dropbox/GMT/c2.cp.biocarta.v6.1.symbols.gmt" # BIOCARTA: BioCarta gene sets
# gmt_fname <- "~/Dropbox/GMT/c2.cp.reactome.v6.1.symbols.gmt" # REACTOME: Reactome gene sets
# gmt_fname <- "~/Dropbox/GMT/c2.cgp.v6.1.symbols.gmt" # chemical and genetic perturbations
# gmt_fname <- "~/Dropbox/GMT/c2.cp.v6.1.symbols.gmt" # Canonical pathways
# gmt_fname <- "~/Dropbox/GMT/c3.all.v6.1.symbols.gmt" # motif gene sets
# gmt_fname <- "~/Dropbox/GMT/c3.mir.v6.1.symbols.gmt" # microRNA targets
# gmt_fname <- "~/Dropbox/GMT/c3.tft.v6.1.symbols.gmt" # transcription factor targets
# gmt_fname <- "~/Dropbox/GMT/c4.all.v6.1.symbols.gmt" # computational gene sets
# gmt_fname <- "~/Dropbox/GMT/c4.cgn.v6.1.symbols.gmt" # cancer gene neighborhoods
# gmt_fname <- "~/Dropbox/GMT/c4.cm.v6.1.symbols.gmt" # cancer modules
# gmt_fname <- "~/Dropbox/GMT/c5.all.v6.1.symbols.gmt" # GO gene sets
# gmt_fname <- "~/Dropbox/GMT/c5.bp.v6.1.symbols.gmt" # GO biological process
# gmt_fname <- "~/Dropbox/GMT/c5.cc.v6.1.symbols.gmt" # GO cellular component
# gmt_fname <- "~/Dropbox/GMT/c5.mf.v6.1.symbols.gmt" # GO molecular function
# gmt_fname <- "~/Dropbox/GMT/c6.all.v6.1.symbols.gmt" # oncogenic signatures
# gmt_fname <- "~/Dropbox/GMT/c7.all.v6.1.symbols.gmt" # immunologic signatures
# gmt_fname <- "~/Dropbox/GMT/h.all.v6.1.symbols.gmt" # hallmark gene sets

# gmt_in$gene[which(is.na(match(gmt_in$gene,Conversion_Human2Human$SYMBOL))==FALSE)] = Conversion_Human2Human$ENSEMBL[match(gmt_in$gene[which(is.na(match(gmt_in$gene,Conversion_Human2Human$SYMBOL))==FALSE)],Conversion_Human2Human$SYMBOL)]
```

```{r}
path_to_gmt <- "~/Dropbox/GMT/WikiPathways_2019_Human.txt" # GO gene sets
gmt_list <- path_to_gmt %>% CEMiTool::read_gmt()
custom_WikiPathways_2019_Human = gmt_list %>% split(f = .$term) %>% map(dplyr::select, gene) %>% map(pull)
```

```{r}
path_to_gmt <- "~/Dropbox/GMT/c5.all.v6.1.symbols.gmt" # GO gene sets
gmt_list <- path_to_gmt %>% CEMiTool::read_gmt()
custom_GO_Pathways = gmt_list %>% split(f = .$term) %>% map(dplyr::select, gene) %>% map(pull)
```

```{r}
path_to_gmt <- "~/Dropbox/GMT/c2.cp.reactome.v6.1.symbols.gmt" # Reactome gene sets
gmt_list <- path_to_gmt %>% CEMiTool::read_gmt()
custom_Reactome_Pathways = gmt_list %>% split(f = .$term) %>% map(dplyr::select, gene) %>% map(pull)
```

```{r}
path_to_gmt <- "~/Dropbox/GMT/c2.cp.kegg.v6.1.symbols.gmt" # KEGG gene sets
gmt_list <- path_to_gmt %>% CEMiTool::read_gmt()
custom_KEGG_Pathways = gmt_list %>% split(f = .$term) %>% map(dplyr::select, gene) %>% map(pull)
```

# from CEMiTool/clusterProfiler
```{r}
# allgenes <- gmt_list[, "gene"] %>% unique
# 
# enrichment <- function(genes, gmt_list) {
#   library(clusterProfiler)
#   allgenes <- gmt_list[, "gene"] %>% unique
#   enriched <- clusterProfiler::enricher(
#     gene = genes,
#     pvalueCutoff = 1,
#     qvalueCutoff = 1,
#     universe = allgenes,
#     TERM2GENE = gmt_list
#   )
#   enrichment_result <- enriched@result %>% filter(qvalue < 0.05)
#   enrichment_result <- enrichment_result %>% filter(Count > 2)
#   return(enrichment_result)
# }
# 
# enrichment_result <- edgeR_pipeline_results$topDE2$`06hx0h`$rowname %>% enrichment(gmt_list)
# 
# 
# 
# 
# # DEGs_edgeR <- edgeR_results$LLAEVEN_TETxCONTROL_TET$rowname
# DEGs_edgeR <- edgeR_results$TNFxCTRL$rowname %>% str_to_upper()
# 
# enriched <- clusterProfiler::enricher(
#   gene = DEGs_edgeR,
#   pvalueCutoff = 1,
#   qvalueCutoff = 1,
#   universe = allgenes,
#   TERM2GENE = gmt_list
# )
# 
# enrichment_result <- enriched@result %>% filter(qvalue < 0.05)
# enrichment_result <- enrichment_result %>% filter(Count > 2)
```


# Enrichment analysis (ORA)
# custom ORA
# Databases
```{r}
WikiPathways_2019_Human <- "~/Dropbox/GMT/WikiPathways_2019_Human.txt" # WikiPathways_2019_Human gene sets
WikiPathways_2019_Mouse <- "~/Dropbox/GMT/WikiPathways_2019_Mouse.txt" # WikiPathways_2019_Mouse gene sets
KEGG_2019_Human <- "~/Dropbox/GMT/KEGG_2019_Human.txt" # KEGG_2019_Human gene sets
KEGG_2019_Mouse <- "~/Dropbox/GMT/KEGG_2019_Mouse.txt" # KEGG_2019_Mouse gene sets
LINCS_L1000_Chem_Pert_down <- "~/Dropbox/GMT/LINCS_L1000_Chem_Pert_down.txt" # LINCS_L1000_Chem_Pert_down gene sets
LINCS_L1000_Chem_Pert_up <- "~/Dropbox/GMT/LINCS_L1000_Chem_Pert_up.txt" # LINCS_L1000_Chem_Pert_up gene sets
Reactome_2016 <- "~/Dropbox/GMT/Reactome_2016.txt" # Reactome_2016 gene sets
GO_Molecular_Function_2018 <- "~/Dropbox/GMT/GO_Molecular_Function_2018.txt" # GO_Molecular_Function_2018 gene sets
GO_Cellular_Component_2018 <- "~/Dropbox/GMT/GO_Cellular_Component_2018.txt" # GO_Cellular_Component_2018 gene sets
GO_Biological_Process_2018 <- "~/Dropbox/GMT/GO_Biological_Process_2018.txt" # GO_Biological_Process_2018 gene sets

# database = c("KEGG_2019_Human")
```

# custom enrichment analysis
```{r}
ora_analysis <- function(genes, database = KEGG_2019_Human, pvalue_cutoff = 1, qvalue_cutoff = 1, p.adjust_cutoff = 0.1, count_at_least = 3){
  library(clusterProfiler)
  gmt_list <- database %>% read.gmt()
  custom_pathways <- gmt_list %>% split(f = .$ont) %>% map(dplyr::select, gene) %>% map(pull)
  allgenes <- gmt_list[, "gene"] %>% unique()
  enrichment_result <- genes %>% map(clusterProfiler::enricher, pvalueCutoff = pvalue_cutoff, qvalueCutoff = qvalue_cutoff, universe = allgenes, TERM2GENE = gmt_list)
  clProf.df <- enrichment_result %>% map(as.data.frame) %>% bind_rows(.id = "Cluster")
  clProf.df <- clProf.df %>% filter(p.adjust < p.adjust_cutoff & Count >= count_at_least)
  geneClusters <- genes
  compareCluster_output <- new("compareClusterResult", compareClusterResult = clProf.df, geneClusters = geneClusters, fun = "enricher",
  .call = match.call(expand.dots = TRUE))
  return(compareCluster_output)
}

# ora_analysis_2 <- function(genes, database = KEGG_2019_Human, pvalue_cutoff = 1, qvalue_cutoff = 1, p.adjust_cutoff = 0.1, count_at_least = 3, universe = universe){
#   library(CEMiTool)
#   library(clusterProfiler)
#   gmt_list <- database %>% CEMiTool::read_gmt()
#   custom_pathways <- gmt_list %>% split(f = .$term) %>% map(dplyr::select, gene) %>% map(pull)
#   allgenes <- gmt_list[, "gene"] %>% unique()
#   enrichment_result <- genes %>% map(clusterProfiler::enricher, pvalueCutoff = pvalue_cutoff, qvalueCutoff = qvalue_cutoff, universe = universe, TERM2GENE = gmt_list)
#   clProf.df <- enrichment_result %>% map(as.data.frame) %>% bind_rows(.id = "Cluster")
#   clProf.df <- clProf.df %>% filter(p.adjust < p.adjust_cutoff & Count >= count_at_least)
#   geneClusters <- genes
#   compareCluster_output <- new("compareClusterResult", compareClusterResult = clProf.df, geneClusters = geneClusters, fun = "enricher",
#   .call = match.call(expand.dots = TRUE))
#   return(compareCluster_output)
# }
```

# ORA results
```{r}
genes <- edgeR_topDE2$TNFxCTRL$rowname %>% toupper() %>% list() %>% purrr::set_names("TNFxCTRL")
ora_results <- genes %>% ora_analysis(database = KEGG_2019_Mouse, p.adjust_cutoff = 0.05)
```

# BDIA Functions
```{r}
calculate_median <- function(countdata_input = countdata_input, control_columns = control_columns, compare_columns = compare_columns){
  library(edgeR)
  countdata_input <- filtered_countData %>% cpm() %>% as.data.frame %>% rownames_to_column()
  median_df <- data.frame(
  rowname = countdata_input %>% dplyr::select(rowname) %>% pull(),
  median_control = countdata_input %>% dplyr::select(c(control_columns)) %>% apply(MARGIN = 1, FUN = median),
  median_compare = countdata_input %>% dplyr::select(c(compare_columns)) %>% apply(MARGIN = 1, FUN = median) )
  median_df <- median_df %>% mutate(pis_control = median_control/(median_control %>% sum),
                                    pis_compare = median_compare/(median_compare %>% sum))
  median_df <- median_df %>% mutate(median_control = median_control %>% formattable::digits(digits = 2),
                                    median_compare = median_compare %>% formattable::digits(digits = 2),
                                    pis_control = pis_control %>% formattable::scientific(digits = 2),
                                    pis_compare = pis_compare %>% formattable::scientific(digits = 2))
  output = list()
  output[["countdata_input"]] <- countdata_input
  output[["median_df"]] <- median_df
  return(output)
}

# calculate_median_output <- calculate_median(countdata_input = filtered_countData, control_columns = dplyr::contains("CONTROL_TET"), compare_columns = dplyr::contains("LLAEVEN_TET"))

rownames(filtered_countData) <- rownames(filtered_countData) %>% toupper()
filtered_countData_sub <- filtered_countData %>% as.data.frame() %>% dplyr::select(-c(dplyr::contains("P4")))
calculate_median_output <- calculate_median(countdata_input = filtered_countData_sub, control_columns = dplyr::contains("CTRL_"), compare_columns = dplyr::contains("TNF_"))

# filtered_countData_sub <- filtered_countData %>% as.data.frame() %>% dplyr::select(-c(dplyr::contains("h12")))
# calculate_median_output <- calculate_median(countdata_input = filtered_countData_sub, control_columns = dplyr::contains("h0"), compare_columns = dplyr::contains("h6"))

classify_genes <- function(calculate_median_output = calculate_median_output, DE_results = DE_results, FDR_limbo = 0.07, LFC_limbo = log2(1.4)) {
  median_df <- calculate_median_output[["median_df"]]
  HC_DEGs <- DE_results %>%
    dplyr::filter(FDR < THRESHOLD & abs(logFC) >= LFC) %>%
    dplyr::select(rowname) %>%
    pull()
  LowVar_df <- median_df %>%
    filter(median_control < 2 & median_compare < 2) %>%
    filter(median_control > 0 & median_compare > 0)
  #---- low Var cannot be DEGs ---------------------
  LowVar_df <- LowVar_df %>% dplyr::slice(-(LowVar_df$rowname %in% HC_DEGs %>% which()))
  midLayer <- LowVar_df %>%
    dplyr::select(rowname) %>%
    pull() %>%
    as.character()
  Limbo_df <- DE_results %>% dplyr::filter(FDR >= THRESHOLD & FDR < FDR_limbo & abs(logFC) > LFC_limbo)
  Limbo_genes <- Limbo_df %>%
    dplyr::select(rowname) %>%
    pull()
  cutoff_median <- LowVar_df %>%
    dplyr::select(median_compare) %>%
    pull() %>%
    median()
  MiddleHighExp_df <- LowVar_df %>% filter(median_compare > cutoff_median)
  No_DEGs <- MiddleHighExp_df %>%
    dplyr::select(rowname) %>%
    pull() %>%
    as.character()
  median_hist <- LowVar_df %>%
    dplyr::select(c(median_control, median_compare)) %>%
    reshape2::melt() %>%
    ggplot(aes(x = log10(value))) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", breaks = seq(from = -1.5, to = 0.5, by = .1)) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(median)") + ggtitle("Limbo frequencies") + facet_wrap(~variable)
  genes_class_list <- list(
    HC_DEGs = HC_DEGs,
    Limbo_genes = Limbo_genes,
    No_DEGs = No_DEGs,
    Histograms = median_hist,
    BDIA_input_genes = c(HC_DEGs, Limbo_genes, No_DEGs) %>% unique(),
    median_df = calculate_median_output[["median_df"]],
    countdata_input = calculate_median_output[["countdata_input"]]
  )
  return(genes_class_list)
}

# DE_results <- edgeR_results$LLAEVEN_TETxCONTROL_TET
DE_results <- edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper())
# DE_results <- edgeR_pipeline_results$results$`06hx0h`

classify_genes_output <- classify_genes(DE_results = DE_results, calculate_median_output = calculate_median_output)

# Rcpp version
#  implement the following using R functions in C++, using the hints provided: %in% using unordered_set and the find() or count() methods.

# calculate_pgs_log10 <- cppFunction('Rcpp::NumericVector pgs_log10 = df{
#   Rcpp::DataFrame df = Rcpp::as<Rcpp::DataFrame>(df);
#   Rcpp::IntegerVector median_control = df["median_control"];
#   Rcpp::IntegerVector median_compare = df["median_compare"];
#   Rcpp::IntegerVector pis_control = df["pis_control"];
#   Rcpp::IntegerVector pis_compare = df["pis_compare"];
#   Rcpp::CharacterVector gene_ids = gene_ids;
#   Rcpp::CharacterVector column = df["column"];
#   
#   std::unordered_set<int>
#   
# 
#   return pgs_log10
# 
#                                       }')

# from https://stackoverflow.com/questions/21359432/a-c-version-of-the-in-operator-in-r/21383041#21383041
# which_in2 <- cppFunction('std::vector<int> which_in2(IntegerVector x, IntegerVector y) {
#   std::vector<int> y_sort(y.size());
#   std::partial_sort_copy (y.begin(), y.end(), y_sort.begin(), y_sort.end());
#   int nx = x.size();
#   std::vector<int> out;
#   for (int i = 0; i < nx; ++i) {
#     std::vector<int>::iterator found =
#       lower_bound(y_sort.begin(), y_sort.end(), x[i]);
#     if (found != y_sort.end()) {
#       out.push_back(i + 1);
#     }
#   }
#   return out;
# }')

# from https://stackoverflow.com/questions/40691823/rcpp-subsetting-rows-of-dataframe
# cppFunction('LogicalVector subset_c(DataFrame df, StringVector column_level) {
#   using namespace std;  
#   StringVector sub = df["rowname"];
#   std::string level = Rcpp::as<std::string>(column_level[0]);
#   Rcpp::LogicalVector ind(sub.size());
#   for (int i = 0; i < sub.size(); i++){
#       ind[i] = (sub[i] == level);
#   }
# 
#   return(ind);
# }')

# cppFunction('LogicalVector BDIA_subset_c(DataFrame df, StringVector column_level) {
#   using namespace std;  
#   StringVector sub = df["rowname"];
#   for (int j = 0; j < column_level.size(); j++){
#   std::string level = Rcpp::as<std::string>(column_level[j]);
#   Rcpp::LogicalVector ind(sub.size());
#   for (int i = 0; i < sub.size(); i++){
#       ind[i] = (sub[i] == level);
#   }
#   }
# 
#   return(ind);
# }')

# calculate_pgs_log10 <- cppFunction('std::vector<double> calculate_pgs_log10(DataFrame df, CharacterVector gene_ids, CharacterVector column){
#   std::vector<double> median_control = df["median_control"];
#   std::vector<double> median_compare = df["median_compare"];
#   std::vector<double> pis_control = df["pis_control"];
#   std::vector<double> pis_compare = df["pis_compare"];
#   std::vector<double> column = df["column"];
#   std::list dimnames = df.attr("dimnames");
#   std::vector<char> rownames = dimnames[0];
#   std::vector<char> pis_rows = which_in2(rownames, gene_ids);
#   std::vector<double> pis = df[]
#   int n = pis.size();
#   for(i=0; i<n; i++){
#       if (pis[i] = 0) pis[i] = 1e-8;
#       if (pis[i] = NA_REAL) pis[i] = 1e-8;
#   }
#   pis = log10(pis);
#   pis_sum = sum(pis);
#   return pis_sum;
#                                       }')

# calculate_pgs_log10_c <- cppFunction('Rcpp::NumericVector calculate_pgs_log10_c(Rcpp::DataFrame df, Rcpp::CharacterVector gene_ids, Rcpp::CharacterVector column){
#   Rcpp::IntegerVector median_control = df["median_control"];
#   Rcpp::IntegerVector median_compare = df["median_compare"];
#   Rcpp::IntegerVector pis_control = df["pis_control"];
#   Rcpp::IntegerVector pis_compare = df["pis_compare"];
#   Rcpp::CharacterVector column = df["column"];
#   Rcpp::List dimnames = df.attr("dimnames");
#   Rcpp::CharacterVector rownames = dimnames[0];
#   Rcpp::CharacterVector pis = which_in2(rownames, gene_ids);
  # sub_df = Rcpp::DataFrame::create(Rcpp::Named("rowname") = rowname[ind],
  #                                  Rcpp::Named("pis_compare") = pis_compare[ind]
  #                                  );
  # 
  # CharacterVector pis = sub_df["pis_compare"];
#   int n = pis.size();
#   for(i=0; i<n; i++){
#       if (pis[i] = 0) pis[i] = 1e-8;
#       if (pis[i] = NA_REAL) pis[i] = 1e-8;
#   }
#   pis = log10(pis);
#   pis_sum = sum(pis);
#   return pis_sum;
#                                       }')

# subset dataframe -> Boolean over gene_ids in pis_compare
# select column as a vector -> subset vector (pis_compare) by boolean ind
# recalculate values if 0 or NA
# calculate log10
# calculate sum
# return calculated value
# from https://stackoverflow.com/questions/43909024/data-type-conversion-error-in-using-rcpp
# cppFunction('Rcpp::NumericVector calculate_pgs_log10_c(Rcpp::DataFrame df, Rcpp::StringVector gene_ids){
#   NumericVector pis_compare = df["pis_compare"];
#   StringVector x = df["rowname"];
#   StringVector y = gene_ids;
#   LogicalVector ind = in(x, y);
#   NumericVector pis = pis_compare[ind];
#   int n = pis.size();
#   for(int i = 0; i < n; i++){
#       if (pis[i] = 0) pis[i] = 1e-8;
#       if (pis[i] = NA_REAL) pis[i] = 1e-8;
#   }
#   NumericVector pis_updated = std::replace(pis.begin(), pis.end(), 0, 1e-8);
#   NumericVector pis_updated = std::replace(pis.begin(), pis.end(), NA_REAL, 1e-8);
#   NumericVector logpis = log10(pis);
#   double pis_sum = sum(logpis);
#   return pis_sum;
#                                       }')



# cppFunction('Rcpp::NumericVector calculate_pgs_log10_c(Rcpp::DataFrame df, Rcpp::StringVector gene_ids){
#   NumericVector pis_compare = df["pis_compare"];
#   StringVector x = df["rowname"];
#   StringVector y = gene_ids;
#   LogicalVector ind = in(x, y);
#   NumericVector pis = pis_compare[ind];
#   std::replace(pis.begin(), pis.end(), 0, 1e-8);
#   return pis;
#                                       }')

  # NumericVector logpis = log10(pis);
  # double pis_sum = sum(logpis);
  # return pis_sum;

calculate_pgs_log10 <- function(df = df, gene_ids = gene_ids, column = column) {
  pis <- df %>% dplyr::filter(rowname %in% gene_ids) %>% dplyr::select(c(column)) %>% pull() %>% as.numeric()
  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8
  pis[is.numeric0(pis)] <- 1e-8
  pis %>% log10() %>% sum() %>% return()
}

# sourceCpp('calculate_pgs_log10_c.cc')
# get_enriched_pathways <- function(enrichment_result = enrichment_result, Description = Description, pathways_list = pathways_list) {
#   pathways <- enrichment_result %>% dplyr::select(c(Description)) %>% pull() %>% as.list() %>% map(as.character())
#   genes_in_pathway <- map(seq_along(pathways), function(i) custom_KEGG_Pathways[[pathways[[i]]]] %>% sort())
#   DEGs_found_in_pathway <- enrichment_result %>% dplyr::select(geneID) %>% map(str_split, pattern = "/") %>% purrr::set_names(NULL) %>% .[[1]]
#   n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% map(length)
#   enriched_pathways <- list()
#   enriched_pathways[["genes_in_pathway"]] <- genes_in_pathway %>% purrr::set_names(pathways)
#   enriched_pathways[["DEGs_found_in_pathway"]] <- DEGs_found_in_pathway %>% purrr::set_names(pathways)
#   enriched_pathways[["n_DEGs_found_in_pathway"]] <- n_DEGs_found_in_pathway %>% purrr::set_names(pathways)
#   return(enriched_pathways)
# }

get_enriched_pathways <- function(enrichment_result = enrichment_result, Description = Description, database = KEGG_2019_Human) {
  gmt_list <- database %>% read.gmt()
  pathways_list <- gmt_list %>% split(f = .$ont) %>% map(dplyr::select, gene) %>% map(pull)
  pathways <- enrichment_result %>% dplyr::select(c(Description)) %>% pull() %>% as.list() %>% map(as.character())
  genes_in_pathway <- map(seq_along(pathways), function(i) pathways_list[[pathways[[i]]]] %>% sort())
  DEGs_found_in_pathway <- enrichment_result %>% dplyr::select(geneID) %>% map(str_split, pattern = "/") %>% purrr::set_names(NULL) %>% .[[1]]
  n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% map(length)
  enriched_pathways <- vector("list", length = length(pathways)) %>% purrr::set_names(pathways)
  for (pathway in seq_along(enriched_pathways)) {
    enriched_pathways[[pathway]][["genes_in_pathway"]] <- genes_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["DEGs_found_in_pathway"]] <- DEGs_found_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["n_DEGs_found_in_pathway"]] <- n_DEGs_found_in_pathway[[pathway]]
  }
  return(enriched_pathways)
}

# get_enriched_pathways_output <- get_enriched_pathways(enrichment_result, Description, database = KEGG_2019_Human)
# get_enriched_pathways_output <- get_enriched_pathways(enrichment_result, Description, database = Reactome_2016)
get_enriched_pathways_output <- get_enriched_pathways(enrichment_result = ora_results@compareClusterResult, Description, database = KEGG_2019_Mouse)

nSample <- 50
nDrawn <- 50
nLoop <- 1000

library(parallelMap)
no_cores <- detectCores() - 1

library(profvis)
# Build df approach
BDIA_bootstrap <- function(classify_genes_output = classify_genes_output, nSample = 50, nDrawn = 50, nLoop = 1000, get_enriched_pathways_output = get_enriched_pathways_output, no_cores = (detectCores() - 1), seed = 123) {
  profvis({
  library(parallelMap)
  # Initiate cluster
  set.seed(seed)
  parallelStartSocket(no_cores)
  countdata_input <- classify_genes_output[["countdata_input"]]
  input_genes <- classify_genes_output[["BDIA_input_genes"]]
  Limbo_genes <- classify_genes_output[["Limbo_genes"]]
  HC_DEGs <- classify_genes_output[["HC_DEGs"]]
  No_DEGs <- classify_genes_output[["No_DEGs"]]
  median_df <- classify_genes_output[["median_df"]]
  pathway_list <- list()
  for (i in seq_along(get_enriched_pathways_output)) {
    pathway <- get_enriched_pathways_output %>% names() %>% .[i]
    print(pathway)
    genes_in_pathway <- get_enriched_pathways_output[[pathway]][["genes_in_pathway"]]
    DEGs_found_in_pathway <- get_enriched_pathways_output[[pathway]][["DEGs_found_in_pathway"]] %>% sort()
    n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% length()
    nGoodGenes <- input_genes %>% length()
    nAllGenes <- countdata_input %>% nrow()
    nBadGenes <- nAllGenes - nGoodGenes
    # No bootstrap for loop
    sample_genes_list <- replicate(n = nLoop, list(sample(input_genes, nSample, replace = FALSE)))
    genes_found <- sample_genes_list %>% map(function(x) x[x %in% get_enriched_pathways_output[[pathway]][["genes_in_pathway"]]])
    nFound <- genes_found %>% map(length)
    # Hypergeometric test
    pORA <- nFound %>% map(phyper, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    log10.pGs <- sample_genes_list %>% map(calculate_pgs_log10, df = median_df, column = "pis_compare")
    post <- map(seq_along(sample_genes_list), function(i) -log10(pORA[[i]]) - log10.pGs[[i]])
    prioriDist <- log10.pGs %>% unlist()

    BDIA_df <- data.frame(
      sample_genes = sample_genes_list %>% map(paste0, collapse = ", ") %>% unlist(),
      maxiGenes = genes_found %>% map(paste0, collapse = ", ") %>% unlist(),
      n_Found = nFound %>% unlist(),
      verossDist = pORA %>% unlist(),
      prioriDist = log10.pGs %>% unlist(),
      postDist = post %>% unlist(),
      stringsAsFactors = FALSE
    )

    postDist2_df <- BDIA_df$postDist %>%
      na.omit() %>%
      sort() %>%
      as.data.frame() %>%
      purrr::set_names("postDist2")
    postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", binwidth = 2) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
    L <- postDist2_df %>%
      pull() %>%
      length()
    up <- L * .95
    postDist2.sig <- postDist2_df$postDist2[1:up]
    limSup <- postDist2.sig %>% max()
    n <- which(postDist2_df$postDist2 < limSup)
    # genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
    genes <- map(seq_along(n), function(i) BDIA_df$maxiGenes[[n[i]]] %>% strsplit(", ")) %>%
      unlist() %>%
      unique() %>%
      sort()
    genes.found.hc.DEGs <- genes %>% .[. %in% HC_DEGs]
    genes.found.limbo <- genes %>% .[. %in% Limbo_genes]
    genes.found.no.DEGs <- genes %>% .[. %in% No_DEGs]
    prev.found <- DEGs_found_in_pathway %>% sort()
    found.DEGs.and.limbo <- c(genes.found.hc.DEGs, genes.found.limbo) %>%
      unique() %>%
      sort()
    common.symbols <- prev.found %>% .[. %in% found.DEGs.and.limbo]
    only.prev.found <- prev.found %>% .[!. %in% found.DEGs.and.limbo]
    only.new.found <- found.DEGs.and.limbo %>% .[!. %in% prev.found]
    # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
    llMu <- postDist2_df$postDist2 %>%
      mean() %>%
      round(digits = 2)
    minLL <- postDist2_df$postDist2 %>%
      min() %>%
      round(digits = 2)
    maxLL <- postDist2_df$postDist2 %>%
      max() %>%
      round(digits = 2)
    BDIA_df <- BDIA_df %>% mutate(
      verossDist = verossDist %>% format(digits = 2, scientific = TRUE),
      prioriDist = prioriDist %>% format(digits = 5),
      postDist = postDist %>% format(digits = 5)
    )
    pathway_list[[pathway]] <- list(
      genes_found_in_pathway = prev.found,
      genes_found_new_algorithm = genes,
      genes_found_DEGs_Limbo = found.DEGs.and.limbo,
      genes_found_in_HC_DEGs = genes.found.hc.DEGs,
      genes_found_in_Limbo = genes.found.limbo,
      genes_found_in_No_DEGs = genes.found.no.DEGs,
      common = common.symbols,
      only_previous_DEGs = only.prev.found,
      only_new_DEGs = only.new.found,
      HC_DEGs = HC_DEGs,
      Limbo_genes = Limbo_genes,
      No_DEGs = No_DEGs,
      Histogram = postDist2_gghist,
      BDIA_summary = data.frame("mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n_genes_found_in_pathway" = prev.found %>% length(), "n_genes_found_new_algorithm" = genes %>% length(), "n_genes_found_DEGs_limbo" = found.DEGs.and.limbo %>% length()),
      BDIA_df = BDIA_df
    )
  }
  return(pathway_list)
  parallelStop()
  })
}
```

# Faster BDIA algorithm with Rcpp
```{r}
is.numeric0 <- function(x){is.numeric(x) && length(x) == 0}

cppFunction('Rcpp::NumericVector extract_pis_c(Rcpp::DataFrame df, Rcpp::StringVector gene_ids){
  NumericVector pis_compare = df["pis_compare"];
  StringVector x = df["rowname"];
  StringVector y = gene_ids;
  LogicalVector ind = in(x, y);
  return pis_compare[ind];
                                      }')

cppFunction('Rcpp::NumericVector log10_c(NumericVector x){
  NumericVector log10x = log10(x);
  return log10x;
                                      }')

conditionally_modify_pis <- function(pis = pis) {
  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8
  pis[is.numeric0(pis)] <- 1e-8
  pis %>% return()
}

calculate_pgs_log10_c <- function(df = df, gene_ids = gene_ids) {
  pgs <- df %>% extract_pis_c %>% conditionally_modify_pis %>% log10_c %>% sum()
  return(pgs)
}

# Build df approach
BDIA_bootstrap <- function(classify_genes_output = classify_genes_output, nSample = 50, nDrawn = 50, nLoop = 1000, get_enriched_pathways_output = get_enriched_pathways_output, no_cores = (detectCores() - 1), seed = 123) {
  profvis({
  library(parallelMap)
  # Initiate cluster
  set.seed(seed)
  parallelStartSocket(no_cores)
  countdata_input <- classify_genes_output[["countdata_input"]]
  input_genes <- classify_genes_output[["BDIA_input_genes"]]
  Limbo_genes <- classify_genes_output[["Limbo_genes"]]
  HC_DEGs <- classify_genes_output[["HC_DEGs"]]
  No_DEGs <- classify_genes_output[["No_DEGs"]]
  median_df <- classify_genes_output[["median_df"]]
  pathway_list <- list()
  for (i in seq_along(get_enriched_pathways_output)) {
    pathway <- get_enriched_pathways_output %>% names() %>% .[i]
    print(pathway)
    genes_in_pathway <- get_enriched_pathways_output[[pathway]][["genes_in_pathway"]]
    DEGs_found_in_pathway <- get_enriched_pathways_output[[pathway]][["DEGs_found_in_pathway"]] %>% sort()
    n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% length()
    nGoodGenes <- input_genes %>% length()
    nAllGenes <- countdata_input %>% nrow()
    nBadGenes <- nAllGenes - nGoodGenes
    # No bootstrap for loop
    sample_genes_list <- replicate(n = nLoop, list(sample(input_genes, nSample, replace = FALSE)))
    genes_found <- sample_genes_list %>% map(function(x) x[x %in% get_enriched_pathways_output[[pathway]][["genes_in_pathway"]]])
    nFound <- genes_found %>% map(length)
    # Hypergeometric test
    pORA <- nFound %>% map(phyper, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    # log10.pGs <- sample_genes_list %>% map(extract_pis_c, df = median_df) %>% map(conditionally_modify_pis) %>% map(log10_c) %>% map(sum)
    log10.pGs <- sample_genes_list %>% map(calculate_pgs_log10_c, df = median_df)
    post <- map(seq_along(sample_genes_list), function(i) -log10(pORA[[i]]) - log10.pGs[[i]])
    prioriDist <- log10.pGs %>% unlist()

    BDIA_df <- data.frame(
      sample_genes = sample_genes_list %>% map(paste0, collapse = ", ") %>% unlist(),
      maxiGenes = genes_found %>% map(paste0, collapse = ", ") %>% unlist(),
      n_Found = nFound %>% unlist(),
      verossDist = pORA %>% unlist(),
      prioriDist = log10.pGs %>% unlist(),
      postDist = post %>% unlist(),
      stringsAsFactors = FALSE
    )

    postDist2_df <- BDIA_df$postDist %>%
      na.omit() %>%
      sort() %>%
      as.data.frame() %>%
      purrr::set_names("postDist2")
    postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", binwidth = 2) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
    L <- postDist2_df %>%
      pull() %>%
      length()
    up <- L * .95
    postDist2.sig <- postDist2_df$postDist2[1:up]
    limSup <- postDist2.sig %>% max()
    n <- which(postDist2_df$postDist2 < limSup)
    # genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
    genes <- map(seq_along(n), function(i) BDIA_df$maxiGenes[[n[i]]] %>% strsplit(", ")) %>%
      unlist() %>%
      unique() %>%
      sort()
    genes.found.hc.DEGs <- genes %>% .[. %in% HC_DEGs]
    genes.found.limbo <- genes %>% .[. %in% Limbo_genes]
    genes.found.no.DEGs <- genes %>% .[. %in% No_DEGs]
    prev.found <- DEGs_found_in_pathway %>% sort()
    found.DEGs.and.limbo <- c(genes.found.hc.DEGs, genes.found.limbo) %>%
      unique() %>%
      sort()
    common.symbols <- prev.found %>% .[. %in% found.DEGs.and.limbo]
    only.prev.found <- prev.found %>% .[!. %in% found.DEGs.and.limbo]
    only.new.found <- found.DEGs.and.limbo %>% .[!. %in% prev.found]
    # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
    llMu <- postDist2_df$postDist2 %>%
      mean() %>%
      round(digits = 2)
    minLL <- postDist2_df$postDist2 %>%
      min() %>%
      round(digits = 2)
    maxLL <- postDist2_df$postDist2 %>%
      max() %>%
      round(digits = 2)
    BDIA_df <- BDIA_df %>% mutate(
      verossDist = verossDist %>% format(digits = 2, scientific = TRUE),
      prioriDist = prioriDist %>% format(digits = 5),
      postDist = postDist %>% format(digits = 5)
    )
    pathway_list[[pathway]] <- list(
      genes_found_in_pathway = prev.found,
      genes_found_new_algorithm = genes,
      genes_found_DEGs_Limbo = found.DEGs.and.limbo,
      genes_found_in_HC_DEGs = genes.found.hc.DEGs,
      genes_found_in_Limbo = genes.found.limbo,
      genes_found_in_No_DEGs = genes.found.no.DEGs,
      common = common.symbols,
      only_previous_DEGs = only.prev.found,
      only_new_DEGs = only.new.found,
      HC_DEGs = HC_DEGs,
      Limbo_genes = Limbo_genes,
      No_DEGs = No_DEGs,
      Histogram = postDist2_gghist,
      BDIA_summary = data.frame("mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n_genes_found_in_pathway" = prev.found %>% length(), "n_genes_found_new_algorithm" = genes %>% length(), "n_genes_found_DEGs_limbo" = found.DEGs.and.limbo %>% length()),
      BDIA_df = BDIA_df
    )
  }
  return(pathway_list)
  parallelStop()
  })
}
```

# BDIA_results
```{r}
BDIA_results <- classify_genes_output %>% BDIA_bootstrap(get_enriched_pathways_output = get_enriched_pathways_output, no_cores = (detectCores() - 1))
```

# microbenchmark R vs Rcpp
```{r}

is.numeric0 <- function(x){is.numeric(x) && length(x) == 0}

library(microbenchmark)
extract_pis <- function(df = df, gene_ids = gene_ids) {
  pis <- df %>% dplyr::filter(rowname %in% gene_ids) %>% dplyr::select(c("pis_compare")) %>% pull() %>% as.numeric()
  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8
  pis[is.numeric0(pis)] <- 1e-8
  pis %>% return()
}

conditionally_modify_pis <- function(pis = pis) {
  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8
  pis[is.numeric0(pis)] <- 1e-8
  pis %>% return()
}

mb_results <- microbenchmark(
  R_based = sample_genes_list %>% map(extract_pis, df = median_df) %>% map(log10) %>% map(sum),
  RCpp_built = sample_genes_list %>% map(extract_pis_c, df = median_df) %>% map(conditionally_modify_pis) %>% map(log10_c) %>% map(sum),
  times = 100
)

mb_results %>% autoplot()
```

```{r}

```

```{r}
mb_results <- microbenchmark(
                              R_based = sample_genes_list %>% map(calculate_pgs_log10, df = median_df, column = "pis_compare") %>% map(log10) %>% map(sum),
                              RCpp_built = sample_genes_list %>% map(calculate_pgs_log10_c, df = median_df) %>% map(log10) %>% map(sum))

mb_results %>% autoplot()
```

# pipeline test
```{r}
genes <- edgeR_topDE2$TNFxCTRL$rowname %>% toupper() %>% list() %>% purrr::set_names("TNFxCTRL")

ora_results <- genes %>% ora_analysis(database = KEGG_2019_Mouse, p.adjust_cutoff = 0.05)
get_enriched_pathways_output <- ora_results@compareClusterResult %>% get_enriched_pathways(Description, database = KEGG_2019_Mouse)

ora_results <- genes %>% ora_analysis(database = WikiPathways_2019_Mouse, p.adjust_cutoff = 0.05)
get_enriched_pathways_output <- ora_results@compareClusterResult %>% get_enriched_pathways(Description, database = WikiPathways_2019_Mouse)

ora_results <- genes %>% ora_analysis(database = Reactome_2016, p.adjust_cutoff = 0.05)
get_enriched_pathways_output <- ora_results@compareClusterResult %>% get_enriched_pathways(Description, database = Reactome_2016)

ora_results <- genes %>% ora_analysis(database = GO_Biological_Process_2018, p.adjust_cutoff = 0.05)
get_enriched_pathways_output <- ora_results@compareClusterResult %>% get_enriched_pathways(Description, database = GO_Biological_Process_2018)

# get_enriched_pathways_output <- enrichment_result %>% get_enriched_pathways(Description, pathways_list = custom_KEGG_Pathways)
# get_enriched_pathways_output <- enrichment_result %>% get_enriched_pathways(Description, pathways_list = custom_Reactome_Pathways)
# get_enriched_pathways_output <- enrichment_result %>% get_enriched_pathways(Description, pathways_list = custom_GO_Pathways)
# calculate_median_output <- set2$normalizedCounts %>% calculate_median(control_columns = dplyr::contains("CONTROL_TET"), compare_columns = dplyr::contains("LLAEVEN_TET"))
calculate_median_output <- filtered_countData_sub %>% calculate_median(control_columns = dplyr::contains("CTRL_"), compare_columns = dplyr::contains("TNF_"))

# genes_class <- calculate_median_output %>% classify_genes(DE_results = edgeR_results$LLAEVEN_TETxCONTROL_TET)
genes_class <- calculate_median_output %>% classify_genes(DE_results = edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper()))


BDIA_results <- genes_class %>% BDIA_bootstrap(get_enriched_pathways_output = get_enriched_pathways_output, no_cores = (detectCores() - 1))

BDIA_results <- set2$normalizedCounts %>% calculate_median(control_columns = dplyr::contains("CONTROL_TET"), compare_columns = dplyr::contains("LLAEVEN_TET")) %>% classify_genes(DE_results = edgeR_results$LLAEVEN_TETxCONTROL_TET) %>% BDIA_bootstrap(nDrawn = 50, nSample = 50, nLoop = 1000, get_enriched_pathways_output = get_enriched_pathways_output, no_cores = (detectCores() - 1))

BDIA_results <- filtered_countData_sub %>% calculate_median(control_columns = dplyr::contains("CTRL_"), compare_columns = dplyr::contains("TNF_")) %>% classify_genes(DE_results = edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper())) %>% BDIA_bootstrap(nDrawn = 50, nSample = 50, nLoop = 1000, get_enriched_pathways_output = get_enriched_pathways_output, no_cores = (detectCores() - 1), seed = 666)
```

```{r}
# KEGG -> "C3AR1"     "C8G"       "GRIN2D"    "GRK4"      "GSTA1"     "SERPINB1B" "SLC8A1"
candidate_genes_KEGG_1 <- map(seq_along(BDIA_results), function(i) BDIA_results[[i]]$only_new_DEGs) %>% unlist() %>% unique() %>% sort()
candidate_genes_KEGG_2 <- map(seq_along(BDIA_results), function(i) BDIA_results[[i]]$only_new_DEGs) %>% unlist() %>% unique() %>% sort()
candidate_genes_KEGG_3 <- map(seq_along(BDIA_results), function(i) BDIA_results[[i]]$only_new_DEGs) %>% unlist() %>% unique() %>% sort()

# Reactome -> "ARTN"   "C3AR1"  "GRIN2D" "NPFF"   "SLC8A1"
candidate_genes_reactome_1 <- map(seq_along(BDIA_results), function(i) BDIA_results[[i]]$only_new_DEGs) %>% unlist() %>% unique() %>% sort()
candidate_genes_reactome_2 <- map(seq_along(BDIA_results), function(i) BDIA_results[[i]]$only_new_DEGs) %>% unlist() %>% unique() %>% sort()

# GO BP -> "ARTN"    "C3AR1"   "CARD14"  "CHRM4"   "DOCK4"   "FABP3"   "GRIN2D"  "GSTA1"   "HRC"     "MGLL"    "NPFF"    "NPTX1"   "POLR3GL" "SHPK"    "SLC8A1"  "TFAP2A" 
candidate_genes_GO_BP_1 <- map(seq_along(BDIA_results), function(i) BDIA_results[[i]]$only_new_DEGs) %>% unlist() %>% unique() %>% sort()
```

# DE results
```{r}
edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper) %>% filter(rowname %in% candidate_genes_GO_BP_1)
```


# BDIA found genes
```{r}
candidate_genes_3 %>% plot_genes_boxplot(df = filtered_countData, design = Design$condition, levels = c("CTRL", "P4", "TNF", "TNF_P4"), converter = Conversion_Mouse2Human %>% mutate(SYMBOL = SYMBOL %>% toupper()), convert_from = "SYMBOL", convert_to = "ENSEMBL" )

candidate_genes_GO_BP_1 %>% plot_genes_boxplot(df = filtered_countData_sub, design = Design %>% filter(condition %in% c("CTRL", "TNF")) %>% .[["condition"]], levels = c("CTRL", "TNF"), converter = Conversion_Mouse2Human %>% mutate(SYMBOL = SYMBOL %>% toupper()), convert_from = "SYMBOL", convert_to = "ENSEMBL" )
```

# Comparison without vs with BDIA
```{r}
# KEGG
new_ora <- genes %>% map(c, candidate_genes_KEGG_1) %>% ora_analysis(database = KEGG_2019_Mouse, p.adjust_cutoff = 0.05)
plot_grid(ora_results %>% dotplot(), new_ora %>% dotplot(), labels = c('KEGG_without_BDIA', 'KEGG_with_BDIA'), label_size = 12)

# KEGG
new_ora <- genes %>% map(c, candidate_genes_reactome_1) %>% ora_analysis(database = Reactome_2016, p.adjust_cutoff = 0.05)
plot_grid(ora_results %>% dotplot(), new_ora %>% dotplot(), labels = c('Reactome_without_BDIA', 'Reactome_with_BDIA'), label_size = 12)

# GO BP
new_ora <- genes %>% map(c, candidate_genes_GO_BP_1) %>% ora_analysis(database = GO_Biological_Process_2018, p.adjust_cutoff = 0.05)
plot_grid(ora_results %>% dotplot(), new_ora %>% dotplot(), labels = c('GO_BP_without_BDIA', 'GO_BP_with_BDIA'), label_size = 12)
```

Obs: So far, functions were developed in S3.

# S4 approach
```{r}
library(methods)
BDIA <- setClass("BDIA", slots = c(countdata_input = "data.frame",
                                   median_df = "data.frame",
                                   genes_class = "list",
                                   input_genes = "vector",
                                   enriched_pathways = "list",
                                   ora_results_genes = "vector",
                                   ora_results_clusters_df = "data.frame",
                                   database = "vector",
                                   pathway_list = "list"))

# BDIA <- setClass("BDIA", slots = c(countdata_input = "data.frame", 
#                                    median_df = "data.frame"), contains = "compareClusterResult")

ora_analysis <- function(genes, database = KEGG_2019_Human, pvalue_cutoff = 1, qvalue_cutoff = 1, p.adjust_cutoff = 0.1, count_at_least = 3){
  library(clusterProfiler)
  gmt_list <- database %>% read.gmt()
  custom_pathways <- gmt_list %>% split(f = .$ont) %>% map(dplyr::select, gene) %>% map(pull)
  allgenes <- gmt_list[, "gene"] %>% unique()
  enrichment_result <- genes %>% map(clusterProfiler::enricher, pvalueCutoff = pvalue_cutoff, qvalueCutoff = qvalue_cutoff, universe = allgenes, TERM2GENE = gmt_list)
  clProf.df <- enrichment_result %>% map(as.data.frame) %>% bind_rows(.id = "Cluster")
  clProf.df <- clProf.df %>% filter(p.adjust < p.adjust_cutoff & Count >= count_at_least)
  geneClusters <- genes
  # compareCluster_output <- new("compareClusterResult", compareClusterResult = clProf.df, geneClusters = geneClusters, fun = "enricher", .call = match.call(expand.dots = TRUE))
  # return(compareCluster_output)
  ora_analysis_output <- new("BDIA", ora_results_clusters_df = clProf.df, ora_results_genes = geneClusters, database = database)
  return(ora_analysis_output)
}

# porcentagem de mediana / total = probabilidade

calculate_median_S4 <- function(countdata_input = countdata_input, control_columns = control_columns, compare_columns = compare_columns){
  library(edgeR)
  countdata_input <- filtered_countData %>% cpm() %>% as.data.frame %>% rownames_to_column()
  median_df <- data.frame(
  rowname = countdata_input %>% dplyr::select(rowname) %>% pull(),
  median_control = countdata_input %>% dplyr::select(c(control_columns)) %>% apply(MARGIN = 1, FUN = median),
  median_compare = countdata_input %>% dplyr::select(c(compare_columns)) %>% apply(MARGIN = 1, FUN = median) )
  median_df <- median_df %>% mutate(pis_control = median_control/(median_control %>% sum),
                                    pis_compare = median_compare/(median_compare %>% sum))
  median_df <- median_df %>% mutate(median_control = median_control %>% formattable::digits(digits = 2),
                                    median_compare = median_compare %>% formattable::digits(digits = 2),
                                    pis_control = pis_control %>% formattable::scientific(digits = 2),
                                    pis_compare = pis_compare %>% formattable::scientific(digits = 2))
  
  # calculate_median_output <- new("BDIA", countdata_input = countdata_input, median_df = median_df, fun = "calculate_median", .call = match.call(expand.dots = TRUE))
  calculate_median_output <- new("BDIA", countdata_input = countdata_input, median_df = median_df)
  return(calculate_median_output)
}

classify_genes_S4 <- function(calculate_median_output = calculate_median_output, DE_results = DE_results, FDR_HC_DEGs = 0.045, LFC_HC_DEGs = log2(2), FDR_limbo = 0.1, LFC_limbo = log2(1.4), min_median = 2) {
  median_df <- calculate_median_output@median_df
  countdata_input = calculate_median_output@countdata_input
  HC_DEGs <- DE_results %>% dplyr::filter(FDR < FDR_HC_DEGs & abs(logFC) >= LFC_HC_DEGs) %>% dplyr::select(rowname) %>% pull()
  # LowVar_df <- median_df %>% filter(median_control < min_median & median_compare < min_median) #%>% filter(median_control > 0 & median_compare > 0)
  #---- low Var cannot be DEGs ---------------------
  # LowVar_df <- LowVar_df %>% dplyr::slice(-(LowVar_df$rowname %in% HC_DEGs %>% which()))
  # midLayer <- LowVar_df %>% dplyr::select(rowname) %>% pull() %>% as.character()
  Limbo_df <- DE_results %>% dplyr::filter(FDR >= THRESHOLD & FDR < FDR_limbo & abs(logFC) >= LFC_limbo)
  Limbo_genes <- Limbo_df %>% dplyr::select(rowname) %>% pull()
  # cutoff_median <- LowVar_df %>% dplyr::select(median_compare) %>% pull() %>% median()
  # MiddleHighExp_df <- LowVar_df %>% filter(median_compare > cutoff_median)
  Not_DEGs <- DE_results %>% dplyr::filter(FDR > FDR_HC_DEGs & abs(logFC) < LFC_HC_DEGs) %>% dplyr::select(rowname) %>% pull() %>% as.character()
  median_hist <- median_df %>% dplyr::select(c(median_control, median_compare)) %>%  reshape2::melt(id.vars = NULL) %>% ggplot(aes(x = log10(value))) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", breaks = seq(from = -1.5, to = 0.5, by = .1)) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(median)") + ggtitle("All genes frequencies") + facet_wrap(~variable)
  HC_DEGs_hist <- median_df %>% filter(rowname %in% HC_DEGs) %>% dplyr::select(c(median_control, median_compare)) %>%  reshape2::melt(id.vars = NULL) %>% ggplot(aes(x = log10(value))) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", breaks = seq(from = -1.5, to = 0.5, by = .1)) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(median)") + ggtitle("High confidence DEGs frequencies") + facet_wrap(~variable)
  limbo_hist <- median_df %>% filter(rowname %in% Limbo_genes) %>% dplyr::select(c(median_control, median_compare)) %>%  reshape2::melt(id.vars = NULL) %>% ggplot(aes(x = log10(value))) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", breaks = seq(from = -1.5, to = 0.5, by = .1)) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(median)") + ggtitle("Limbo frequencies") + facet_wrap(~variable)
  genes_class_list <- list(
    HC_DEGs = HC_DEGs,
    Limbo_genes = Limbo_genes,
    Not_DEGs = Not_DEGs,
    histograms = list(median_hist = median_hist,
                      HC_DEGs_hist = HC_DEGs_hist,
                      limbo_hist = limbo_hist)
  )
  input_genes = c(genes_class_list$HC_DEGs, genes_class_list$Limbo_genes, genes_class_list$Not_DEGs) %>% unique()
  
  # BDIA_gene_classes <- setClass("BDIA_gene_classes", slots = c(
  #                                  genes_class = "list", 
  #                                  BDIA_input_genes = "vector"), contains = "BDIA")
  # BDIA_gene_classes <- new("BDIA_gene_classes", genes_class = genes_class_list, BDIA_input_genes = BDIA_input_genes, fun = "classify_genes", .call = match.call(expand.dots = TRUE))
  # BDIA_gene_classes <- new("BDIA", countdata_input = countdata_input, median_df = median_df, genes_class = genes_class_list, BDIA_input_genes = BDIA_input_genes, fun = "classify_genes", .call = match.call(expand.dots = TRUE))  
  classify_genes_output <- new("BDIA", genes_class = genes_class_list, input_genes = input_genes, median_df = median_df, countdata_input = countdata_input)

  # classify_genes_output <- list(
  #   histograms = median_hist,
  #   BDIA_gene_classes = BDIA_gene_classes
  # )
  return(classify_genes_output)
}

get_enriched_pathways_S4 <- function(ora_results = ora_results, Description = Description) {
  database = ora_results@database
  ora_results_clusters_df = ora_results@ora_results_clusters_df
  ora_results_genes = ora_results@ora_results_genes
  gmt_list <- database %>% read.gmt()
  pathways_list <- gmt_list %>% split(f = .$ont) %>% map(dplyr::select, gene) %>% map(pull)
  # pathways <- ora_results@compareClusterResult %>% dplyr::select(c(Description)) %>% pull() %>% as.list() %>% map(as.character())
  pathways <- ora_results@ora_results_clusters_df %>% dplyr::select(c(Description)) %>% pull() %>% as.list() %>% map(as.character())
  genes_in_pathway <- map(seq_along(pathways), function(i) pathways_list[[pathways[[i]]]] %>% sort())
  DEGs_found_in_pathway <- ora_results@ora_results_clusters_df %>% dplyr::select(geneID) %>% map(str_split, pattern = "/") %>% purrr::set_names(NULL) %>% .[[1]]
  n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% map(length)
  enriched_pathways <- vector("list", length = length(pathways)) %>% purrr::set_names(pathways)
  for (pathway in seq_along(enriched_pathways)) {
    enriched_pathways[[pathway]][["genes_in_pathway"]] <- genes_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["DEGs_found_in_pathway"]] <- DEGs_found_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["n_DEGs_found_in_pathway"]] <- n_DEGs_found_in_pathway[[pathway]]
  }

  # get_enriched_pathways_output  <- setClass("BDIA_enriched_pathways", slots = c(
  #                                                                           enriched_pathways = "list"                                                                                                          ), contains = "BDIA")
  get_enriched_pathways_output <- new("BDIA", ora_results_clusters_df = ora_results_clusters_df, ora_results_genes = ora_results_genes, enriched_pathways = enriched_pathways, database = database)
  return(get_enriched_pathways_output)
}

is.numeric0 <- function(x){is.numeric(x) && length(x) == 0}

cppFunction('Rcpp::NumericVector extract_pis_c(Rcpp::DataFrame df, Rcpp::StringVector gene_ids){
  NumericVector pis_compare = df["pis_compare"];
  StringVector x = df["rowname"];
  StringVector y = gene_ids;
  LogicalVector ind = in(x, y);
  return pis_compare[ind];
                                      }')

cppFunction('Rcpp::NumericVector log10_c(NumericVector x){
  NumericVector log10x = log10(x);
  return log10x;
                                      }')

conditionally_modify_pis <- function(pis = pis) {
  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8
  pis[is.numeric0(pis)] <- 1e-8
  pis %>% return()
}

calculate_pgs_log10_c <- function(df = df, gene_ids = gene_ids) {
  pgs <- df %>% extract_pis_c(gene_ids = gene_ids) %>% conditionally_modify_pis %>% log10_c %>% sum()
  return(pgs)
}


# Build df approach
BDIA_bootstrap_S4 <- function(classify_genes_S4_output = classify_genes_S4_output, get_enriched_pathways_S4_output = get_enriched_pathways_S4_output, nSample = 50, nDrawn = 50, nLoop = 1000, no_cores = (detectCores() - 1), seed = 123) {
  # profvis({
  library(parallelMap)
  # Initiate cluster
  set.seed(seed)
  parallelStartSocket(no_cores)
  countdata_input <- classify_genes_S4_output@countdata_input
  median_df <- classify_genes_S4_output@median_df
  genes_class <- classify_genes_S4_output@genes_class
  Limbo_genes <- genes_class[["Limbo_genes"]]
  HC_DEGs <- genes_class[["HC_DEGs"]]
  No_DEGs <- genes_class[["No_DEGs"]]
  input_genes <- classify_genes_S4_output@input_genes
  enriched_pathways <- get_enriched_pathways_S4_output@enriched_pathways
  ora_results_genes <- get_enriched_pathways_S4_output@ora_results_genes
  ora_results_clusters_df <- get_enriched_pathways_S4_output@ora_results_clusters_df
  database <- get_enriched_pathways_S4_output@database  
  pathway_list <- list()
  for (i in seq_along(enriched_pathways)) {
    pathway <- enriched_pathways %>% names() %>% .[i]
    print(pathway)
    genes_in_pathway <- enriched_pathways[[pathway]][["genes_in_pathway"]]
    DEGs_found_in_pathway <- enriched_pathways[[pathway]][["DEGs_found_in_pathway"]] %>% sort()
    n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% length()
    nGoodGenes <- input_genes %>% length()
    nAllGenes <- countdata_input %>% nrow()
    nBadGenes <- nAllGenes - nGoodGenes
    # No bootstrap for loop
    sample_genes_list <- replicate(n = nLoop, list(sample(input_genes, nSample, replace = FALSE)))
    genes_found <- sample_genes_list %>% map(function(x) x[x %in% enriched_pathways[[pathway]][["genes_in_pathway"]]])
    nFound <- genes_found %>% map(length)
    # Hypergeometric test
    pORA <- nFound %>% map(phyper, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    # log10.pGs <- sample_genes_list %>% map(extract_pis_c, df = median_df) %>% map(conditionally_modify_pis) %>% map(log10_c) %>% map(sum)
    log10.pGs <- sample_genes_list %>% map(calculate_pgs_log10_c, df = median_df)
    post <- map(seq_along(sample_genes_list), function(i) -log10(pORA[[i]]) - log10.pGs[[i]])
    prioriDist <- log10.pGs %>% unlist()

    BDIA_df <- data.frame(
      sample_genes = sample_genes_list %>% map(paste0, collapse = ", ") %>% unlist(),
      maxiGenes = genes_found %>% map(paste0, collapse = ", ") %>% unlist(),
      n_Found = nFound %>% unlist(),
      verossDist = pORA %>% unlist(),
      prioriDist = log10.pGs %>% unlist(),
      postDist = post %>% unlist(),
      stringsAsFactors = FALSE
    )

    postDist2_df <- BDIA_df$postDist %>%
      na.omit() %>%
      sort() %>%
      as.data.frame() %>%
      purrr::set_names("postDist2")
    postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", binwidth = 2) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
    L <- postDist2_df %>%
      pull() %>%
      length()
    up <- L * .95
    postDist2.sig <- postDist2_df$postDist2[1:up]
    limSup <- postDist2.sig %>% max()
    n <- which(postDist2_df$postDist2 < limSup)
    # genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
    genes <- map(seq_along(n), function(i) BDIA_df$maxiGenes[[n[i]]] %>% strsplit(", ")) %>%
      unlist() %>%
      unique() %>%
      sort()
    genes.found.hc.DEGs <- genes %>% .[. %in% HC_DEGs]
    genes.found.limbo <- genes %>% .[. %in% Limbo_genes]
    genes.found.no.DEGs <- genes %>% .[. %in% No_DEGs]
    prev.found <- DEGs_found_in_pathway %>% sort()
    found.DEGs.and.limbo <- c(genes.found.hc.DEGs, genes.found.limbo) %>%
      unique() %>%
      sort()
    common.symbols <- prev.found %>% .[. %in% found.DEGs.and.limbo]
    only.prev.found <- prev.found %>% .[!. %in% found.DEGs.and.limbo]
    only.new.found <- found.DEGs.and.limbo %>% .[!. %in% prev.found]
    # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
    llMu <- postDist2_df$postDist2 %>%
      mean() %>%
      round(digits = 2)
    minLL <- postDist2_df$postDist2 %>%
      min() %>%
      round(digits = 2)
    maxLL <- postDist2_df$postDist2 %>%
      max() %>%
      round(digits = 2)
    BDIA_df <- BDIA_df %>% mutate(
      verossDist = verossDist %>% format(digits = 2, scientific = TRUE),
      prioriDist = prioriDist %>% format(digits = 5),
      postDist = postDist %>% format(digits = 5)
    )
    pathway_list[[pathway]] <- list(
      genes_found_in_pathway = prev.found,
      genes_found_new_algorithm = genes,
      genes_found_DEGs_Limbo = found.DEGs.and.limbo,
      genes_found_in_HC_DEGs = genes.found.hc.DEGs,
      genes_found_in_Limbo = genes.found.limbo,
      genes_found_in_No_DEGs = genes.found.no.DEGs,
      common = common.symbols,
      only_previous_DEGs = only.prev.found,
      only_new_DEGs = only.new.found,
      HC_DEGs = HC_DEGs,
      Limbo_genes = Limbo_genes,
      No_DEGs = No_DEGs,
      Histogram = postDist2_gghist,
      BDIA_summary = data.frame("mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n_genes_found_in_pathway" = prev.found %>% length(), "n_genes_found_new_algorithm" = genes %>% length(), "n_genes_found_DEGs_limbo" = found.DEGs.and.limbo %>% length()),
      BDIA_df = BDIA_df
    )
  }
  BDIA_bootstrap_S4_output <- new("BDIA", countdata_input = countdata_input, median_df = median_df, genes_class = genes_class, input_genes = input_genes, ora_results_clusters_df = ora_results_clusters_df, ora_results_genes = ora_results_genes, enriched_pathways = enriched_pathways, database = database, pathway_list = pathway_list)
  return(BDIA_bootstrap_S4_output)
  parallelStop()
  # })
}

```

```{r}
genes <- edgeR_topDE2$TNFxCTRL$rowname %>% toupper() %>% list() %>% purrr::set_names("TNFxCTRL")
# genes <- edgeR_topDE2$TNF_P4xCTRL$rowname %>% toupper() %>% list() %>% purrr::set_names("TNF_P4xCTRL")

# ora_results <- genes %>% ora_analysis(database = KEGG_2019_Mouse, p.adjust_cutoff = 0.05)
ora_results <- genes %>% ora_analysis(database = Reactome_2016, p.adjust_cutoff = 0.05)
# ora_results <- genes %>% ora_analysis(database = GO_Biological_Process_2018, p.adjust_cutoff = 0.05)
# ora_results <- genes %>% ora_analysis(database = GO_Molecular_Function_2018, p.adjust_cutoff = 0.05)
# ora_results <- genes %>% ora_analysis(database = WikiPathways_2019_Mouse, p.adjust_cutoff = 0.05)


filtered_countData_sub <- filtered_countData %>% as.data.frame() %>% dplyr::select(-c(dplyr::contains("P4")))
calculate_median_S4_output <- calculate_median_S4(countdata_input = filtered_countData_sub, control_columns = dplyr::contains("CTRL_"), compare_columns = dplyr::contains("TNF_"))

DE_results <- edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper())
# DE_results <- edgeR_pipeline_results$results$`06hx0h`

classify_genes_S4_output <- classify_genes_S4(DE_results = DE_results, calculate_median_output = calculate_median_S4_output)

# get_enriched_pathways_output <- get_enriched_pathways(enrichment_result, Description, database = KEGG_2019_Human)
# get_enriched_pathways_output <- get_enriched_pathways(enrichment_result, Description, database = Reactome_2016)
get_enriched_pathways_S4_output <- get_enriched_pathways_S4(ora_results = ora_results, Description)

# genes_class <- calculate_median_output %>% classify_genes(DE_results = edgeR_results$LLAEVEN_TETxCONTROL_TET)
# genes_class <- calculate_median_output %>% classify_genes(DE_results = edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper()))

BDIA_results_S4 <- classify_genes_S4_output %>% BDIA_bootstrap_S4(get_enriched_pathways_S4_output = get_enriched_pathways_S4_output, no_cores = (detectCores() - 1))
```

```{r}
KEGG_genes <- BDIA_results_S4@pathway_list %>% map(function(x) x %>% .[["only_new_DEGs"]]) %>% unlist() %>% unique()
edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% KEGG_genes)
filtered_countData_sub %>% cpm() %>% as.data.frame() %>% rownames_to_column() %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% KEGG_genes)
```

```{r}
Reactome_genes <- BDIA_results_S4@pathway_list %>% map(function(x) x %>% .[["only_new_DEGs"]]) %>% unlist() %>% unique()
edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% Reactome_genes)
filtered_countData_sub %>% cpm() %>% as.data.frame() %>% rownames_to_column() %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% Reactome_genes)
```

```{r}
GO_BP_genes <- BDIA_results_S4@pathway_list %>% map(function(x) x %>% .[["only_new_DEGs"]]) %>% unlist() %>% unique()
edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% GO_BP_genes)
filtered_countData_sub %>% cpm() %>% as.data.frame() %>% rownames_to_column() %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% GO_BP_genes)
```

```{r}
WikiPathways_genes <- BDIA_results_S4@pathway_list %>% map(function(x) x %>% .[["only_new_DEGs"]]) %>% unlist() %>% unique()
edgeR_results$TNFxCTRL %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% WikiPathways_genes)
filtered_countData_sub %>% cpm() %>% as.data.frame() %>% rownames_to_column() %>% mutate(rowname = rowname %>% toupper()) %>% filter(rowname %in% WikiPathways_genes)
```

# R6 object oriented
# S4 approach
```{r}
library(R6)
BDIA <- R6Class("BDIA",
  private = list(database = KEGG_2019_Mouse),
  public = list(
    countdata_input = data.frame(),
    median_df = data.frame(),
    genes_class = list(),
    input_genes = vector(),
    enriched_pathways = list(),
    ora_results_genes = vector(),
    ora_results_clusters_df = data.frame(),
    pathway_list = list()
  )
)

# BDIA <- setClass("BDIA", slots = c(countdata_input = "data.frame", 
#                                    median_df = "data.frame"), contains = "compareClusterResult")

ora_analysis <- function(genes, database = KEGG_2019_Human, pvalue_cutoff = 1, qvalue_cutoff = 1, p.adjust_cutoff = 0.1, count_at_least = 3){
  library(clusterProfiler)
  gmt_list <- database %>% read.gmt()
  custom_pathways <- gmt_list %>% split(f = .$ont) %>% map(dplyr::select, gene) %>% map(pull)
  allgenes <- gmt_list[, "gene"] %>% unique()
  enrichment_result <- genes %>% map(clusterProfiler::enricher, pvalueCutoff = pvalue_cutoff, qvalueCutoff = qvalue_cutoff, universe = allgenes, TERM2GENE = gmt_list)
  clProf.df <- enrichment_result %>% map(as.data.frame) %>% bind_rows(.id = "Cluster")
  clProf.df <- clProf.df %>% filter(p.adjust < p.adjust_cutoff & Count >= count_at_least)
  geneClusters <- genes
  # compareCluster_output <- new("compareClusterResult", compareClusterResult = clProf.df, geneClusters = geneClusters, fun = "enricher", .call = match.call(expand.dots = TRUE))
  # return(compareCluster_output)
  ora_analysis_output <- new("BDIA", ora_results_clusters_df = clProf.df, ora_results_genes = geneClusters, database = database)
  return(ora_analysis_output)
}

calculate_median_R6 <- function(countdata_input = countdata_input, control_columns = control_columns, compare_columns = compare_columns){
  library(edgeR)
  countdata_input <- filtered_countData %>% cpm() %>% as.data.frame %>% rownames_to_column()
  median_df <- data.frame(
  rowname = countdata_input %>% dplyr::select(rowname) %>% pull(),
  median_control = countdata_input %>% dplyr::select(c(control_columns)) %>% apply(MARGIN = 1, FUN = median),
  median_compare = countdata_input %>% dplyr::select(c(compare_columns)) %>% apply(MARGIN = 1, FUN = median) )
  median_df <- median_df %>% mutate(pis_control = median_control/(median_control %>% sum),
                                    pis_compare = median_compare/(median_compare %>% sum))
  median_df <- median_df %>% mutate(median_control = median_control %>% formattable::digits(digits = 2),
                                    median_compare = median_compare %>% formattable::digits(digits = 2),
                                    pis_control = pis_control %>% formattable::scientific(digits = 2),
                                    pis_compare = pis_compare %>% formattable::scientific(digits = 2))
  
  # calculate_median_output <- new("BDIA", countdata_input = countdata_input, median_df = median_df, fun = "calculate_median", .call = match.call(expand.dots = TRUE))
  calculate_median_output <- new("BDIA", countdata_input = countdata_input, median_df = median_df)
  return(calculate_median_output)
}

classify_genes_R6 <- function(calculate_median_output = calculate_median_output, DE_results = DE_results, FDR_limbo = 0.07, LFC_limbo = log2(1.4)) {
  median_df <- calculate_median_output@median_df
  countdata_input = calculate_median_output@countdata_input
  HC_DEGs <- DE_results %>% dplyr::filter(FDR < THRESHOLD & abs(logFC) >= LFC) %>% dplyr::select(rowname) %>% pull()
  LowVar_df <- median_df %>% filter(median_control < 2 & median_compare < 2) %>% filter(median_control > 0 & median_compare > 0)
  #---- low Var cannot be DEGs ---------------------
  LowVar_df <- LowVar_df %>% dplyr::slice(-(LowVar_df$rowname %in% HC_DEGs %>% which()))
  midLayer <- LowVar_df %>% dplyr::select(rowname) %>% pull() %>% as.character()
  Limbo_df <- DE_results %>% dplyr::filter(FDR >= THRESHOLD & FDR < FDR_limbo & abs(logFC) > LFC_limbo)
  Limbo_genes <- Limbo_df %>% dplyr::select(rowname) %>% pull()
  cutoff_median <- LowVar_df %>% dplyr::select(median_compare) %>% pull() %>% median()
  MiddleHighExp_df <- LowVar_df %>% filter(median_compare > cutoff_median)
  Not_DEGs <- MiddleHighExp_df %>% dplyr::select(rowname) %>% pull() %>% as.character()
  median_hist <- LowVar_df %>%  dplyr::select(c(median_control, median_compare)) %>%  reshape2::melt(id.vars = NULL) %>% ggplot(aes(x = log10(value))) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", breaks = seq(from = -1.5, to = 0.5, by = .1)) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(median)") + ggtitle("Limbo frequencies") + facet_wrap(~variable)
  genes_class_list <- list(
    HC_DEGs = HC_DEGs,
    Limbo_genes = Limbo_genes,
    Not_DEGs = Not_DEGs,
    histograms = median_hist
  )
  input_genes = c(genes_class_list$HC_DEGs, genes_class_list$Limbo_genes, genes_class_list$Not_DEGs) %>% unique()
  
  # BDIA_gene_classes <- setClass("BDIA_gene_classes", slots = c(
  #                                  genes_class = "list", 
  #                                  BDIA_input_genes = "vector"), contains = "BDIA")
  # BDIA_gene_classes <- new("BDIA_gene_classes", genes_class = genes_class_list, BDIA_input_genes = BDIA_input_genes, fun = "classify_genes", .call = match.call(expand.dots = TRUE))
  # BDIA_gene_classes <- new("BDIA", countdata_input = countdata_input, median_df = median_df, genes_class = genes_class_list, BDIA_input_genes = BDIA_input_genes, fun = "classify_genes", .call = match.call(expand.dots = TRUE))  
  classify_genes_output <- new("BDIA", genes_class = genes_class_list, input_genes = input_genes, median_df = median_df, countdata_input = countdata_input)

  # classify_genes_output <- list(
  #   histograms = median_hist,
  #   BDIA_gene_classes = BDIA_gene_classes
  # )
  return(classify_genes_output)
}

get_enriched_pathways_R6 <- function(ora_results = ora_results, Description = Description) {
  database = ora_results@database
  ora_results_clusters_df = ora_results@ora_results_clusters_df
  ora_results_genes = ora_results@ora_results_genes
  gmt_list <- database %>% read.gmt()
  pathways_list <- gmt_list %>% split(f = .$ont) %>% map(dplyr::select, gene) %>% map(pull)
  # pathways <- ora_results@compareClusterResult %>% dplyr::select(c(Description)) %>% pull() %>% as.list() %>% map(as.character())
  pathways <- ora_results@ora_results_clusters_df %>% dplyr::select(c(Description)) %>% pull() %>% as.list() %>% map(as.character())
  genes_in_pathway <- map(seq_along(pathways), function(i) pathways_list[[pathways[[i]]]] %>% sort())
  DEGs_found_in_pathway <- ora_results@ora_results_clusters_df %>% dplyr::select(geneID) %>% map(str_split, pattern = "/") %>% purrr::set_names(NULL) %>% .[[1]]
  n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% map(length)
  enriched_pathways <- vector("list", length = length(pathways)) %>% purrr::set_names(pathways)
  for (pathway in seq_along(enriched_pathways)) {
    enriched_pathways[[pathway]][["genes_in_pathway"]] <- genes_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["DEGs_found_in_pathway"]] <- DEGs_found_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["n_DEGs_found_in_pathway"]] <- n_DEGs_found_in_pathway[[pathway]]
  }

  # get_enriched_pathways_output  <- setClass("BDIA_enriched_pathways", slots = c(
  #                                                                           enriched_pathways = "list"                                                                                                          ), contains = "BDIA")
  get_enriched_pathways_output <- new("BDIA", ora_results_clusters_df = ora_results_clusters_df, ora_results_genes = ora_results_genes, enriched_pathways = enriched_pathways, database = database)
  return(get_enriched_pathways_output)
}

is.numeric0 <- function(x){is.numeric(x) && length(x) == 0}

cppFunction('Rcpp::NumericVector extract_pis_c(Rcpp::DataFrame df, Rcpp::StringVector gene_ids){
  NumericVector pis_compare = df["pis_compare"];
  StringVector x = df["rowname"];
  StringVector y = gene_ids;
  LogicalVector ind = in(x, y);
  return pis_compare[ind];
                                      }')

cppFunction('Rcpp::NumericVector log10_c(NumericVector x){
  NumericVector log10x = log10(x);
  return log10x;
                                      }')

conditionally_modify_pis <- function(pis = pis) {
  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8
  pis[is.numeric0(pis)] <- 1e-8
  pis %>% return()
}

calculate_pgs_log10_c <- function(df = df, gene_ids = gene_ids) {
  pgs <- df %>% extract_pis_c(gene_ids = gene_ids) %>% conditionally_modify_pis %>% log10_c %>% sum()
  return(pgs)
}


# Build df approach
BDIA_bootstrap_R6 <- function(classify_genes_R6_output = classify_genes_R6_output, get_enriched_pathways_R6_output = get_enriched_pathways_R6_output, nSample = 50, nDrawn = 50, nLoop = 1000, no_cores = (detectCores() - 1), seed = 123) {
  # profvis({
  library(parallelMap)
  # Initiate cluster
  set.seed(seed)
  parallelStartSocket(no_cores)
  countdata_input <- classify_genes_R6_output@countdata_input
  median_df <- classify_genes_R6_output@median_df
  genes_class <- classify_genes_R6_output@genes_class
  Limbo_genes <- genes_class[["Limbo_genes"]]
  HC_DEGs <- genes_class[["HC_DEGs"]]
  No_DEGs <- genes_class[["No_DEGs"]]
  input_genes <- classify_genes_R6_output@input_genes
  enriched_pathways <- get_enriched_pathways_R6_output@enriched_pathways
  ora_results_genes <- get_enriched_pathways_R6_output@ora_results_genes
  ora_results_clusters_df <- get_enriched_pathways_R6_output@ora_results_clusters_df
  database <- get_enriched_pathways_R6_output@database  
  pathway_list <- list()
  for (i in seq_along(enriched_pathways)) {
    pathway <- enriched_pathways %>% names() %>% .[i]
    print(pathway)
    genes_in_pathway <- enriched_pathways[[pathway]][["genes_in_pathway"]]
    DEGs_found_in_pathway <- enriched_pathways[[pathway]][["DEGs_found_in_pathway"]] %>% sort()
    n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% length()
    nGoodGenes <- input_genes %>% length()
    nAllGenes <- countdata_input %>% nrow()
    nBadGenes <- nAllGenes - nGoodGenes
    # No bootstrap for loop
    sample_genes_list <- replicate(n = nLoop, list(sample(input_genes, nSample, replace = FALSE)))
    genes_found <- sample_genes_list %>% map(function(x) x[x %in% enriched_pathways[[pathway]][["genes_in_pathway"]]])
    nFound <- genes_found %>% map(length)
    # Hypergeometric test
    pORA <- nFound %>% map(phyper, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    # log10.pGs <- sample_genes_list %>% map(extract_pis_c, df = median_df) %>% map(conditionally_modify_pis) %>% map(log10_c) %>% map(sum)
    log10.pGs <- sample_genes_list %>% map(calculate_pgs_log10_c, df = median_df)
    post <- map(seq_along(sample_genes_list), function(i) -log10(pORA[[i]]) - log10.pGs[[i]])
    prioriDist <- log10.pGs %>% unlist()

    BDIA_df <- data.frame(
      sample_genes = sample_genes_list %>% map(paste0, collapse = ", ") %>% unlist(),
      maxiGenes = genes_found %>% map(paste0, collapse = ", ") %>% unlist(),
      n_Found = nFound %>% unlist(),
      verossDist = pORA %>% unlist(),
      prioriDist = log10.pGs %>% unlist(),
      postDist = post %>% unlist(),
      stringsAsFactors = FALSE
    )

    postDist2_df <- BDIA_df$postDist %>%
      na.omit() %>%
      sort() %>%
      as.data.frame() %>%
      purrr::set_names("postDist2")
    postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", binwidth = 2) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
    L <- postDist2_df %>%
      pull() %>%
      length()
    up <- L * .95
    postDist2.sig <- postDist2_df$postDist2[1:up]
    limSup <- postDist2.sig %>% max()
    n <- which(postDist2_df$postDist2 < limSup)
    # genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
    genes <- map(seq_along(n), function(i) BDIA_df$maxiGenes[[n[i]]] %>% strsplit(", ")) %>%
      unlist() %>%
      unique() %>%
      sort()
    genes.found.hc.DEGs <- genes %>% .[. %in% HC_DEGs]
    genes.found.limbo <- genes %>% .[. %in% Limbo_genes]
    genes.found.no.DEGs <- genes %>% .[. %in% No_DEGs]
    prev.found <- DEGs_found_in_pathway %>% sort()
    found.DEGs.and.limbo <- c(genes.found.hc.DEGs, genes.found.limbo) %>%
      unique() %>%
      sort()
    common.symbols <- prev.found %>% .[. %in% found.DEGs.and.limbo]
    only.prev.found <- prev.found %>% .[!. %in% found.DEGs.and.limbo]
    only.new.found <- found.DEGs.and.limbo %>% .[!. %in% prev.found]
    # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
    llMu <- postDist2_df$postDist2 %>%
      mean() %>%
      round(digits = 2)
    minLL <- postDist2_df$postDist2 %>%
      min() %>%
      round(digits = 2)
    maxLL <- postDist2_df$postDist2 %>%
      max() %>%
      round(digits = 2)
    BDIA_df <- BDIA_df %>% mutate(
      verossDist = verossDist %>% format(digits = 2, scientific = TRUE),
      prioriDist = prioriDist %>% format(digits = 5),
      postDist = postDist %>% format(digits = 5)
    )
    pathway_list[[pathway]] <- list(
      genes_found_in_pathway = prev.found,
      genes_found_new_algorithm = genes,
      genes_found_DEGs_Limbo = found.DEGs.and.limbo,
      genes_found_in_HC_DEGs = genes.found.hc.DEGs,
      genes_found_in_Limbo = genes.found.limbo,
      genes_found_in_No_DEGs = genes.found.no.DEGs,
      common = common.symbols,
      only_previous_DEGs = only.prev.found,
      only_new_DEGs = only.new.found,
      HC_DEGs = HC_DEGs,
      Limbo_genes = Limbo_genes,
      No_DEGs = No_DEGs,
      Histogram = postDist2_gghist,
      BDIA_summary = data.frame("mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n_genes_found_in_pathway" = prev.found %>% length(), "n_genes_found_new_algorithm" = genes %>% length(), "n_genes_found_DEGs_limbo" = found.DEGs.and.limbo %>% length()),
      BDIA_df = BDIA_df
    )
  }
  BDIA_bootstrap_R6_output <- new("BDIA", countdata_input = countdata_input, median_df = median_df, genes_class = genes_class, input_genes = input_genes, ora_results_clusters_df = ora_results_clusters_df, ora_results_genes = ora_results_genes, enriched_pathways = enriched_pathways, database = database, pathway_list = pathway_list)
  return(BDIA_bootstrap_R6_output)
  parallelStop()
  # })
}

```
























# Loop over pathways enrichment results
```{r}
library(profvis)
profvis({
nSample <- nDrawn <- 50
nLoop <- 1000

comparing <- "LLAEVEN_TETxCONTROL_TET"

  # Calculate the number of cores
  no_cores <- detectCores() - 1
  # Initiate cluster
  parallelStartSocket(no_cores)

    listPathways <- list()
for (i in 1:nrow(enrichment_result)) {
  
  pathway <- enrichment_result$Description[i] %>% as.character()
  print(pathway)
  genes.in.pathway <- custom_KEGG_Pathways[[pathway]] %>% sort()
  genes.found.in.pathway <- enrichment_result %>% filter(Description == pathway) %>% dplyr::select(geneID) %>% str_split(pattern = "/") %>% unlist() %>% as.character()
  nGenesFoundInPathway <- genes.found.in.pathway %>% length()

    all.sel.gene_ids <- c(HC_DEGs, Limbo, NO.DEGs) %>% unique()
  nGoodGenes <- all.sel.gene_ids %>% length()
  nAllGenes <- countdata_input$rowname %>% length()
  nBadGenes <- nAllGenes - nGoodGenes

  #------------------------------------------------------------------------------------.
  #--- bootstrap -- bayesian ----------------------------------------------------------
  #------------------------------------------------------------------------------------.

  postDist <- c()
  prioriDist <- c()
  verossDist <- c()
  maxiGenes <- c()
  sel.genes <- c()

  for (r in 1:nLoop) {
    # print(r)
    #-- sample hk --------------------------------
    sample.symbols <- sample(all.sel.gene_ids, nSample, replace = FALSE)
    genes.found <- sample.symbols[sample.symbols %in% genes.in.pathway]
    nFound <- genes.found %>% length()
    # Hypergeometric test
    pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    verossDist <- c(verossDist, pORA)
    log10.pGs <- median_df %>% calculate_pgs_log10(gene_ids = sample.gene_ids, column = pis_2)
    post <- -log10(pORA) - log10.pGs
    prioriDist <- c(prioriDist, log10.pGs)
    postDist <- c(postDist, post)
    maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
  }
  #------------- log -------------------
  # postDist = postDist/sum(postDist)
  postDist2 <- postDist %>% na.omit() %>% sort()
  # postDist2_hist <- postDist2 %>% hist(breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")
  postDist2_df = data.frame(postDist2 = postDist2)
  postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y=..density..), colour="black", fill="white", binwidth = 2) + geom_density(alpha=.2, fill="#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
  postDist2_hist$xname <- pathway
  L <- postDist2 %>% length()
  up <- L * .95
  postDist2.sig <- postDist2[1:up]
  limSup <- postDist2.sig %>% max()
  n <- which(postDist2 < limSup)
  genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
  genes.found.hc.degs <- genes %>% .[. %in% HC_DEGs]
  genes.found.limbo <- genes %>% .[. %in% Limbo]
  genes.found.no.degs <- genes %>% .[. %in% NO.DEGs]
  prev.found <- genes.found.in.pathway %>% sort()
  found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo) %>% unique() %>% sort()
  common.symbols <- prev.found %>% .[. %in% found.degs.and.limbo]
  only.prev.found <- prev.found %>% .[!. %in% found.degs.and.limbo]
  only.new.found <- found.degs.and.limbo %>% .[!. %in% prev.found]
  # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
  sel.genes <- c(sel.genes, only.new.found)
  llMu <- postDist2 %>% mean() %>% round(digits = 2)
  minLL <- postDist2 %>% min() %>% round(digits = 2)
  maxLL <- postDist2 %>% max() %>% round(digits = 2)

  listPathways[[pathway]] <- list()
  listPathways[[pathway]][["genes.found.in.pathway"]] <- prev.found
  listPathways[[pathway]][["genes.found.new.algorithm"]] <- genes
  listPathways[[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
  listPathways[[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
  listPathways[[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
  listPathways[[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
  listPathways[[pathway]][["common"]] <- common.symbols
  listPathways[[pathway]][["only.previous.degs"]] <- only.prev.found
  listPathways[[pathway]][["only.new.degs"]] <- only.new.found
  listPathways[[pathway]][["ll.posteriori"]] <- postDist
  listPathways[[pathway]][["HC_DEGs"]] <- HC_DEGs
  listPathways[[pathway]][["Limbo"]] <- Limbo
  listPathways[[pathway]][["NO_DEGs"]] <- NO.DEGs
  listPathways[[pathway]][["Histogram"]] <- postDist2_gghist
  listPathways[[pathway]][["BDIA_summary"]] <- data.frame( "mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n.genes.found.in.pathway" = prev.found %>% length(), "n.genes.found.new.algorithm" = genes %>% length(), "n.genes.found.degs.limbo" = found.degs.and.limbo %>% length())
}
  
parallelStop()
})


map(seq_along(listPathways), function(i) listPathways[[i]]$Histogram)
map(seq_along(listPathways), function(i) listPathways[[i]]$only.new.degs)

sel.genes <- sel.genes %>% unique
    

# dfLFC5 <- dfLFC5[dfLFC5$gene.id %in% all.sel.gene_ids, ]
#     dfAux <- dfLFC5[dfLFC5$symbol %in% sel.genes, ]
#     dfLFC.bayes <- rbind(dfLFC, dfAux)
#     length(sel.genes)
#     nrow(dfLFC)
#     length(sel.genes) + nrow(dfLFC)
#     #-- may have parologs ------.
#     nrow(dfLFC.bayes)
# 
#     #-------- dfLFC.bayes does not allow repeated symbols ---------
#     dfLFC.bayes <- dfLFC.bayes[order(dfLFC.bayes$symbol, dfLFC.bayes$fdr), ]
#     # dfLFC.bayes[, c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
#     dfLFC.bayes[dfLFC.bayes$symbol == "CXCL1", ]
#     nrow(dfLFC.bayes)
#     # dfAux2 = dfLFC.bayes
#     keep <- c()
#     previous <- ""
#     for (m in 1:nrow(dfLFC.bayes)) {
#       symbol <- dfLFC.bayes[m, "symbol"]
#       if (symbol != previous) {
#         previous <- symbol
#         keep <- c(keep, m)
#       }
#     }
```


# Test without for loop for pathways (v2)
```{r}
get_enriched_pathways_v2 <- function(enrichment_result = enrichment_result, Description = Description, pathways_list = pathways_list) {
  pathways <- enrichment_result %>% dplyr::select(c(Description)) %>% pull() %>% as.list() %>% map(as.character())
  genes_in_pathway <- map(seq_along(pathways), function(i) custom_KEGG_Pathways[[pathways[[i]]]] %>% sort())
  DEGs_found_in_pathway <- enrichment_result %>% dplyr::select(geneID) %>% map(str_split, pattern = "/") %>% purrr::set_names(NULL) %>% .[[1]]
  n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% map(length)
  enriched_pathways <- vector("list", length = length(pathways)) %>% purrr::set_names(pathways)
  for (pathway in seq_along(enriched_pathways)) {
    enriched_pathways[[pathway]][["genes_in_pathway"]] <- genes_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["DEGs_found_in_pathway"]] <- DEGs_found_in_pathway[[pathway]]
    enriched_pathways[[pathway]][["n_DEGs_found_in_pathway"]] <- n_DEGs_found_in_pathway[[pathway]]
  }
  # map(seq_along(enriched_pathways), function(i){
  #   enriched_pathways[[i]][["genes_in_pathway"]] <- genes_in_pathway[[i]]
  #   enriched_pathways[[i]][["DEGs_found_in_pathway"]] <- DEGs_found_in_pathway[[i]]
  #   enriched_pathways[[i]][["n_DEGs_found_in_pathway"]] <- n_DEGs_found_in_pathway[[i]]
  # })
  return(enriched_pathways)
}

get_enriched_pathways_v2_output <- get_enriched_pathways_v2(enrichment_result, Description, pathways_list = custom_KEGG_Pathways)

get_enriched_pathways_v2_output <- get_enriched_pathways_v2_output[[1]]

BDIA_bootstrap_v2 <- function(classify_genes_output = classify_genes_output, nSample = 50, nDrawn = 50, nLoop = 1000, get_enriched_pathways_v2_output = get_enriched_pathways_v2_output, no_cores = (detectCores() - 1)) {
  # profvis({
  library(parallelMap)
  # Initiate cluster
  parallelStartSocket(no_cores)
  countdata_input <- classify_genes_output[["countdata_input"]]
  input_genes <- classify_genes_output[["BDIA_input_genes"]]
  Limbo_genes <- classify_genes_output[["Limbo_genes"]]
  HC_DEGs <- classify_genes_output[["HC_DEGs"]]
  No_DEGs <- classify_genes_output[["No_DEGs"]]
  median_df <- classify_genes_output[["median_df"]]
  pathway_list <- list()
  # for (i in seq_along(get_enriched_pathways_v2_output[["genes_in_pathway"]])) {
  pathway <- get_enriched_pathways_v2_output %>% names() %>% .[[1]]
  print(pathway)
  genes_in_pathway <- get_enriched_pathways_v2_output[[pathway]][["genes_in_pathway"]]
  DEGs_found_in_pathway <- get_enriched_pathways_v2_output[[pathway]] [["DEGs_found_in_pathway"]] %>% sort()
  n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% length()
  nGoodGenes <- input_genes %>% length()
  nAllGenes <- countdata_input %>% nrow()
  nBadGenes <- nAllGenes - nGoodGenes
  # No bootstrap for loop
  sample_genes_list <- replicate(n = nLoop, list(sample(input_genes, nSample, replace = FALSE)))
  genes_found <- sample_genes_list %>% map(function(x) x[x %in% get_enriched_pathways_v2_output[["genes_in_pathway"]][[pathway]]])
  nFound <- genes_found %>% map(length)
  # Hypergeometric test
  pORA <- nFound %>% map(phyper, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
  log10.pGs <- sample_genes_list %>% lapply(calculate_pgs_log10, df = median_df, column = pis_compare)
  post <- map(seq_along(sample_genes_list), function(i) -log10(pORA[[i]]) - log10.pGs[[i]])
  prioriDist <- log10.pGs %>% unlist()

  BDIA_df <- data.frame(
    sample_genes = sample_genes_list %>% map(paste0, collapse = ", ") %>% unlist(),
    maxiGenes = genes_found %>% map(paste0, collapse = ", ") %>% unlist(),
    n_Found = nFound %>% unlist(),
    verossDist = pORA %>% unlist(),
    prioriDist = log10.pGs %>% unlist(),
    postDist = post %>% unlist()
  )

  postDist2_df <- BDIA_df$postDist %>% na.omit() %>% sort() %>% as.data.frame() %>% rename("." = "postDist2")
  postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", binwidth = 2) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
  L <- postDist2_df %>% pull() %>% length()
  up <- L * .95
  postDist2.sig <- postDist2_df$postDist2[1:up]
  limSup <- postDist2.sig %>% max()
  n <- which(postDist2_df$postDist2 < limSup)
  # genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
  genes <- map(seq_along(n), function(i) BDIA_df$maxiGenes[[n[i]]] %>% strsplit(", ")) %>% unlist() %>% unique() %>% sort()
  genes.found.hc.DEGs <- genes %>% .[. %in% HC_DEGs]
  genes.found.limbo <- genes %>% .[. %in% Limbo_genes]
  genes.found.no.DEGs <- genes %>% .[. %in% No_DEGs]
  prev.found <- DEGs_found_in_pathway %>% sort()
  found.DEGs.and.limbo <- c(genes.found.hc.DEGs, genes.found.limbo) %>% unique() %>% sort()
  common.symbols <- prev.found %>% .[. %in% found.DEGs.and.limbo]
  only.prev.found <- prev.found %>% .[!. %in% found.DEGs.and.limbo]
  only.new.found <- found.DEGs.and.limbo %>% .[!. %in% prev.found]
  # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
  llMu <- postDist2_df$postDist2 %>% mean() %>% round(digits = 2)
  minLL <- postDist2_df$postDist2 %>% min() %>% round(digits = 2)
  maxLL <- postDist2_df$postDist2 %>% max() %>% round(digits = 2)
  pathway_list[[pathway]] <- list()
  pathway_list[[pathway]][["genes_found_in_pathway"]] <- prev.found
  pathway_list[[pathway]][["genes_found_new_algorithm"]] <- genes
  pathway_list[[pathway]][["genes_found_DEGs_Limbo"]] <- found.DEGs.and.limbo
  pathway_list[[pathway]][["genes_found_in_HC_DEGs"]] <- genes.found.hc.DEGs
  pathway_list[[pathway]][["genes_found_in_Limbo"]] <- genes.found.limbo
  pathway_list[[pathway]][["genes_found_in_No_DEGs"]] <- genes.found.no.DEGs
  pathway_list[[pathway]][["common"]] <- common.symbols
  pathway_list[[pathway]][["only_previous_DEGs"]] <- only.prev.found
  pathway_list[[pathway]][["only_new_DEGs"]] <- only.new.found
  # pathway_list[[pathway]][["ll_posteriori"]] <- postDist
  pathway_list[[pathway]][["HC_DEGs"]] <- HC_DEGs
  pathway_list[[pathway]][["Limbo_genes"]] <- Limbo_genes
  pathway_list[[pathway]][["No_DEGs"]] <- No_DEGs
  pathway_list[[pathway]][["Histogram"]] <- postDist2_gghist
  pathway_list[[pathway]][["BDIA_summary"]] <- data.frame("mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n_genes_found_in_pathway" = prev.found %>% length(), "n_genes_found_new_algorithm" = genes %>% length(), "n_genes_found_DEGs_limbo" = found.DEGs.and.limbo %>% length())
  pathway_list[[pathway]][["BDIA_df"]] <- BDIA_df
  # }
  return(pathway_list)
  parallelStop()
  # })
}

sub_get_enriched_pathways_output <- get_enriched_pathways_output$KEGG_SPLICEOSOME
names(sub_get_enriched_pathways_output)

BDIA_results_v2 <- get_enriched_pathways_output$KEGG_SPLICEOSOME %>% BDIA_bootstrap_v2(classify_genes_output = genes_class, nDrawn = 50, nSample = 50, nLoop = 1000, no_cores = (detectCores() - 1))
```

```{r}
BDIA_bootstrap <- function(input_genes = input_genes, nSample = 50, nDrawn = 50, countdata_input = countdata_input, median_df = median_df, nLoop = 1000, get_enriched_pathways_output = get_enriched_pathways_output) {
				
profvis({
  library(parallelMap)
  # Calculate the number of cores
  no_cores <- detectCores() - 1
  # Initiate cluster
  parallelStartSocket(no_cores)
  listPathways <- list()
  for (i in seq_along(get_enriched_pathways_output[["genes_in_pathway"]])) {
    pathway <- get_enriched_pathways_output[["genes_in_pathway"]] %>% names() %>% .[i]
    print(pathway)
    genes_in_pathway <- get_enriched_pathways_output[["genes_in_pathway"]][[pathway]]
    DEGs_found_in_pathway <- get_enriched_pathways_output[["DEGs_found_in_pathway"]][[pathway]] %>% sort()
    n_DEGs_found_in_pathway <- DEGs_found_in_pathway %>% length()
    nGoodGenes <- input_genes %>% length()
    nAllGenes <- countdata_input %>% nrow()
    nBadGenes <- nAllGenes - nGoodGenes
    prioriDist <- postDist <- verossDist <- maxiGenes <- sel.genes <- c()
    for (r in 1:nLoop) {
      print(r)
      sample_genes <- sample(input_genes, nSample, replace = FALSE)
      genes.found <- sample_genes[sample_genes %in% get_enriched_pathways_output[["genes_in_pathway"]][[pathway]]]
      nFound <- genes.found %>% length()
      # Hypergeometric test
      pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
      verossDist <- c(verossDist, pORA)
      log10.pGs <- median_df %>% calculate_pgs_log10(gene_ids = sample_genes, column = pis_2)
      post <- -log10(pORA) - log10.pGs
      prioriDist <- c(prioriDist, log10.pGs)
      postDist <- c(postDist, post)
      maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
    }
    postDist2 <- postDist %>% na.omit() %>% sort()
    # postDist2_hist <- postDist2 %>% hist(breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")
    # postDist2_hist$xname <- pathway
    postDist2_df <- data.frame(postDist2 = postDist2)
    postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y = ..density..), colour = "black", fill = "white", binwidth = 2) + geom_density(alpha = .2, fill = "#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
    L <- postDist2 %>% length()
    up <- L * .95
    postDist2.sig <- postDist2[1:up]
    limSup <- postDist2.sig %>% max()
    n <- which(postDist2 < limSup)
    genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
    genes.found.hc.degs <- genes %>% .[. %in% HC_DEGs]
    genes.found.limbo <- genes %>% .[. %in% Limbo]
    genes.found.no.degs <- genes %>% .[. %in% NO.DEGs]
    prev.found <- DEGs_found_in_pathway %>% sort()
    found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo) %>% unique() %>% sort()
    common.symbols <- prev.found %>% .[. %in% found.degs.and.limbo]
    only.prev.found <- prev.found %>% .[!. %in% found.degs.and.limbo]
    only.new.found <- found.degs.and.limbo %>% .[!. %in% prev.found]
    # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
    sel.genes <- c(sel.genes, only.new.found)
    llMu <- postDist2 %>% mean() %>% round(digits = 2)
    minLL <- postDist2 %>% min() %>% round(digits = 2)
    maxLL <- postDist2 %>% max() %>% round(digits = 2)

    listPathways[[pathway]] <- list()
    listPathways[[pathway]][["genes.found.in.pathway"]] <- prev.found
    listPathways[[pathway]][["genes.found.new.algorithm"]] <- genes
    listPathways[[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
    listPathways[[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
    listPathways[[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
    listPathways[[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
    listPathways[[pathway]][["common"]] <- common.symbols
    listPathways[[pathway]][["only.previous.degs"]] <- only.prev.found
    listPathways[[pathway]][["only.new.degs"]] <- only.new.found
    listPathways[[pathway]][["ll.posteriori"]] <- postDist
    listPathways[[pathway]][["HC_DEGs"]] <- HC_DEGs
    listPathways[[pathway]][["Limbo"]] <- Limbo
    listPathways[[pathway]][["NO_DEGs"]] <- NO.DEGs
    listPathways[[pathway]][["Histogram"]] <- postDist2_gghist
    listPathways[[pathway]][["BDIA_summary"]] <- data.frame("mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n.genes.found.in.pathway" = prev.found %>% length(), "n.genes.found.new.algorithm" = genes %>% length(), "n.genes.found.degs.limbo" = found.degs.and.limbo %>% length())
  }
  return(listPathways)
})
}

BDIA_results <- BDIA_bootstrap(input_genes = all.sel.gene_ids, nSample = 50, nLoop = 1000, countdata_input = countdata_input, median_df = median_df, get_enriched_pathways_output = a)

  # Calculate the number of cores
  no_cores <- detectCores() - 1
  # Initiate cluster
  parallelStartSocket(no_cores) 
  


parallelStop()

  for (r in 1:nLoop) {
    # print(r)
    #-- sample hk --------------------------------
    sample.symbols <- sample(all.sel.gene_ids, nSample, replace = FALSE)
    genes.found <- sample.symbols[sample.symbols %in% genes.in.pathway]
    nFound <- genes.found %>% length()
    # Hypergeometric test
    pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    verossDist <- c(verossDist, pORA)
    log10.pGs <- median_df %>% calculate_pgs_log10(gene_ids = sample.gene_ids, column = pis_2)
    post <- -log10(pORA) - log10.pGs
    prioriDist <- c(prioriDist, log10.pGs)
    postDist <- c(postDist, post)
    maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
  }
  #------------- log -------------------
  # postDist = postDist/sum(postDist)
  postDist2 <- postDist %>% na.omit() %>% sort()
  # postDist2_hist <- postDist2 %>% hist(breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")
  # postDist2_hist$xname <- pathway
  postDist2_df = data.frame(postDist2 = postDist2)
  postDist2_gghist <- postDist2_df %>% ggplot(aes(x = postDist2)) + geom_histogram(aes(y=..density..), colour="black", fill="white", binwidth = 2) + geom_density(alpha=.2, fill="#377EB8") + xlab("-log10(posterior distribution)") + ggtitle(pathway)
  L <- postDist2 %>% length()
  up <- L * .95
  postDist2.sig <- postDist2[1:up]
  limSup <- postDist2.sig %>% max()
  n <- which(postDist2 < limSup)
  genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()
  genes.found.hc.degs <- genes %>% .[. %in% HC_DEGs]
  genes.found.limbo <- genes %>% .[. %in% Limbo]
  genes.found.no.degs <- genes %>% .[. %in% NO.DEGs]
  prev.found <- genes.found.in.pathway %>% sort()
  found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo) %>% unique() %>% sort()
  common.symbols <- prev.found %>% .[. %in% found.degs.and.limbo]
  only.prev.found <- prev.found %>% .[!. %in% found.degs.and.limbo]
  only.new.found <- found.degs.and.limbo %>% .[!. %in% prev.found]
  # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?
  sel.genes <- c(sel.genes, only.new.found)
  llMu <- postDist2 %>% mean() %>% round(digits = 2)
  minLL <- postDist2 %>% min() %>% round(digits = 2)
  maxLL <- postDist2 %>% max() %>% round(digits = 2)

  listPathways[[pathway]] <- list()
  listPathways[[pathway]][["genes.found.in.pathway"]] <- prev.found
  listPathways[[pathway]][["genes.found.new.algorithm"]] <- genes
  listPathways[[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
  listPathways[[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
  listPathways[[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
  listPathways[[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
  listPathways[[pathway]][["common"]] <- common.symbols
  listPathways[[pathway]][["only.previous.degs"]] <- only.prev.found
  listPathways[[pathway]][["only.new.degs"]] <- only.new.found
  listPathways[[pathway]][["ll.posteriori"]] <- postDist
  listPathways[[pathway]][["HC_DEGs"]] <- HC_DEGs
  listPathways[[pathway]][["Limbo"]] <- Limbo
  listPathways[[pathway]][["NO_DEGs"]] <- NO.DEGs
  listPathways[[pathway]][["Histogram"]] <- postDist2_gghist
  listPathways[[pathway]][["BDIA_summary"]] <- data.frame( "mu" = llMu, "minLL" = minLL, "maxLL" = maxLL, "n.genes.found.in.pathway" = prev.found %>% length(), "n.genes.found.new.algorithm" = genes %>% length(), "n.genes.found.degs.limbo" = found.degs.and.limbo %>% length())

```

```{r}
 #----------- Enriched Pathways db = KEGG -------------------------
    df.pathways <- open_Enrichment.string.table(comparing, db, 0)
    colnames(df.pathways)
    df.pathways[1:7, 1:9]

    #----------- Loop -------------------------
    stri <- paste("\n\nBegining loop of pathways:\n", sep = "")
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    # i = which(df.pathways$pathway == "PI3K-Akt signaling pathway")
    # i = which(df.pathways$pathway == "Dopaminergic synapse")

    degs.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% hc.degs, "symbol"])
    limbo.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% limbo, "symbol"])

    #------------------  loop second pathways  ---------------------------------
     #------------------  loop second pathways  ---------------------------------
    cat("loop pathways \n")
    sel.genes <- c()
    i <- 1
    for (i in 1:nrow(df.pathways)) { #  1:3

      dfPath <- df.pathways[i, ]
      pathway <- as.character(dfPath$pathway)
      stri_ini <- sprintf(">> %s - for %s:\n", pathway, comparing)
      stri_ini

      genes.in.pathway <- as.character(dfPath$genes.path)
      genes.in.pathway <- strsplit(genes.in.pathway, ";")[[1]]
      genes.in.pathway <- genes.in.pathway[order(genes.in.pathway)]
      genes.in.pathway
      if (length(genes.in.pathway) == 0) next

      # n.genes.found.in.pathway = as.character(dfPath$n.genes.found)
      genes.found.in.pathway <- as.character(dfPath$genes.found)
      nGenesFoundInPathway <- length(genes.found.in.pathway)

      all.sel.gene_ids <- c(hc.degs, limbo, no.degs)

      all.sel.gene_ids <- all.sel.gene_ids[all.sel.gene_ids != ""]
      all.sel.gene_ids <- unique(all.sel.gene_ids)
      length(all.sel.gene_ids)
      c(length(hc.degs), length(no.degs), length(limbo), length(all.sel.gene_ids))

      # length(midLayer)
      postDist <- c()
      prioriDist <- c()
      verossDist <- c()
      maxiGenes <- c()
      nAllGenes <- length(all.symbols)

      nGoodGenes <- length(all.sel.gene_ids)
      nBadGenes <- nAllGenes - nGoodGenes

      if (i == 1) {
        stri <- sprintf("<<--->>\n\nThere are %d High Confident DEGs.\n", nHC.genes)
        stri <- sprintf("%sThere are %d medium confident DEGs (limbo).\n", stri, length(limbo))
        stri <- sprintf("%sThere are %d 'housekeeping' genes.\n", stri, length(midLayer))
        stri <- sprintf("%sThere are %d 'good genes' = hc.degs + limbo + HK.\n", stri, nGoodGenes)
        stri <- sprintf("%sWe will draw a maximum of %d genes.\n", stri, nSample)
        stri <- sprintf("%sTherefore, there are %d 'bad genes'.\n", stri, nBadGenes)
        stri <- sprintf("%sWe will perform %d loop as bootstrap'.\n", stri, nLoop)
        stri <- sprintf("%sOnly enrichment with %d or more genes will be selected'.\n", stri, min.EnrGenes)
        stri <- sprintf("%s\n\n%s", stri, stri_ini)
      } else {
        stri <- sprintf("<<--->>\n\n%s", stri_ini)
      }
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      #------------------------------------------------------------------------------------.
      #--- bootstrap -- bayesian ----------------------------------------------------------
      #------------------------------------------------------------------------------------.

      for (r in 1:nLoop) {
        #-- sample hk --------------------------------
        sample.gene_ids <- sample(all.sel.gene_ids, nSample, replace = FALSE)
        sample.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% sample.gene_ids, "symbol"])
        sample.symbols <- unique(sample.symbols)
        sample.symbols <- sample.symbols[sample.symbols != ""]
        sample.symbols <- sample.symbols[order(sample.symbols)]
        length(sample.symbols)

        nDrawn <- length(sample.symbols)

        # length(sample.symbols %in% all.symbols)

        genes.found <- sample.symbols[sample.symbols %in% genes.in.pathway]
        nFound <- length(genes.found)

        if (nFound < min.EnrGenes) {
          verossDist <- c(verossDist, NA)
          prioriDist <- c(prioriDist, NA)
          postDist <- c(postDist, NA)
          maxiGenes <- c(maxiGenes, NA)
          next
        }

        # ?phyper
        pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
        verossDist <- c(verossDist, pORA)

        #----------- log10(p(Gset)) -------------
        thisCol <- colCase
        log10.pGs <- calc.pGs.log10(dfPis, sample.gene_ids, thisCol = colCase)
        # print(log10.pGs)

        # post = -log10(pORA) - log10.pGs - (-log10(pORA + not.pORA))
        post <- -log10(pORA) - log10.pGs

        prioriDist <- c(prioriDist, log10.pGs)
        postDist <- c(postDist, post)
        maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
      }
      # postDist = postDist/sum(postDist)

      #------------- log -------------------

      # postDist = postDist/sum(postDist)
      postDist2 <- postDist[!is.na(postDist)]
      postDist2 <- postDist2[order(postDist2)]
      head(postDist2)
      tail(postDist2)
      hist(postDist2, breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")

      L <- length(postDist2)
      L
      up <- L * .95
      up
      stri <- sprintf("Maximum log posteriori %.1f.\n", up)

      postDist2.sig <- postDist2[1:up]

      limSup <- max(postDist2.sig)
      limSup

      n <- which(postDist2 < limSup)

      genes <- unlist(strsplit(maxiGenes[n], ";"))
      genes <- genes[!is.na(genes)]
      length(genes)
      genes <- unique(genes)
      length(genes)
      genes <- genes[order(genes)]

      stri <- sprintf("%sSelected genes: %s.\n", stri, paste(genes, collapse = ", "))


      genes.found.hc.degs <- genes[which(genes %in% degs.symbols)]
      stri <- sprintf("%sSelected genes in HC DEGs: %s.\n", stri, paste(genes.found.hc.degs, collapse = ", "))

      genes.found.limbo <- genes[which(genes %in% limbo.symbols)]
      stri <- sprintf("%sSelected genes in Limbo: %s.\n", stri, paste(genes.found.limbo, collapse = ", "))

      no.degs.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% no.degs, "symbol"])
      genes.found.no.degs <- genes[which(genes %in% no.degs.symbols)]
      stri <- sprintf("%sSelected genes in 'housekeeping like' genes: %s.\n", stri, paste(genes.found.no.degs, collapse = ", "))

      prev.found <- strsplit(as.character(dfPath$genes.found), ";")[[1]]
      prev.found <- prev.found[order(prev.found)]
      stri <- sprintf("%sPrevious found gene in pathway '%s': %s.\n", stri, pathway, paste(prev.found, collapse = ", "))

      found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo)
      found.degs.and.limbo <- unique(found.degs.and.limbo)
      found.degs.and.limbo <- found.degs.and.limbo[order(found.degs.and.limbo)]
      stri <- sprintf("%sHC DEGs + Limbo genes: %s.\n", stri, paste(found.degs.and.limbo, collapse = ", "))

      common.symbols <- prev.found[prev.found %in% found.degs.and.limbo]

      if (length(common.symbols) > 0) {
        stri <- sprintf("%s\nCommon DEGs: %s.\n", stri, paste(common.symbols, collapse = ", "))
      } else {
        stri <- sprintf("%s\nThere are no common DEGs.\n", stri)
      }

      only.prev.found <- prev.found[!prev.found %in% found.degs.and.limbo]
      if (length(common.symbols) > 0) {
        stri <- sprintf("%sDEGs not present in new algorithm: %s.\n", stri, paste(only.prev.found, collapse = ", "))
      } else {
        stri <- sprintf("%sAll previous DEGs were found using the new algorithm.\n", stri)
      }

      only.new.found <- found.degs.and.limbo[!found.degs.and.limbo %in% prev.found]
      if (length(common.symbols) > 0) {
        stri <- sprintf("%sNew DEGs found using the new algorithm: %s.\n", stri, paste(only.new.found, collapse = ", "))
      } else {
        stri <- sprintf("%sThere are NO NEW DEGs found using the new algorithm.\n", stri)
      }

      cat("\n------ new found ----------\n")
      print(length(only.new.found))
      only.new.found <- only.new.found[ !only.new.found %in% dfLFC$symbol ]
      print(length(only.new.found))
      cat("-------------------\n\n")

      sel.genes <- c(sel.genes, only.new.found)
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      #------------- end log ---------------
      pathway2 <- gsub("/", "_", pathway)
      pathway2 <- gsub(" ", "_", pathway2)
      pathway2 <- gsub("  ", " ", pathway2)
      pathway2 <- gsub(" _ ", "_", pathway2)
      pathway2 <- gsub("__", "_", pathway2)
      fileBayes <- sprintf("%s%s/hist_ll_%s_%s_%s_%s_%s_db'%s'_%s.png", main.result.bayes, db, tech, case, comparing, study, study_aux, db, pathway2)
      fileBayes
      file.exists(fileBayes)

      tryCatch({
        png(fileBayes, width = 10 * dpi, height = 8 * dpi, res = dpi)
        hist(postDist2, breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")
        dev.off()
      }, error = function(e) {
        stri <- sprintf("\n\nThere was an error: saving %s\n\n\n", fileBayes)
        text <- write.log(text = text, filename = fileLog, f1 = stri)
      })


      llMu <- round(mean(postDist2), 2)
      minLL <- round(min(postDist2), 2)
      maxLL <- round(max(postDist2), 2)
      stri <- sprintf("\n\nmean(Post): %.2f\n", llMu)
      stri <- sprintf("%smax(Post) = min(-log10(Post)): %.2f\n", stri, minLL)
      stri <- sprintf("%smin(Post) = max(-log10(Post)): %.2f\n", stri, maxLL)
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      listPathways[[comparing]][[pathway]] <- list()
      listPathways[[comparing]][[pathway]][["mu"]] <- llMu
      listPathways[[comparing]][[pathway]][["minLL"]] <- minLL
      listPathways[[comparing]][[pathway]][["maxLL"]] <- maxLL
      listPathways[[comparing]][[pathway]][["n.genes.found.in.pathway"]] <- length(prev.found)
      listPathways[[comparing]][[pathway]][["genes.found.in.pathway"]] <- prev.found
      listPathways[[comparing]][[pathway]][["HC-DEGs"]] <- degs.symbols
      listPathways[[comparing]][[pathway]][["Limbo"]] <- limbo.symbols
      listPathways[[comparing]][[pathway]][["NO-DEGs"]] <- no.degs.symbols
      listPathways[[comparing]][[pathway]][["n.genes.found.new.algorithm"]] <- length(genes)
      listPathways[[comparing]][[pathway]][["genes.found.new.algorithm"]] <- genes
      listPathways[[comparing]][[pathway]][["n.genes.found.degs.limbo"]] <- length(found.degs.and.limbo)
      listPathways[[comparing]][[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
      listPathways[[comparing]][[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
      listPathways[[comparing]][[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
      listPathways[[comparing]][[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
      listPathways[[comparing]][[pathway]][["common"]] <- common.symbols
      listPathways[[comparing]][[pathway]][["only.previous.degs"]] <- only.prev.found
      listPathways[[comparing]][[pathway]][["only.new.degs"]] <- only.new.found
      listPathways[[comparing]][[pathway]][["ll.posteriori"]] <- postDist
    }

    
```

# KEGG
```{r}
if (!file.exists(fileLL) | force) {
  listPathways <- list()
  #---- loop first comparings ---------------------------------
  comparing <- comparings[1]

  for (comparing in comparings[1:2]) {
    cat(sprintf("\n\n>> %s \n", comparing))
    listPathways[[comparing]] <- list()

    if (comparing == "t6xt0") {
      mean.cols <- c("mean1", "mean2")
      pis.cols <- c("pisCtrl", "pisCase1")
      colControl <- pis.cols[1]
      colCase <- pis.cols[2]

      sample.cols <- good.cols[ c(grep("h0", good.cols), grep("h6", good.cols))]
      control_cols <- good.cols[grep("h0", good.cols)]
    } else if (comparing == "t12xt0") {
      mean.cols <- c("mean1", "mean3")
      pis.cols <- c("pisCtrl", "pisCase2")
      colControl <- pis.cols[1]
      colCase <- pis.cols[2]

      sample.cols <- good.cols[ c(grep("h0", good.cols), grep("h12", good.cols))]
      control_cols <- good.cols[grep("h0", good.cols)]
    }


    #--------- open files -------------
    dfLFC5 <- open_dfLFC5new(comparing)
    dfLFC5 <- dfLFC5[dfLFC5$gene.id != "", ]
    dfLFC5 <- dfLFC5[order(dfLFC5$fdr), ]
    nrow(dfLFC5)

    all.symbols <- as.character(dfLFC5$symbol)
    all.symbols <- all.symbols[!is.na(all.symbols)]
    all.symbols <- all.symbols[ all.symbols != ""]
    all.symbols <- unique(all.symbols)
    all.symbols <- all.symbols[order(all.symbols)]
    length(all.symbols)
    #-- 13667 ... 6h

    geneids <- as.character(dfLFC5$gene.id)

    stri <- sprintf("Opening dfLFC5 for %s\n", comparing)
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    stri <- sprintf("There are %d valid (ortholog) symbols from %d gene IDs.\n", length(all.symbols), length(geneids))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    # There are 13667 valid (ortholog) symbols from 14867 gene IDs.
    stri <- sprintf("that represents %.2f%% ortholog conversion.\n", length(all.symbols) * 100 / length(geneids))
    text <- write.log(text = text, filename = fileLog, f1 = stri)
    # that represents 91.93% ortholog conversion.

    filePis <- sprintf("%sdistribution_pis_geneID_ensembl_%s_%s_%s_%s_%s_LRT.tsv", main.result.bayes, tech, case, comparing, study, study_aux)
    filePis

    if (!file.exists(filePis)) {
      unique.groups
      totMed1 <- sum(dfLFC5$median1, na.rm = T)
      totMed2 <- sum(dfLFC5$median2, na.rm = T)
      totMed3 <- sum(dfLFC5$median3, na.rm = T)

      tots <- c(totMed1, totMed2, totMed3)
      names(tots) <- unique.groups
      tots

      pisCtrl <- as.numeric(dfLFC5$median1) / totMed1
      pisCase1 <- as.numeric(dfLFC5$median2) / totMed2
      pisCase2 <- as.numeric(dfLFC5$median3) / totMed3

      dfPis <- cbind(dfLFC5, pisCtrl, pisCase1, pisCase2)
      ret <- my.write.csv(dfPis, filePis)
    } else {
      dfPis <- my.read.csv(filePis)
    }



    #--- special product to evaluate high anova with high total mean -----
    n <- grep("median", colnames(dfPis))
    if (length(n) == 0) {
      stop("Didn't find median columns!!! Fatal error.")
      return(NULL)
    }

    #---- parameter:  median * anova
    meds <- apply(dfPis[, n], 1, sum) / length(n)
    anova.med.all <- -log10(as.numeric(dfPis$pvals.anova)) * meds


    dfPis <- cbind(dfPis, anova.med.all)
    colnames(dfPis)

    dfPis <- dfPis[order(-dfPis$anova.med.all), ]
    ncol(dfPis)
    dfPis[1:30, c(1:5, 28:32)]

    #--- parameter: bayes.min.median = 2
    if (comparing == comparings[1]) {
      dfLowVar <- dfPis[dfPis$pvals.anova >= bayes.pvals.anova &
        dfPis$median1 > bayes.min.median & dfPis$median2 > bayes.min.median, ]
    } else {
      dfLowVar <- dfPis[dfPis$pvals.anova >= bayes.pvals.anova &
        dfPis$median1 > bayes.min.median & dfPis$median3 > bayes.min.median, ]
    }

    nrow(dfPis)
    nrow(dfLowVar)
    colnames(dfLowVar)

    #-- high confidence ---------------------------------------------------
    # fdr <- .05; cutoffLods = -log10(fdr)
    # foldChange = 2; cutoffLogFC = log2(foldChange)

    dfLFC <- open_dfLFC(comparing)
    dfLFC[dfLFC$symbol == "CXCL1", ]
    nDEG.species <- nrow(dfLFC)

    dfLFC <- dfLFC[dfLFC$gene.id != "", ]
    nDEG.human <- nrow(dfLFC)
    stri <- sprintf("There are %d ortholog human DEGs in %d DEGs", nDEG.human, nDEG.species)
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    normal.degs <- as.character(dfLFC$gene.id)
    normal.symbol.degs <- unique_symbols_geneids(dfLFC$symbol)

    stri <- sprintf("There are %d valid ortholog DEGs - with symbol - DEGs from %d gene IDs.\n", length(normal.symbol.degs), length(normal.degs))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    # There are 13667 valid (ortholog) symbols from 14867 gene IDs.
    stri <- sprintf("that represents %.2f%% ortholog conversion.\n", length(normal.symbol.degs) * 100 / length(normal.degs))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    #-- high confidence ---------------------------------------------------
    #-- parameter: fdr.hc = 0.01 (usually) ;  cutoff.lfc.hc = 1
    dfLFC.HC <- dfPis[ (dfPis$fdr < fdr.hc) & (abs(dfPis$logFC) >= cutoff.lfc.hc), ]

    nHC.genes <- nrow(dfLFC.HC)
    nHC.genes

    #---- getting the minimum possible DEGs -------------------------------
    hc.degs <- as.character(dfLFC.HC$gene.id)
    length(hc.degs)

    want.to.cut <- F

    if (want.to.cut) {
      if (length(hc.degs) > 150) {
        hc.degs <- hc.degs[1:150]
        stri <- paste(stri, "Cutting to 150 DEGs.\n\n", sep = "")
        nHC.genes <- 150
      }
    }
    stri <- sprintf("There are %s high confident DEGs, with fdr < %.2f and cutoffLogFC > %.2f.\n", length(hc.degs), fdr.hc, cutoff.lfc.hc)
    text <- write.log(text = text, filename = fileLog, f1 = stri)


    dfPis <- dfPis[order(dfPis$fdr), ]
    nrow(dfPis)
    nrow(dfLowVar)
    nrow(dfLFC.HC)

    #-- high confidence -------------------------------------------------
    #-- parameter: fdr.limbo = 0.07,  cutoffLFC.limbo = log2(1.4) =  0.4854268
    dfLimbo <- dfPis[ (dfPis$fdr >= fdr.hc) & (dfPis$fdr < fdr.limbo) & (abs(dfPis$logFC) > cutoffLFC.limbo), ]
    nLimbo <- nrow(dfLimbo)

    #---- calc Limbo -----------------------------------------------------
    #---- getting the minimum possible DEGs ---------------.
    limbo <- as.character(dfLimbo$gene.id)

    sprintf("There are %s high confident DEGs, with fdr < %.2f and cutoffLogFC > %.2f.\n", length(hc.degs), fdr.hc, cutoff.lfc.hc)
    stri <- sprintf("There are %d low Var genes.\n", nrow(dfLowVar))
    stri <- sprintf("%sThere are %d limbo genes, with fdr >= %.2f and fdr < %.2f.\n", stri, length(limbo), fdr.hc, fdr.limbo)
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    #---- low Var cannot be DEGs ---------------------
    n <- which(dfLowVar$gene.id %in% hc.degs)
    length(n)

    # dfLowVar[n, 1:5]
    # dfLFC.HC[dfLFC.HC$symbol %in% c("LILRB3"), ]
    # dfLFC.HC[dfLFC.HC$gene.id %in%  dfLowVar[n, "gene.id"], ]
    # dfLowVar[n, ]

    #--- low Var should no be in Limbo ---------------------
    n <- which(dfLowVar$gene.id %in% limbo)
    if (length(n) > 0) {
      stri <- "\nUauuu: why hc.degs with low anova???\n\n"
      text <- write.log(text = text, filename = fileLog, f1 = stri)
      dfLowVar <- dfLowVar[-n, ]
    }
    midLayer <- as.character(dfLowVar$gene.id)

    par(mfrow = c(2, 2))
    hist(log10(dfLowVar$median1), main = "median1 - 'limbo' frequencies", breaks = 50)
    hist(log10(dfLowVar$median2), main = "median2 - 'limbo' frequencies", breaks = 50)
    hist(log10(dfLowVar$median3), main = "median3 - 'limbo' frequencies", breaks = 50)
    hist(-log10(dfLowVar$pvals.anova), main = "-log10(anova) - middle layer ANOVA distribution (limbo)", breaks = 50)
    par(mfrow = c(1, 1))

    cutoff.median <- median(dfLowVar$median2)
    nrow(dfLowVar[dfLowVar$median2 > cutoff.median, ])

    if (comparing == comparing[1]) {
      dfMiddleHighExp <- dfLowVar[dfLowVar$median2 > cutoff.median, ]
    } else {
      dfMiddleHighExp <- dfLowVar[dfLowVar$median3 > cutoff.median, ]
    }
    nrow(dfMiddleHighExp)

    no.degs <- as.character(dfMiddleHighExp$gene.id)

    stri <- sprintf("There are %d NO DEGs~Middle layer High Expression/High Variance.\n", length(no.degs))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    length(hc.degs)
    length(limbo)
    length(no.degs)

    stri <- sprintf(
      "All genes = %d, DEGs+Limbo+NoDegs = %d (%d, %d, %d).\n",
      nrow(dfPis), length(hc.degs) + length(limbo) + length(no.degs), length(hc.degs), length(limbo), length(no.degs)
    )
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    #-- we will put allways the degs, and sorte 50% of limbo degs and HK
    #-- N is the variable parte - were we will sample

    #----------- Enriched Pathways db = KEGG -------------------------
    df.pathways <- open_Enrichment.string.table(comparing, db, 0)
    colnames(df.pathways)
    df.pathways[1:7, 1:9]

    #----------- Loop -------------------------
    stri <- paste("\n\nBegining loop of pathways:\n", sep = "")
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    # i = which(df.pathways$pathway == "PI3K-Akt signaling pathway")
    # i = which(df.pathways$pathway == "Dopaminergic synapse")

    degs.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% hc.degs, "symbol"])
    limbo.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% limbo, "symbol"])

    #------------------  loop second pathways  ---------------------------------
    cat("loop pathways \n")
    sel.genes <- c()
    i <- 1
    for (i in 1:nrow(df.pathways)) { #  1:3

      dfPath <- df.pathways[i, ]
      pathway <- as.character(dfPath$pathway)
      stri_ini <- sprintf(">> %s - for %s:\n", pathway, comparing)
      stri_ini

      genes.in.pathway <- as.character(dfPath$genes.path)
      genes.in.pathway <- strsplit(genes.in.pathway, ";")[[1]]
      genes.in.pathway <- genes.in.pathway[order(genes.in.pathway)]
      genes.in.pathway
      if (length(genes.in.pathway) == 0) next

      # n.genes.found.in.pathway = as.character(dfPath$n.genes.found)
      genes.found.in.pathway <- as.character(dfPath$genes.found)
      nGenesFoundInPathway <- length(genes.found.in.pathway)

      all.sel.gene_ids <- c(hc.degs, limbo, no.degs)

      all.sel.gene_ids <- all.sel.gene_ids[all.sel.gene_ids != ""]
      all.sel.gene_ids <- unique(all.sel.gene_ids)
      length(all.sel.gene_ids)
      c(length(hc.degs), length(no.degs), length(limbo), length(all.sel.gene_ids))

      genes.found.in.pathway <- as.character(dfPath$genes.found)
      nGenesFoundInPathway <- length(genes.found.in.pathway)

      # length(midLayer)
      postDist <- c()
      prioriDist <- c()
      verossDist <- c()
      maxiGenes <- c()
      c(length(hc.degs), length(no.degs), length(limbo), length(all.sel.gene_ids))

      all.sel.gene_ids <- c(hc.degs, no.degs, limbo)
      all.sel.gene_ids

      nAllGenes <- length(all.symbols)

      nGoodGenes <- length(all.sel.gene_ids)
      nBadGenes <- nAllGenes - nGoodGenes

      if (i == 1) {
        stri <- sprintf("<<--->>\n\nThere are %d High Confident DEGs.\n", nHC.genes)
        stri <- sprintf("%sThere are %d medium confident DEGs (limbo).\n", stri, length(limbo))
        stri <- sprintf("%sThere are %d 'housekeeping' genes.\n", stri, length(midLayer))
        stri <- sprintf("%sThere are %d 'good genes' = hc.degs + limbo + HK.\n", stri, nGoodGenes)
        stri <- sprintf("%sWe will draw a maximum of %d genes.\n", stri, nSample)
        stri <- sprintf("%sTherefore, there are %d 'bad genes'.\n", stri, nBadGenes)
        stri <- sprintf("%sWe will perform %d loop as bootstrap'.\n", stri, nLoop)
        stri <- sprintf("%sOnly enrichment with %d or more genes will be selected'.\n", stri, min.EnrGenes)
        stri <- sprintf("%s\n\n%s", stri, stri_ini)
      } else {
        stri <- sprintf("<<--->>\n\n%s", stri_ini)
      }
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      #------------------------------------------------------------------------------------.
      #--- bootstrap -- bayesian ----------------------------------------------------------
      #------------------------------------------------------------------------------------.

      for (r in 1:nLoop) {
        #-- sample hk --------------------------------
        sample.gene_ids <- sample(all.sel.gene_ids, nSample, replace = FALSE)
        sample.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% sample.gene_ids, "symbol"])
        sample.symbols <- unique(sample.symbols)
        sample.symbols <- sample.symbols[sample.symbols != ""]
        sample.symbols <- sample.symbols[order(sample.symbols)]
        length(sample.symbols)

        nDrawn <- length(sample.symbols)

        # length(sample.symbols %in% all.symbols)

        genes.found <- sample.symbols[sample.symbols %in% genes.in.pathway]
        nFound <- length(genes.found)

        if (nFound < min.EnrGenes) {
          verossDist <- c(verossDist, NA)
          prioriDist <- c(prioriDist, NA)
          postDist <- c(postDist, NA)
          maxiGenes <- c(maxiGenes, NA)
          next
        }

        # ?phyper
        pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
        verossDist <- c(verossDist, pORA)

        #----------- log10(p(Gset)) -------------
        thisCol <- colCase
        log10.pGs <- calc.pGs.log10(dfPis, sample.gene_ids, thisCol = colCase)
        # print(log10.pGs)

        # post = -log10(pORA) - log10.pGs - (-log10(pORA + not.pORA))
        post <- -log10(pORA) - log10.pGs

        prioriDist <- c(prioriDist, log10.pGs)
        postDist <- c(postDist, post)
        maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
      }
      # postDist = postDist/sum(postDist)

      #------------- log -------------------

      # postDist = postDist/sum(postDist)
      postDist2 <- postDist[!is.na(postDist)]
      postDist2 <- postDist2[order(postDist2)]
      head(postDist2)
      tail(postDist2)
      hist(postDist2, breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")

      L <- length(postDist2)
      L
      up <- L * .95
      up
      stri <- sprintf("Maximum log posteriori %.1f.\n", up)

      postDist2.sig <- postDist2[1:up]

      limSup <- max(postDist2.sig)
      limSup

      n <- which(postDist2 < limSup)

      genes <- unlist(strsplit(maxiGenes[n], ";"))
      genes <- genes[!is.na(genes)]
      length(genes)
      genes <- unique(genes)
      length(genes)
      genes <- genes[order(genes)]

      stri <- sprintf("%sSelected genes: %s.\n", stri, paste(genes, collapse = ", "))


      genes.found.hc.degs <- genes[which(genes %in% degs.symbols)]
      stri <- sprintf("%sSelected genes in HC DEGs: %s.\n", stri, paste(genes.found.hc.degs, collapse = ", "))

      genes.found.limbo <- genes[which(genes %in% limbo.symbols)]
      stri <- sprintf("%sSelected genes in Limbo: %s.\n", stri, paste(genes.found.limbo, collapse = ", "))

      no.degs.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% no.degs, "symbol"])
      genes.found.no.degs <- genes[which(genes %in% no.degs.symbols)]
      stri <- sprintf("%sSelected genes in 'housekeeping like' genes: %s.\n", stri, paste(genes.found.no.degs, collapse = ", "))

      prev.found <- strsplit(as.character(dfPath$genes.found), ";")[[1]]
      prev.found <- prev.found[order(prev.found)]
      stri <- sprintf("%sPrevious found gene in pathway '%s': %s.\n", stri, pathway, paste(prev.found, collapse = ", "))

      found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo)
      found.degs.and.limbo <- unique(found.degs.and.limbo)
      found.degs.and.limbo <- found.degs.and.limbo[order(found.degs.and.limbo)]
      stri <- sprintf("%sHC DEGs + Limbo genes: %s.\n", stri, paste(found.degs.and.limbo, collapse = ", "))

      common.symbols <- prev.found[prev.found %in% found.degs.and.limbo]

      if (length(common.symbols) > 0) {
        stri <- sprintf("%s\nCommon DEGs: %s.\n", stri, paste(common.symbols, collapse = ", "))
      } else {
        stri <- sprintf("%s\nThere are no common DEGs.\n", stri)
      }

      only.prev.found <- prev.found[!prev.found %in% found.degs.and.limbo]
      if (length(common.symbols) > 0) {
        stri <- sprintf("%sDEGs not present in new algorithm: %s.\n", stri, paste(only.prev.found, collapse = ", "))
      } else {
        stri <- sprintf("%sAll previous DEGs were found using the new algorithm.\n", stri)
      }

      only.new.found <- found.degs.and.limbo[!found.degs.and.limbo %in% prev.found]
      if (length(common.symbols) > 0) {
        stri <- sprintf("%sNew DEGs found using the new algorithm: %s.\n", stri, paste(only.new.found, collapse = ", "))
      } else {
        stri <- sprintf("%sThere are NO NEW DEGs found using the new algorithm.\n", stri)
      }

      cat("\n------ new found ----------\n")
      print(length(only.new.found))
      only.new.found <- only.new.found[ !only.new.found %in% dfLFC$symbol ]
      print(length(only.new.found))
      cat("-------------------\n\n")

      sel.genes <- c(sel.genes, only.new.found)
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      #------------- end log ---------------
      pathway2 <- gsub("/", "_", pathway)
      pathway2 <- gsub(" ", "_", pathway2)
      pathway2 <- gsub("  ", " ", pathway2)
      pathway2 <- gsub(" _ ", "_", pathway2)
      pathway2 <- gsub("__", "_", pathway2)
      fileBayes <- sprintf("%s%s/hist_ll_%s_%s_%s_%s_%s_db'%s'_%s.png", main.result.bayes, db, tech, case, comparing, study, study_aux, db, pathway2)
      fileBayes
      file.exists(fileBayes)

      tryCatch({
        png(fileBayes, width = 10 * dpi, height = 8 * dpi, res = dpi)
        hist(postDist2, breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")
        dev.off()
      }, error = function(e) {
        stri <- sprintf("\n\nThere was an error: saving %s\n\n\n", fileBayes)
        text <- write.log(text = text, filename = fileLog, f1 = stri)
      })


      llMu <- round(mean(postDist2), 2)
      minLL <- round(min(postDist2), 2)
      maxLL <- round(max(postDist2), 2)
      stri <- sprintf("\n\nmean(Post): %.2f\n", llMu)
      stri <- sprintf("%smax(Post) = min(-log10(Post)): %.2f\n", stri, minLL)
      stri <- sprintf("%smin(Post) = max(-log10(Post)): %.2f\n", stri, maxLL)
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      listPathways[[comparing]][[pathway]] <- list()
      listPathways[[comparing]][[pathway]][["mu"]] <- llMu
      listPathways[[comparing]][[pathway]][["minLL"]] <- minLL
      listPathways[[comparing]][[pathway]][["maxLL"]] <- maxLL
      listPathways[[comparing]][[pathway]][["n.genes.found.in.pathway"]] <- length(prev.found)
      listPathways[[comparing]][[pathway]][["genes.found.in.pathway"]] <- prev.found
      listPathways[[comparing]][[pathway]][["HC-DEGs"]] <- degs.symbols
      listPathways[[comparing]][[pathway]][["Limbo"]] <- limbo.symbols
      listPathways[[comparing]][[pathway]][["NO-DEGs"]] <- no.degs.symbols
      listPathways[[comparing]][[pathway]][["n.genes.found.new.algorithm"]] <- length(genes)
      listPathways[[comparing]][[pathway]][["genes.found.new.algorithm"]] <- genes
      listPathways[[comparing]][[pathway]][["n.genes.found.degs.limbo"]] <- length(found.degs.and.limbo)
      listPathways[[comparing]][[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
      listPathways[[comparing]][[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
      listPathways[[comparing]][[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
      listPathways[[comparing]][[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
      listPathways[[comparing]][[pathway]][["common"]] <- common.symbols
      listPathways[[comparing]][[pathway]][["only.previous.degs"]] <- only.prev.found
      listPathways[[comparing]][[pathway]][["only.new.degs"]] <- only.new.found
      listPathways[[comparing]][[pathway]][["ll.posteriori"]] <- postDist
    }

    sel.genes <- unique_symbols_geneids(sel.genes)
    sel.genes
    fileGenes <- sprintf("%snew_degs_bayes_symbol_db'%s'_%s_%s_%s_%s.txt", main.result.bayes, db, study, study_aux, case, comparing)
    fileGenes
    ret <- my.write.txt(sel.genes, fileGenes)

    fileEnsSamp <- sprintf("%sensembl_samples_bayes_db'%s'_%s_%s_%s_%s.txt", main.result.bayes, db, study, study_aux, case, comparing)
    fileEnsSamp
    ret <- my.write.txt(paste(all.sel.gene_ids, collapse = ", "), fileEnsSamp)

    dfLFC5 <- dfLFC5[dfLFC5$gene.id %in% all.sel.gene_ids, ]
    dfAux <- dfLFC5[dfLFC5$symbol %in% sel.genes, ]
    dfLFC.bayes <- rbind(dfLFC, dfAux)
    length(sel.genes)
    nrow(dfLFC)
    length(sel.genes) + nrow(dfLFC)
    #-- may have parologs ------.
    nrow(dfLFC.bayes)

    #-------- dfLFC.bayes does not allow repeated symbols ---------
    dfLFC.bayes <- dfLFC.bayes[order(dfLFC.bayes$symbol, dfLFC.bayes$fdr), ]
    # dfLFC.bayes[, c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    dfLFC.bayes[dfLFC.bayes$symbol == "CXCL1", ]
    nrow(dfLFC.bayes)
    # dfAux2 = dfLFC.bayes
    keep <- c()
    previous <- ""
    for (m in 1:nrow(dfLFC.bayes)) {
      symbol <- dfLFC.bayes[m, "symbol"]
      if (symbol != previous) {
        previous <- symbol
        keep <- c(keep, m)
      }
    }
    dfLFC.bayes <- dfLFC.bayes[keep, ]
    # dfLFC.bayes[, c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    # dfLFC.bayes[dfLFC.bayes$symbol == "CXCL1", ]
    nrow(dfLFC.bayes)
    #--------------------------------------------------------.

    dfLFC[dfLFC$symbol == "CXCL1", c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    dfLFC.bayes[dfLFC.bayes$symbol == "CXCL1", c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    fileLFCb <- sprintf("%sLFC_bayes_ensembl_%s_%s_%s_db'%s'_%s_%s.tsv", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileLFCb
    ret <- my.write.csv(dfLFC.bayes, fileLFCb)

    fileLFCb3c <- sprintf("%sLFC_bayes_ensembl_%s_%s_%s_db'%s'_%s_%s_3cols.tsv", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileLFCb3c
    dfLFC.bayes <- dfLFC.bayes[, c("gene.id", "symbol", "logFC")]
    dfLFC.bayes <- unique(dfLFC.bayes)
    ret <- my.write.csv(dfLFC.bayes, fileLFCb3c)

    all.new.degs <- c(all.new.degs, sel.genes)

    symbols <- unique_symbols_geneids(dfLFC.bayes$symbol)
    stri.symbols <- paste(symbols, collapse = "\n")
    fileNewDegs <- sprintf("%sdegs_bayes_symbol_%s_%s_%s_db'%s'%s_%s.txt", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileNewDegs
    ret <- my.write.txt(stri.symbols, fileNewDegs)

    geneids <- unique_symbols_geneids(dfLFC.bayes$gene.id)
    stri.geneids <- paste(geneids, collapse = "\n")
    fileNewDegs <- sprintf("%sdegs_bayes_ensembl_%s_%s_%s_db'%s'_%s_%s.txt", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileNewDegs
    ret <- my.write.txt(stri.geneids, fileNewDegs)
  }

  #----------- End Loop -------------------------
  ret <- my.write.list(listPathways, fileLL)
} else {
  stri <- sprintf("\nFile already exists: '%s'\n", fileLL)
  text <- write.log(text = text, filename = fileLog, f1 = stri)
}

stri <- "--- end ---------------\n"
text <- write.log(text = text, filename = fileLog, f1 = stri)

comparing <- comparings[1]
names(listPathways)
names(listPathways[[comparings[1]]])
names(listPathways[[comparings[2]]])

names(listPathways[[comparings[1]]][["Hypertrophic cardiomyopathy (HCM)"]])
listPathways[[comparings[1]]][["Hypertrophic cardiomyopathy (HCM)"]][["mu"]]
listPathways[[comparings[2]]][["Measles"]][["mu"]]

#----------- loop of listPathways -----------------------
listPathways <- my.read.list(fileLL)
names(listPathways)

comparing <- comparings[1]

df <- data.frame()
comparing <- comparings[1]

for (comparing in comparings[1:2]) {
  cat(sprintf(">> comparing %s\n", comparing))
  nms <- names(listPathways[[comparing]])
  pathway <- nms[1]

  for (pathway in nms) {
    mu <- listPathways[[comparing]][[pathway]][["mu"]]
    minLL <- listPathways[[comparing]][[pathway]][["minLL"]]
    maxLL <- listPathways[[comparing]][[pathway]][["maxLL"]]
    genes.found.in.pathway <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.pathway"]], collapse = ", ")
    genes.found.in.hc.degs <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.hc-degs" ]], collapse = ", ")
    genes.found.in.limbo <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.limbo"]], collapse = ", ")
    genes.found.in.no.degs <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.no-degs"]], collapse = ", ")
    common <- paste(listPathways[[comparing]][[pathway]][["common"]], collapse = ", ")
    only.previous.degs <- paste(listPathways[[comparing]][[pathway]][["only.previous.degs"]], collapse = ", ")

    new.degs <- listPathways[[comparing]][[pathway]][["only.new.degs"]]
    only.new.degs <- paste(new.degs, collapse = ", ")

    genes2 <- new.degs
    genes2 <- genes2[genes2 != ""]
    # print(only.new.degs)
    sel.genes <- c(sel.genes, genes2)

    dfAux <- data.frame(
      comparing, pathway, mu, minLL, maxLL, genes.found.in.pathway,
      genes.found.in.hc.degs, genes.found.in.limbo, genes.found.in.no.degs,
      common, only.previous.degs, only.new.degs
    )

    df <- rbind(df, dfAux)
  }
}

file.df.ll <- sprintf("%sLL_db'%s'_%s_%s.tsv", main.result.bayes, db, study, study_aux)
file.df.ll
colnames(df)
dim(df)
ret <- my.write.csv(df, file.df.ll)

head(df[, 1:5])
tail(df[, 1:5])

all.new.degs <- unique_symbols_geneids(all.new.degs)
all.new.degs

stri.all.new.degs <- paste(all.new.degs, collapse = "\n")
fileNewDegs <- sprintf("%sall_new_degs_%s_%s_db'%s'_%s_%s.txt", main.result.bayes, tech, case, db, study, study_aux)
fileNewDegs
ret <- my.write.txt(all.new.degs, fileNewDegs)

#--------------------- histogram posterioris -----------------------------
# dev.off()
par(mfrow = c(2, 2))
hist(df[df$comparing == comparings[1], "mu"], main = "mean posteriori distribution", xlab = "mean(posteriori)")
hist(df[df$comparing == comparings[2], "mu"], main = "mean posteriori distribution", xlab = "mean(posteriori)")
hist(df[df$comparing == comparings[1], "minLL"], main = "minimum posteriori distribution", xlab = "minimum(posteriori)")
hist(df[df$comparing == comparings[2], "minLL"], main = "minimum posteriori distribution", xlab = "minimum(posteriori)")
par(mfrow = c(1, 1))


df <- df[order(df$minLL), ]
df[1:8, 1:4]

nrow(df)

for (comparing in comparings[1:2]) {
  cat(sprintf("\n\n>> %s \n", comparing))

  dfAux <- df[df$comparing == comparing, ]
  for (i in 1:nrow(dfAux)) {
    cat(sprintf("\n ## pathway: '%s'\n", dfAux[i, "pathway"]))
    cat(sprintf("    previous found:    '%s'\n", dfAux[i, "genes.found.in.pathway"]))
    cat(sprintf("    common:            '%s'\n", dfAux[i, "common"]))
    cat(sprintf("    only.previous.degs:'%s'\n", dfAux[i, "only.previous.degs"]))
    cat(sprintf("    only.new.degs:     '%s'\n\n\n", dfAux[i, "only.new.degs"]))
  }
}

```

# Reactome
```{r}
#--------------------------------------------------------------.
#------------------ bayes_gene_improvement_lib.R ---------------
#--------------------------------------------------------------.

tab <- "\t"

# thisCol = colCase
calc.pGs.log10 <- function(dfPis, gene_ids, thisCol) {
  pis <- as.numeric(paste(dfPis[dfPis$gene.id %in% gene_ids, thisCol]))

  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8

  return(sum(log10(pis)))
}


DE_results <- DE_results %>% mutate(pGs_log10 = PValue %>% log10)

enrichment_result <- enrichment_result %>% mutate(pGs_log10 = pvalue %>% log10)

#--- bad ----
#----------- perturbed ------------------------
# thisCol = colCase
#----------- control ------------------------
# thisCol = colControl
calc.pGs <- function(dfPis, degs, sample.symb, not.sample.symb, comb, thisCol) {
  dfx <- dfPis[dfPis$symbol %in% c(degs, sample.symb, not.sample.symb), c("symbol", thisCol)]
  # nrow(dfx)
  # head(dfx)
  tot <- sum(as.numeric(dfx[, 2]))
  dfx[, 2] <- as.numeric(dfx[, 2]) / tot
  # head(dfx,100)
  # sum(as.numeric(dfx[,2]))

  pis <- dfx[ dfx$symbol %in% sample.symb, thisCol]
  length(pis)
  p.Gs.given.pert <- 0
  for (pi in pis) {
    p.Gs.given.pert <- p.Gs.given.pert + log(pi)
  }
  p.Gs.given.pert

  pis.not <- dfx[ dfx$symbol %in% not.sample.symb, thisCol]
  for (pi in pis.not) {
    p.Gs.given.pert <- p.Gs.given.pert + log(1 - pi)
  }
  p.Gs.given.pert

  p.Gs.given.pert <- log(comb) + p.Gs.given.pert
  p.Gs.given.pert
  return(exp(p.Gs.given.pert))
}

stirling <- function(n) {
  n * log(n) - n
}
# nrow(dfPis)

calc.pGs_wrong <- function(dfPis, my.symbols, not.sample.symb, thisCol) {
  pis <- dfPis[dfPis$symbol %in% my.symbols, thisCol]
  length(pis)
  p.Gs.given.pert <- 0
  for (pi in pis) {
    p.Gs.given.pert <- p.Gs.given.pert + log10(pi)
  }
  p.Gs.given.pert

  pis.not <- dfPis[dfPis$symbol %in% not.sample.symb, thisCol]
  length(pis.not)
  for (pi in pis.not) {
    p.Gs.given.pert <- p.Gs.given.pert + log10(1 - pi)
  }
  p.Gs.given.pert
  p.Gs.given.pert <- exp(p.Gs.given.pert)


  return(p.Gs.given.pert)
}


genes.wo.upstream <- function() {
  ids <- as.character(dfEntries$id.elem)

  id <- ids[1]
  no.upstreams <- c()

  for (id in ids) {
    cat(paste(id, "\n"))

    stri <- as.character(dfEntries[dfEntries$id.elem == id, "name"])
    entrezids <- gsub("hsa:", "", strsplit(stri, " ")[[1]])

    entrezid <- entrezid[1]
    have.usptream <- F
    for (entrezid in entrezids) {
      upstreams <- dfRelations[ dfRelations$entry2 == id, c("symb1", "functional", "symbs1", "posTranslational", "posTransl.symb")]
      if (nrow(upstreams) > 0) {
        have.usptream <- T
        break
      }
    }

    if (!have.usptream) {
      no.upstreams <- c(no.upstreams, id)
    }
  }

  return(no.upstreams)
}


symb <- "PIK3CA"
id <- NULL
find_up_stream <- function(symb = NULL, id = NULL) {
  if (!is.null(symb)) {
    dfIdSymb[dfIdSymb$symbol == symb, ]
    entrezid <- dfIdSymb[dfIdSymb$symbol == symb, "entrezID"]
    hsa <- paste("hsa:", entrezid, sep = "")
    ids <- dfEntries[grep(hsa, dfEntries$name), "id.elem"]
  } else {
    ids <- c(id)
  }

  id <- ids[1]

  for (id in ids) {
    # dfEntries[dfEntries$id.elem == id, ]
    stri <- as.character(dfEntries[dfEntries$id.elem == id, "name"])
    entrezids <- gsub("hsa:", "", strsplit(stri, " ")[[1]])

    entrezid <- entrezid[1]
    lista <- c()
    for (entrezid in entrezids) {
      symb <- as.character(dfIdSymb[dfIdSymb$entrezID == entrezid, "symbol"])
      stri <- paste("\n\n<<<", symb, "entrezid:", entrezid, "\n")
      text <- paste(text, stri)

      acc <- dfEntries[dfEntries$id.elem == id, "kegg.acc"]
      stri <- paste(tab, "### ", id, "\n", sep = "")
      text <- paste(text, stri)

      upstreams <- dfRelations[ dfRelations$entry2 == id, c("entry1", "symb1", "functional", "symbs1", "posTranslational", "posTransl.symb")]
      if (nrow(upstreams) > 0) {
        stri <- paste(tab, tab, "Upstream:\n")
        for (i in 1:nrow(upstreams)) {
          ups <- as.matrix(upstreams[i, ])
          stri <- paste(stri, tab, tab, tab, ups[1], " ", ups[2], " ", ups[3], " ", ups[6], " ", symb, ".\n", sep = "")
          lista <- c(lista, ups[1])
        }
      }
    }
  }

  return(lista)
}


# symb = "PIK3CA"; id=NULL; do.upstream=T
# lista = list(); symb=NULL; id=idInis[1]; do.upstream=T
find_up_down_path <- function(id, lista, text, do.upstream = F) {

  # if (! is.null(symb)) {
  #   text = paste(text, "Starting with symbol ", symb,"\n")
  #
  #   entrezid = as.character(dfIdSymb[dfIdSymb$symbol == symb, "entrezID"])
  #   hsa = paste("hsa:", entrezid,sep="")
  #   ids = dfEntries[grep(hsa, dfEntries$name), "id.elem"]
  # } else {
  # text = paste(text, "\n. Starting with id ", id,"\n", sep="")

  # dfEntries[dfEntries$id.elem == id, ]
  stri <- as.character(dfEntries[dfEntries$id.elem == id, "name"])
  entrezids <- gsub("hsa:", "", strsplit(stri, " ")[[1]])
  print(paste(entrezids, collapse = ";"))

  if (id %in% names(lista)) return(list(text = text, lista = lista))

  lista[[id]] <- list()
  listaID <- lista[[id]]

  entrezid <- entrezids[1]
  for (entrezid in entrezids[1]) {
    symb <- as.character(dfIdSymb[dfIdSymb$entrezID == entrezid, "symbol"])
    acc <- dfEntries[dfEntries$id.elem == id, "kegg.acc"]

    dfa <- dfLFC5[entrezid, ]
    if (nrow(dfa) == 0) {
      striAux <- ""
    } else if (entrezid == "undefined") {
      striAux <- ""
    } else if (length(grep("cpd:", entrezid)) > 0) {
      striAux <- "compound."
    } else if (is.na(dfa$mean1)) {
      striAux <- "was not expressed."
    } else {
      striAux <- sprintf("Gene=%s, EntrezID=%s, 0h=%5.1f,  6h=%5.1f,  12h=%5.1f CPM.", dfa$symbol, dfa$entrezid, dfa$mean1, dfa$mean2, dfa$mean3)
    }

    stri <- paste("\n<<< id=", id, " ", symb, ", entrezid: ", entrezid, " ... ", striAux, "\n", sep = "")
    text <- paste(text, stri)

    # dfRelations[dfRelations$entry2==id,]
    if (do.upstream) {
      listaID[["upstream"]] <- c()

      upstreams <- dfRelations[ dfRelations$entry2 == id, c("symb1", "functional", "symbs1", "posTranslational", "posTransl.symb", "entry1")]
      if (nrow(upstreams) == 0) {
        stri <- paste(tab, tab, "There is no uptreams.\n")
      } else {
        stri <- paste(tab, tab, "Upstream:\n")
        for (i in 1:nrow(upstreams)) {
          ups <- as.matrix(upstreams[i, ])

          funcStri <- gsub("activation", "beeing activated by ", ups[2])
          stri <- paste(stri, tab, tab, tab, funcStri, ups[1], ".\n", sep = "")
          listaID[["upstream"]] <- c(listaID[["upstream"]], ups[6])
        }
      }
      text <- paste(text, stri)
    }

    listaID[["downstream"]] <- c()
    downstream <- dfRelations[ dfRelations$entry1 == id, c("symb2", "functional", "symbs2", "posTranslational", "posTransl.symb", "entry2")]
    arrDown <- c()
    if (nrow(downstream) == 0) {
      stri <- paste(tab, tab, "There is no downstream.\n")
    } else {
      stri <- paste(tab, tab, "Downstream:\n")
      for (i in 1:nrow(downstream)) {
        downs <- as.matrix(downstream[i, ])

        funcStri <- gsub("activation", "activates", downs[2])
        funcStri <- gsub("inhibition", "inhibites", funcStri)

        to.symbol <- strsplit(as.character(dfEntries[dfEntries$id.elem == downs[6], "symbols"]), ";")[[1]][1]

        if (downs[5] == "") {
          striVia <- ""
        } else {
          striVia <- paste("via", downs[5])
        }
        stri <- paste(stri, tab, tab, tab, i, ") ", funcStri, " ", to.symbol, " (", downs[6], ") ", striVia, "\n", sep = "")
        #-- entry2 --> output downstreaming
        arrDown <- c(arrDown, downs[6])
        listaID[["downstream"]] <- c(listaID[["downstream"]], downs[6])
      }
    }

    text <- paste(text, stri)

    if (length(arrDown) > 0) {
      for (id in arrDown) {
        lista2 <- find_up_down_path(id = id, lista = lista, text = text, do.upstream = F)
        text <- lista2[["text"]]
        lista <- lista2[["lista"]]
        # if (!is.null(arr2)) arrDown = c(arrDown, arr2)
      }
    }
  }

  return(list(text = text, lista = lista))
}
```

# Fl√°vio libs
```{r}
calc.pGs.log10 <- function(dfPis, gene_ids, thisCol) {
  pis <- as.numeric(paste(dfPis[dfPis$gene.id %in% gene_ids, thisCol]))

  pis[pis == 0] <- 1e-8
  pis[is.na(pis)] <- 1e-8

  return(sum(log10(pis)))
}

#--- bad ----
#----------- perturbed ------------------------
# thisCol = colCase
#----------- control ------------------------
# thisCol = colControl
calc.pGs <- function(dfPis, degs, sample.symb, not.sample.symb, comb, thisCol) {
  dfx <- dfPis[dfPis$symbol %in% c(degs, sample.symb, not.sample.symb), c("symbol", thisCol)]
  # nrow(dfx)
  # head(dfx)
  tot <- sum(as.numeric(dfx[, 2]))
  dfx[, 2] <- as.numeric(dfx[, 2]) / tot
  # head(dfx,100)
  # sum(as.numeric(dfx[,2]))

  pis <- dfx[ dfx$symbol %in% sample.symb, thisCol]
  length(pis)
  p.Gs.given.pert <- 0
  for (pi in pis) {
    p.Gs.given.pert <- p.Gs.given.pert + log(pi)
  }
  p.Gs.given.pert

  pis.not <- dfx[ dfx$symbol %in% not.sample.symb, thisCol]
  for (pi in pis.not) {
    p.Gs.given.pert <- p.Gs.given.pert + log(1 - pi)
  }
  p.Gs.given.pert

  p.Gs.given.pert <- log(comb) + p.Gs.given.pert
  p.Gs.given.pert
  return(exp(p.Gs.given.pert))
}

stirling <- function(n) {
  n * log(n) - n
}
# nrow(dfPis)

calc.pGs_wrong <- function(dfPis, my.symbols, not.sample.symb, thisCol) {
  pis <- dfPis[dfPis$symbol %in% my.symbols, thisCol]
  length(pis)
  p.Gs.given.pert <- 0
  for (pi in pis) {
    p.Gs.given.pert <- p.Gs.given.pert + log10(pi)
  }
  p.Gs.given.pert

  pis.not <- dfPis[dfPis$symbol %in% not.sample.symb, thisCol]
  length(pis.not)
  for (pi in pis.not) {
    p.Gs.given.pert <- p.Gs.given.pert + log10(1 - pi)
  }
  p.Gs.given.pert
  p.Gs.given.pert <- exp(p.Gs.given.pert)


  return(p.Gs.given.pert)
}


genes.wo.upstream <- function() {
  ids <- as.character(dfEntries$id.elem)

  id <- ids[1]
  no.upstreams <- c()

  for (id in ids) {
    cat(paste(id, "\n"))

    stri <- as.character(dfEntries[dfEntries$id.elem == id, "name"])
    entrezids <- gsub("hsa:", "", strsplit(stri, " ")[[1]])

    entrezid <- entrezid[1]
    have.usptream <- F
    for (entrezid in entrezids) {
      upstreams <- dfRelations[ dfRelations$entry2 == id, c("symb1", "functional", "symbs1", "posTranslational", "posTransl.symb")]
      if (nrow(upstreams) > 0) {
        have.usptream <- T
        break
      }
    }

    if (!have.usptream) {
      no.upstreams <- c(no.upstreams, id)
    }
  }

  return(no.upstreams)
}


symb <- "PIK3CA"
id <- NULL
find_up_stream <- function(symb = NULL, id = NULL) {
  if (!is.null(symb)) {
    dfIdSymb[dfIdSymb$symbol == symb, ]
    entrezid <- dfIdSymb[dfIdSymb$symbol == symb, "entrezID"]
    hsa <- paste("hsa:", entrezid, sep = "")
    ids <- dfEntries[grep(hsa, dfEntries$name), "id.elem"]
  } else {
    ids <- c(id)
  }

  id <- ids[1]

  for (id in ids) {
    # dfEntries[dfEntries$id.elem == id, ]
    stri <- as.character(dfEntries[dfEntries$id.elem == id, "name"])
    entrezids <- gsub("hsa:", "", strsplit(stri, " ")[[1]])

    entrezid <- entrezid[1]
    lista <- c()
    for (entrezid in entrezids) {
      symb <- as.character(dfIdSymb[dfIdSymb$entrezID == entrezid, "symbol"])
      stri <- paste("\n\n<<<", symb, "entrezid:", entrezid, "\n")
      text <- paste(text, stri)

      acc <- dfEntries[dfEntries$id.elem == id, "kegg.acc"]
      stri <- paste(tab, "### ", id, "\n", sep = "")
      text <- paste(text, stri)

      upstreams <- dfRelations[ dfRelations$entry2 == id, c("entry1", "symb1", "functional", "symbs1", "posTranslational", "posTransl.symb")]
      if (nrow(upstreams) > 0) {
        stri <- paste(tab, tab, "Upstream:\n")
        for (i in 1:nrow(upstreams)) {
          ups <- as.matrix(upstreams[i, ])
          stri <- paste(stri, tab, tab, tab, ups[1], " ", ups[2], " ", ups[3], " ", ups[6], " ", symb, ".\n", sep = "")
          lista <- c(lista, ups[1])
        }
      }
    }
  }

  return(lista)
}


# symb = "PIK3CA"; id=NULL; do.upstream=T
# lista = list(); symb=NULL; id=idInis[1]; do.upstream=T
find_up_down_path <- function(id, lista, text, do.upstream = F) {

  # if (! is.null(symb)) {
  #   text = paste(text, "Starting with symbol ", symb,"\n")
  #
  #   entrezid = as.character(dfIdSymb[dfIdSymb$symbol == symb, "entrezID"])
  #   hsa = paste("hsa:", entrezid,sep="")
  #   ids = dfEntries[grep(hsa, dfEntries$name), "id.elem"]
  # } else {
  # text = paste(text, "\n. Starting with id ", id,"\n", sep="")

  # dfEntries[dfEntries$id.elem == id, ]
  stri <- as.character(dfEntries[dfEntries$id.elem == id, "name"])
  entrezids <- gsub("hsa:", "", strsplit(stri, " ")[[1]])
  print(paste(entrezids, collapse = ";"))

  if (id %in% names(lista)) return(list(text = text, lista = lista))

  lista[[id]] <- list()
  listaID <- lista[[id]]

  entrezid <- entrezids[1]
  for (entrezid in entrezids[1]) {
    symb <- as.character(dfIdSymb[dfIdSymb$entrezID == entrezid, "symbol"])
    acc <- dfEntries[dfEntries$id.elem == id, "kegg.acc"]

    dfa <- dfLFC5[entrezid, ]
    if (nrow(dfa) == 0) {
      striAux <- ""
    } else if (entrezid == "undefined") {
      striAux <- ""
    } else if (length(grep("cpd:", entrezid)) > 0) {
      striAux <- "compound."
    } else if (is.na(dfa$mean1)) {
      striAux <- "was not expressed."
    } else {
      striAux <- sprintf("Gene=%s, EntrezID=%s, 0h=%5.1f,  6h=%5.1f,  12h=%5.1f CPM.", dfa$symbol, dfa$entrezid, dfa$mean1, dfa$mean2, dfa$mean3)
    }

    stri <- paste("\n<<< id=", id, " ", symb, ", entrezid: ", entrezid, " ... ", striAux, "\n", sep = "")
    text <- paste(text, stri)

    # dfRelations[dfRelations$entry2==id,]
    if (do.upstream) {
      listaID[["upstream"]] <- c()

      upstreams <- dfRelations[ dfRelations$entry2 == id, c("symb1", "functional", "symbs1", "posTranslational", "posTransl.symb", "entry1")]
      if (nrow(upstreams) == 0) {
        stri <- paste(tab, tab, "There is no uptreams.\n")
      } else {
        stri <- paste(tab, tab, "Upstream:\n")
        for (i in 1:nrow(upstreams)) {
          ups <- as.matrix(upstreams[i, ])

          funcStri <- gsub("activation", "beeing activated by ", ups[2])
          stri <- paste(stri, tab, tab, tab, funcStri, ups[1], ".\n", sep = "")
          listaID[["upstream"]] <- c(listaID[["upstream"]], ups[6])
        }
      }
      text <- paste(text, stri)
    }

    listaID[["downstream"]] <- c()
    downstream <- dfRelations[ dfRelations$entry1 == id, c("symb2", "functional", "symbs2", "posTranslational", "posTransl.symb", "entry2")]
    arrDown <- c()
    if (nrow(downstream) == 0) {
      stri <- paste(tab, tab, "There is no downstream.\n")
    } else {
      stri <- paste(tab, tab, "Downstream:\n")
      for (i in 1:nrow(downstream)) {
        downs <- as.matrix(downstream[i, ])

        funcStri <- gsub("activation", "activates", downs[2])
        funcStri <- gsub("inhibition", "inhibites", funcStri)

        to.symbol <- strsplit(as.character(dfEntries[dfEntries$id.elem == downs[6], "symbols"]), ";")[[1]][1]

        if (downs[5] == "") {
          striVia <- ""
        } else {
          striVia <- paste("via", downs[5])
        }
        stri <- paste(stri, tab, tab, tab, i, ") ", funcStri, " ", to.symbol, " (", downs[6], ") ", striVia, "\n", sep = "")
        #-- entry2 --> output downstreaming
        arrDown <- c(arrDown, downs[6])
        listaID[["downstream"]] <- c(listaID[["downstream"]], downs[6])
      }
    }

    text <- paste(text, stri)

    if (length(arrDown) > 0) {
      for (id in arrDown) {
        lista2 <- find_up_down_path(id = id, lista = lista, text = text, do.upstream = F)
        text <- lista2[["text"]]
        lista <- lista2[["lista"]]
        # if (!is.null(arr2)) arrDown = c(arrDown, arr2)
      }
    }
  }

  return(list(text = text, lista = lista))
}

```

```{r}
if (!file.exists(fileLL) | force) {
  listPathways <- list()
  #---- loop first comparings ---------------------------------
  comparing <- comparings[1]

  for (comparing in comparings[1:2]) {
    cat(sprintf("\n\n>> %s \n", comparing))
    listPathways[[comparing]] <- list()

    if (comparing == "t6xt0") {
      mean.cols <- c("mean1", "mean2")
      pis.cols <- c("pisCtrl", "pisCase1")
      colControl <- pis.cols[1]
      colCase <- pis.cols[2]

      sample.cols <- good.cols[ c(grep("h0", good.cols), grep("h6", good.cols))]
      control_cols <- good.cols[grep("h0", good.cols)]
    } else if (comparing == "t12xt0") {
      mean.cols <- c("mean1", "mean3")
      pis.cols <- c("pisCtrl", "pisCase2")
      colControl <- pis.cols[1]
      colCase <- pis.cols[2]

      sample.cols <- good.cols[ c(grep("h0", good.cols), grep("h12", good.cols))]
      control_cols <- good.cols[grep("h0", good.cols)]
    }


    #--------- open files -------------
    dfLFC5 <- open_dfLFC5new(comparing)
    dfLFC5 <- dfLFC5[dfLFC5$gene.id != "", ]
    dfLFC5 <- dfLFC5[order(dfLFC5$fdr), ]
    nrow(dfLFC5)

    all.symbols <- as.character(dfLFC5$symbol)
    all.symbols <- all.symbols[!is.na(all.symbols)]
    all.symbols <- all.symbols[ all.symbols != ""]
    all.symbols <- unique(all.symbols)
    all.symbols <- all.symbols[order(all.symbols)]
    length(all.symbols)
    #-- 13667 ... 6h

    geneids <- as.character(dfLFC5$gene.id)

    stri <- sprintf("Opening dfLFC5 for %s\n", comparing)
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    stri <- sprintf("There are %d valid (ortholog) symbols from %d gene IDs.\n", length(all.symbols), length(geneids))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    # There are 13667 valid (ortholog) symbols from 14867 gene IDs.
    stri <- sprintf("that represents %.2f%% ortholog conversion.\n", length(all.symbols) * 100 / length(geneids))
    text <- write.log(text = text, filename = fileLog, f1 = stri)
    # that represents 91.93% ortholog conversion.

    filePis <- sprintf("%sdistribution_pis_geneID_ensembl_%s_%s_%s_%s_%s_LRT.tsv", main.result.bayes, tech, case, comparing, study, study_aux)
    filePis

    if (!file.exists(filePis)) {
      unique.groups
      totMed1 <- sum(dfLFC5$median1, na.rm = T)
      totMed2 <- sum(dfLFC5$median2, na.rm = T)
      totMed3 <- sum(dfLFC5$median3, na.rm = T)

      tots <- c(totMed1, totMed2, totMed3)
      names(tots) <- unique.groups
      tots

      pisCtrl <- as.numeric(dfLFC5$median1) / totMed1
      pisCase1 <- as.numeric(dfLFC5$median2) / totMed2
      pisCase2 <- as.numeric(dfLFC5$median3) / totMed3

      dfPis <- cbind(dfLFC5, pisCtrl, pisCase1, pisCase2)
      ret <- my.write.csv(dfPis, filePis)
    } else {
      dfPis <- my.read.csv(filePis)
    }



    #--- special product to evaluate high anova with high total mean -----
    n <- grep("median", colnames(dfPis))
    if (length(n) == 0) {
      stop("Didn't find median columns!!! Fatal error.")
      return(NULL)
    }

    #---- parameter:  median * anova
    meds <- apply(dfPis[, n], 1, sum) / length(n)
    anova.med.all <- -log10(as.numeric(dfPis$pvals.anova)) * meds


    dfPis <- cbind(dfPis, anova.med.all)
    colnames(dfPis)

    dfPis <- dfPis[order(-dfPis$anova.med.all), ]
    ncol(dfPis)
    dfPis[1:30, c(1:5, 28:32)]

    #--- parameter: bayes.min.median = 2
    if (comparing == comparings[1]) {
      dfLowVar <- dfPis[dfPis$pvals.anova >= bayes.pvals.anova &
        dfPis$median1 > bayes.min.median & dfPis$median2 > bayes.min.median, ]
    } else {
      dfLowVar <- dfPis[dfPis$pvals.anova >= bayes.pvals.anova &
        dfPis$median1 > bayes.min.median & dfPis$median3 > bayes.min.median, ]
    }

    nrow(dfPis)
    nrow(dfLowVar)
    colnames(dfLowVar)

    #-- high confidence ---------------------------------------------------
    # fdr <- .05; cutoffLods = -log10(fdr)
    # foldChange = 2; cutoffLogFC = log2(foldChange)

    dfLFC <- open_dfLFC(comparing)
    dfLFC[dfLFC$symbol == "CXCL1", ]
    nDEG.species <- nrow(dfLFC)

    dfLFC <- dfLFC[dfLFC$gene.id != "", ]
    nDEG.human <- nrow(dfLFC)
    stri <- sprintf("There are %d ortholog human DEGs in %d DEGs", nDEG.human, nDEG.species)
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    normal.degs <- as.character(dfLFC$gene.id)
    normal.symbol.degs <- unique_symbols_geneids(dfLFC$symbol)

    stri <- sprintf("There are %d valid ortholog DEGs - with symbol - DEGs from %d gene IDs.\n", length(normal.symbol.degs), length(normal.degs))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    # There are 13667 valid (ortholog) symbols from 14867 gene IDs.
    stri <- sprintf("that represents %.2f%% ortholog conversion.\n", length(normal.symbol.degs) * 100 / length(normal.degs))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    #-- high confidence ---------------------------------------------------
    #-- parameter: fdr.hc = 0.01 (usually) ;  cutoff.lfc.hc = 1
    dfLFC.HC <- dfPis[ (dfPis$fdr < fdr.hc) & (abs(dfPis$logFC) >= cutoff.lfc.hc), ]

    nHC.genes <- nrow(dfLFC.HC)
    nHC.genes

    #---- getting the minimum possible DEGs -------------------------------
    hc.degs <- as.character(dfLFC.HC$gene.id)
    length(hc.degs)

    want.to.cut <- F

    if (want.to.cut) {
      if (length(hc.degs) > 150) {
        hc.degs <- hc.degs[1:150]
        stri <- paste(stri, "Cutting to 150 DEGs.\n\n", sep = "")
        nHC.genes <- 150
      }
    }
    stri <- sprintf("There are %s high confident DEGs, with fdr < %.2f and cutoffLogFC > %.2f.\n", length(hc.degs), fdr.hc, cutoff.lfc.hc)
    text <- write.log(text = text, filename = fileLog, f1 = stri)


    dfPis <- dfPis[order(dfPis$fdr), ]
    nrow(dfPis)
    nrow(dfLowVar)
    nrow(dfLFC.HC)

    #-- high confidence -------------------------------------------------
    #-- parameter: fdr.limbo = 0.07,  cutoffLFC.limbo = log2(1.4) =  0.4854268
    dfLimbo <- dfPis[ (dfPis$fdr >= fdr.hc) & (dfPis$fdr < fdr.limbo) & (abs(dfPis$logFC) > cutoffLFC.limbo), ]
    nLimbo <- nrow(dfLimbo)

    #---- calc Limbo -----------------------------------------------------
    #---- getting the minimum possible DEGs ---------------.
    limbo <- as.character(dfLimbo$gene.id)

    sprintf("There are %s high confident DEGs, with fdr < %.2f and cutoffLogFC > %.2f.\n", length(hc.degs), fdr.hc, cutoff.lfc.hc)
    stri <- sprintf("There are %d low Var genes.\n", nrow(dfLowVar))
    stri <- sprintf("%sThere are %d limbo genes, with fdr >= %.2f and fdr < %.2f.\n", stri, length(limbo), fdr.hc, fdr.limbo)
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    #---- low Var cannot be DEGs ---------------------
    n <- which(dfLowVar$gene.id %in% hc.degs)
    length(n)

    # dfLowVar[n, 1:5]
    # dfLFC.HC[dfLFC.HC$symbol %in% c("LILRB3"), ]
    # dfLFC.HC[dfLFC.HC$gene.id %in%  dfLowVar[n, "gene.id"], ]
    # dfLowVar[n, ]

    #--- low Var should no be in Limbo ---------------------
    n <- which(dfLowVar$gene.id %in% limbo)
    if (length(n) > 0) {
      stri <- "\nUauuu: why hc.degs with low anova???\n\n"
      text <- write.log(text = text, filename = fileLog, f1 = stri)
      dfLowVar <- dfLowVar[-n, ]
    }
    midLayer <- as.character(dfLowVar$gene.id)

    par(mfrow = c(2, 2))
    hist(log10(dfLowVar$median1), main = "median1 - 'limbo' frequencies", breaks = 50)
    hist(log10(dfLowVar$median2), main = "median2 - 'limbo' frequencies", breaks = 50)
    hist(log10(dfLowVar$median3), main = "median3 - 'limbo' frequencies", breaks = 50)
    hist(-log10(dfLowVar$pvals.anova), main = "-log10(anova) - middle layer ANOVA distribution (limbo)", breaks = 50)
    par(mfrow = c(1, 1))

    cutoff.median <- median(dfLowVar$median2)
    nrow(dfLowVar[dfLowVar$median2 > cutoff.median, ])

    if (comparing == comparing[1]) {
      dfMiddleHighExp <- dfLowVar[dfLowVar$median2 > cutoff.median, ]
    } else {
      dfMiddleHighExp <- dfLowVar[dfLowVar$median3 > cutoff.median, ]
    }
    nrow(dfMiddleHighExp)

    no.degs <- as.character(dfMiddleHighExp$gene.id)

    stri <- sprintf("There are %d NO DEGs~Middle layer High Expression/High Variance.\n", length(no.degs))
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    length(hc.degs)
    length(limbo)
    length(no.degs)

    stri <- sprintf(
      "All genes = %d, DEGs+Limbo+NoDegs = %d (%d, %d, %d).\n",
      nrow(dfPis), length(hc.degs) + length(limbo) + length(no.degs), length(hc.degs), length(limbo), length(no.degs)
    )
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    #-- we will put allways the degs, and sorte 50% of limbo degs and HK
    #-- N is the variable parte - were we will sample

    #----------- Enriched Pathways db = KEGG -------------------------
    df.pathways <- open_Enrichment.string.table(comparing, db, 0)
    colnames(df.pathways)
    df.pathways[1:7, 1:9]

    #----------- Loop -------------------------
    stri <- paste("\n\nBegining loop of pathways:\n", sep = "")
    text <- write.log(text = text, filename = fileLog, f1 = stri)

    # i = which(df.pathways$pathway == "PI3K-Akt signaling pathway")
    # i = which(df.pathways$pathway == "Dopaminergic synapse")

    degs.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% hc.degs, "symbol"])
    limbo.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% limbo, "symbol"])

    #------------------  loop second pathways  ---------------------------------
    cat("loop pathways \n")
    sel.genes <- c()
    i <- 1
    for (i in 1:nrow(df.pathways)) { #  1:3

      dfPath <- df.pathways[i, ]
      pathway <- as.character(dfPath$pathway)
      stri_ini <- sprintf(">> %s - for %s:\n", pathway, comparing)
      stri_ini

      genes.in.pathway <- as.character(dfPath$genes.path)
      genes.in.pathway <- strsplit(genes.in.pathway, ";")[[1]]
      genes.in.pathway <- genes.in.pathway[order(genes.in.pathway)]
      genes.in.pathway
      if (length(genes.in.pathway) == 0) next

      # n.genes.found.in.pathway = as.character(dfPath$n.genes.found)
      genes.found.in.pathway <- as.character(dfPath$genes.found)
      nGenesFoundInPathway <- length(genes.found.in.pathway)

      all.sel.gene_ids <- c(hc.degs, limbo, no.degs)

      all.sel.gene_ids <- all.sel.gene_ids[all.sel.gene_ids != ""]
      all.sel.gene_ids <- unique(all.sel.gene_ids)
      length(all.sel.gene_ids)
      c(length(hc.degs), length(no.degs), length(limbo), length(all.sel.gene_ids))

      genes.found.in.pathway <- as.character(dfPath$genes.found)
      nGenesFoundInPathway <- length(genes.found.in.pathway)

      # length(midLayer)
      postDist <- c()
      prioriDist <- c()
      verossDist <- c()
      maxiGenes <- c()
      c(length(hc.degs), length(no.degs), length(limbo), length(all.sel.gene_ids))

      all.sel.gene_ids <- c(hc.degs, no.degs, limbo)
      all.sel.gene_ids

      nAllGenes <- length(all.symbols)

      nGoodGenes <- length(all.sel.gene_ids)
      nBadGenes <- nAllGenes - nGoodGenes

      if (i == 1) {
        stri <- sprintf("<<--->>\n\nThere are %d High Confident DEGs.\n", nHC.genes)
        stri <- sprintf("%sThere are %d medium confident DEGs (limbo).\n", stri, length(limbo))
        stri <- sprintf("%sThere are %d 'housekeeping' genes.\n", stri, length(midLayer))
        stri <- sprintf("%sThere are %d 'good genes' = hc.degs + limbo + HK.\n", stri, nGoodGenes)
        stri <- sprintf("%sWe will draw a maximum of %d genes.\n", stri, nSample)
        stri <- sprintf("%sTherefore, there are %d 'bad genes'.\n", stri, nBadGenes)
        stri <- sprintf("%sWe will perform %d loop as bootstrap'.\n", stri, nLoop)
        stri <- sprintf("%sOnly enrichment with %d or more genes will be selected'.\n", stri, min.EnrGenes)
        stri <- sprintf("%s\n\n%s", stri, stri_ini)
      } else {
        stri <- sprintf("<<--->>\n\n%s", stri_ini)
      }
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      #------------------------------------------------------------------------------------.
      #--- bootstrap -- bayesian ----------------------------------------------------------
      #------------------------------------------------------------------------------------.

      for (r in 1:nLoop) {
        #-- sample hk --------------------------------
        sample.gene_ids <- sample(all.sel.gene_ids, nSample, replace = FALSE)
        sample.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% sample.gene_ids, "symbol"])
        sample.symbols <- unique(sample.symbols)
        sample.symbols <- sample.symbols[sample.symbols != ""]
        sample.symbols <- sample.symbols[order(sample.symbols)]
        length(sample.symbols)

        nDrawn <- length(sample.symbols)

        # length(sample.symbols %in% all.symbols)

        genes.found <- sample.symbols[sample.symbols %in% genes.in.pathway]
        nFound <- length(genes.found)

        if (nFound < min.EnrGenes) {
          verossDist <- c(verossDist, NA)
          prioriDist <- c(prioriDist, NA)
          postDist <- c(postDist, NA)
          maxiGenes <- c(maxiGenes, NA)
          next
        }

        # ?phyper
        pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
        verossDist <- c(verossDist, pORA)

        #----------- log10(p(Gset)) -------------
        thisCol <- colCase
        log10.pGs <- calc.pGs.log10(dfPis, sample.gene_ids, thisCol = colCase)
        # print(log10.pGs)

        # post = -log10(pORA) - log10.pGs - (-log10(pORA + not.pORA))
        post <- -log10(pORA) - log10.pGs

        prioriDist <- c(prioriDist, log10.pGs)
        postDist <- c(postDist, post)
        maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
      }
      # postDist = postDist/sum(postDist)

      #------------- log -------------------

      # postDist = postDist/sum(postDist)
      postDist2 <- postDist[!is.na(postDist)]
      postDist2 <- postDist2[order(postDist2)]
      head(postDist2)
      tail(postDist2)
      hist(postDist2, breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")

      L <- length(postDist2)
      L
      up <- L * .95
      up
      stri <- sprintf("Maximum log posteriori %.1f.\n", up)

      postDist2.sig <- postDist2[1:up]

      limSup <- max(postDist2.sig)
      limSup

      n <- which(postDist2 < limSup)

      genes <- unlist(strsplit(maxiGenes[n], ";"))
      genes <- genes[!is.na(genes)]
      length(genes)
      genes <- unique(genes)
      length(genes)
      genes <- genes[order(genes)]

      stri <- sprintf("%sSelected genes: %s.\n", stri, paste(genes, collapse = ", "))


      genes.found.hc.degs <- genes[which(genes %in% degs.symbols)]
      stri <- sprintf("%sSelected genes in HC DEGs: %s.\n", stri, paste(genes.found.hc.degs, collapse = ", "))

      genes.found.limbo <- genes[which(genes %in% limbo.symbols)]
      stri <- sprintf("%sSelected genes in Limbo: %s.\n", stri, paste(genes.found.limbo, collapse = ", "))

      no.degs.symbols <- as.character(dfLFC5[dfLFC5$gene.id %in% no.degs, "symbol"])
      genes.found.no.degs <- genes[which(genes %in% no.degs.symbols)]
      stri <- sprintf("%sSelected genes in 'housekeeping like' genes: %s.\n", stri, paste(genes.found.no.degs, collapse = ", "))

      prev.found <- strsplit(as.character(dfPath$genes.found), ";")[[1]]
      prev.found <- prev.found[order(prev.found)]
      stri <- sprintf("%sPrevious found gene in pathway '%s': %s.\n", stri, pathway, paste(prev.found, collapse = ", "))

      found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo)
      found.degs.and.limbo <- unique(found.degs.and.limbo)
      found.degs.and.limbo <- found.degs.and.limbo[order(found.degs.and.limbo)]
      stri <- sprintf("%sHC DEGs + Limbo genes: %s.\n", stri, paste(found.degs.and.limbo, collapse = ", "))

      common.symbols <- prev.found[prev.found %in% found.degs.and.limbo]

      if (length(common.symbols) > 0) {
        stri <- sprintf("%s\nCommon DEGs: %s.\n", stri, paste(common.symbols, collapse = ", "))
      } else {
        stri <- sprintf("%s\nThere are no common DEGs.\n", stri)
      }

      only.prev.found <- prev.found[!prev.found %in% found.degs.and.limbo]
      if (length(common.symbols) > 0) {
        stri <- sprintf("%sDEGs not present in new algorithm: %s.\n", stri, paste(only.prev.found, collapse = ", "))
      } else {
        stri <- sprintf("%sAll previous DEGs were found using the new algorithm.\n", stri)
      }

      only.new.found <- found.degs.and.limbo[!found.degs.and.limbo %in% prev.found]
      if (length(common.symbols) > 0) {
        stri <- sprintf("%sNew DEGs found using the new algorithm: %s.\n", stri, paste(only.new.found, collapse = ", "))
      } else {
        stri <- sprintf("%sThere are NO NEW DEGs found using the new algorithm.\n", stri)
      }

      cat("\n------ new found ----------\n")
      print(length(only.new.found))
      only.new.found <- only.new.found[ !only.new.found %in% dfLFC$symbol ]
      print(length(only.new.found))
      cat("-------------------\n\n")

      sel.genes <- c(sel.genes, only.new.found)
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      #------------- end log ---------------
      pathway2 <- gsub("/", "_", pathway)
      pathway2 <- gsub(" ", "_", pathway2)
      pathway2 <- gsub("  ", " ", pathway2)
      pathway2 <- gsub(" _ ", "_", pathway2)
      pathway2 <- gsub("__", "_", pathway2)
      fileBayes <- sprintf("%s%s/hist_ll_%s_%s_%s_%s_%s_db'%s'_%s.png", main.result.bayes, db, tech, case, comparing, study, study_aux, db, pathway2)
      fileBayes
      file.exists(fileBayes)

      tryCatch({
        png(fileBayes, width = 10 * dpi, height = 8 * dpi, res = dpi)
        hist(postDist2, breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")
        dev.off()
      }, error = function(e) {
        stri <- sprintf("\n\nThere was an error: saving %s\n\n\n", fileBayes)
        text <- write.log(text = text, filename = fileLog, f1 = stri)
      })


      llMu <- round(mean(postDist2), 2)
      minLL <- round(min(postDist2), 2)
      maxLL <- round(max(postDist2), 2)
      stri <- sprintf("\n\nmean(Post): %.2f\n", llMu)
      stri <- sprintf("%smax(Post) = min(-log10(Post)): %.2f\n", stri, minLL)
      stri <- sprintf("%smin(Post) = max(-log10(Post)): %.2f\n", stri, maxLL)
      text <- write.log(text = text, filename = fileLog, f1 = stri)

      listPathways[[comparing]][[pathway]] <- list()
      listPathways[[comparing]][[pathway]][["mu"]] <- llMu
      listPathways[[comparing]][[pathway]][["minLL"]] <- minLL
      listPathways[[comparing]][[pathway]][["maxLL"]] <- maxLL
      listPathways[[comparing]][[pathway]][["n.genes.found.in.pathway"]] <- length(prev.found)
      listPathways[[comparing]][[pathway]][["genes.found.in.pathway"]] <- prev.found
      listPathways[[comparing]][[pathway]][["HC-DEGs"]] <- degs.symbols
      listPathways[[comparing]][[pathway]][["Limbo"]] <- limbo.symbols
      listPathways[[comparing]][[pathway]][["NO-DEGs"]] <- no.degs.symbols
      listPathways[[comparing]][[pathway]][["n.genes.found.new.algorithm"]] <- length(genes)
      listPathways[[comparing]][[pathway]][["genes.found.new.algorithm"]] <- genes
      listPathways[[comparing]][[pathway]][["n.genes.found.degs.limbo"]] <- length(found.degs.and.limbo)
      listPathways[[comparing]][[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
      listPathways[[comparing]][[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
      listPathways[[comparing]][[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
      listPathways[[comparing]][[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
      listPathways[[comparing]][[pathway]][["common"]] <- common.symbols
      listPathways[[comparing]][[pathway]][["only.previous.degs"]] <- only.prev.found
      listPathways[[comparing]][[pathway]][["only.new.degs"]] <- only.new.found
      listPathways[[comparing]][[pathway]][["ll.posteriori"]] <- postDist
    }

    sel.genes <- unique_symbols_geneids(sel.genes)
    sel.genes
    fileGenes <- sprintf("%snew_degs_bayes_symbol_db'%s'_%s_%s_%s_%s.txt", main.result.bayes, db, study, study_aux, case, comparing)
    fileGenes
    ret <- my.write.txt(sel.genes, fileGenes)

    fileEnsSamp <- sprintf("%sensembl_samples_bayes_db'%s'_%s_%s_%s_%s.txt", main.result.bayes, db, study, study_aux, case, comparing)
    fileEnsSamp
    ret <- my.write.txt(paste(all.sel.gene_ids, collapse = ", "), fileEnsSamp)

    dfLFC5 <- dfLFC5[dfLFC5$gene.id %in% all.sel.gene_ids, ]
    dfAux <- dfLFC5[dfLFC5$symbol %in% sel.genes, ]
    dfLFC.bayes <- rbind(dfLFC, dfAux)
    length(sel.genes)
    nrow(dfLFC)
    length(sel.genes) + nrow(dfLFC)
    #-- may have parologs ------.
    nrow(dfLFC.bayes)

    #-------- dfLFC.bayes does not allow repeated symbols ---------
    dfLFC.bayes <- dfLFC.bayes[order(dfLFC.bayes$symbol, dfLFC.bayes$fdr), ]
    # dfLFC.bayes[, c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    dfLFC.bayes[dfLFC.bayes$symbol == "CXCL1", ]
    nrow(dfLFC.bayes)
    # dfAux2 = dfLFC.bayes
    keep <- c()
    previous <- ""
    for (m in 1:nrow(dfLFC.bayes)) {
      symbol <- dfLFC.bayes[m, "symbol"]
      if (symbol != previous) {
        previous <- symbol
        keep <- c(keep, m)
      }
    }
    dfLFC.bayes <- dfLFC.bayes[keep, ]
    # dfLFC.bayes[, c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    # dfLFC.bayes[dfLFC.bayes$symbol == "CXCL1", ]
    nrow(dfLFC.bayes)
    #--------------------------------------------------------.

    dfLFC[dfLFC$symbol == "CXCL1", c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    dfLFC.bayes[dfLFC.bayes$symbol == "CXCL1", c("gene.id", "gene.id.spec", "symbol", "logFC", "fdr")]
    fileLFCb <- sprintf("%sLFC_bayes_ensembl_%s_%s_%s_db'%s'_%s_%s.tsv", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileLFCb
    ret <- my.write.csv(dfLFC.bayes, fileLFCb)

    fileLFCb3c <- sprintf("%sLFC_bayes_ensembl_%s_%s_%s_db'%s'_%s_%s_3cols.tsv", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileLFCb3c
    dfLFC.bayes <- dfLFC.bayes[, c("gene.id", "symbol", "logFC")]
    dfLFC.bayes <- unique(dfLFC.bayes)
    ret <- my.write.csv(dfLFC.bayes, fileLFCb3c)

    all.new.degs <- c(all.new.degs, sel.genes)

    symbols <- unique_symbols_geneids(dfLFC.bayes$symbol)
    stri.symbols <- paste(symbols, collapse = "\n")
    fileNewDegs <- sprintf("%sdegs_bayes_symbol_%s_%s_%s_db'%s'%s_%s.txt", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileNewDegs
    ret <- my.write.txt(stri.symbols, fileNewDegs)

    geneids <- unique_symbols_geneids(dfLFC.bayes$gene.id)
    stri.geneids <- paste(geneids, collapse = "\n")
    fileNewDegs <- sprintf("%sdegs_bayes_ensembl_%s_%s_%s_db'%s'_%s_%s.txt", main.result.bayes, tech, case, comparing, db, study, study_aux)
    fileNewDegs
    ret <- my.write.txt(stri.geneids, fileNewDegs)
  }

  #----------- End Loop -------------------------
  ret <- my.write.list(listPathways, fileLL)
} else {
  stri <- sprintf("\nFile already exists: '%s'\n", fileLL)
  text <- write.log(text = text, filename = fileLog, f1 = stri)
}

stri <- "--- end ---------------\n"
text <- write.log(text = text, filename = fileLog, f1 = stri)

comparing <- comparings[1]
names(listPathways)
names(listPathways[[comparings[1]]])
names(listPathways[[comparings[2]]])

names(listPathways[[comparings[1]]][["Hypertrophic cardiomyopathy (HCM)"]])
listPathways[[comparings[1]]][["Hypertrophic cardiomyopathy (HCM)"]][["mu"]]
listPathways[[comparings[2]]][["Measles"]][["mu"]]

#----------- loop of listPathways -----------------------
listPathways <- my.read.list(fileLL)
names(listPathways)

comparing <- comparings[1]

df <- data.frame()
comparing <- comparings[1]

for (comparing in comparings[1:2]) {
  cat(sprintf(">> comparing %s\n", comparing))
  nms <- names(listPathways[[comparing]])
  pathway <- nms[1]

  for (pathway in nms) {
    mu <- listPathways[[comparing]][[pathway]][["mu"]]
    minLL <- listPathways[[comparing]][[pathway]][["minLL"]]
    maxLL <- listPathways[[comparing]][[pathway]][["maxLL"]]
    genes.found.in.pathway <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.pathway"]], collapse = ", ")
    genes.found.in.hc.degs <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.hc-degs" ]], collapse = ", ")
    genes.found.in.limbo <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.limbo"]], collapse = ", ")
    genes.found.in.no.degs <- paste(listPathways[[comparing]][[pathway]][["genes.found.in.no-degs"]], collapse = ", ")
    common <- paste(listPathways[[comparing]][[pathway]][["common"]], collapse = ", ")
    only.previous.degs <- paste(listPathways[[comparing]][[pathway]][["only.previous.degs"]], collapse = ", ")

    new.degs <- listPathways[[comparing]][[pathway]][["only.new.degs"]]
    only.new.degs <- paste(new.degs, collapse = ", ")

    genes2 <- new.degs
    genes2 <- genes2[genes2 != ""]
    # print(only.new.degs)
    sel.genes <- c(sel.genes, genes2)

    dfAux <- data.frame(
      comparing, pathway, mu, minLL, maxLL, genes.found.in.pathway,
      genes.found.in.hc.degs, genes.found.in.limbo, genes.found.in.no.degs,
      common, only.previous.degs, only.new.degs
    )

    df <- rbind(df, dfAux)
  }
}

file.df.ll <- sprintf("%sLL_db'%s'_%s_%s.tsv", main.result.bayes, db, study, study_aux)
file.df.ll
colnames(df)
dim(df)
ret <- my.write.csv(df, file.df.ll)

head(df[, 1:5])
tail(df[, 1:5])

all.new.degs <- unique_symbols_geneids(all.new.degs)
all.new.degs

stri.all.new.degs <- paste(all.new.degs, collapse = "\n")
fileNewDegs <- sprintf("%sall_new_degs_%s_%s_db'%s'_%s_%s.txt", main.result.bayes, tech, case, db, study, study_aux)
fileNewDegs
ret <- my.write.txt(all.new.degs, fileNewDegs)

#--------------------- histogram posterioris -----------------------------
# dev.off()
par(mfrow = c(2, 2))
hist(df[df$comparing == comparings[1], "mu"], main = "mean posteriori distribution", xlab = "mean(posteriori)")
hist(df[df$comparing == comparings[2], "mu"], main = "mean posteriori distribution", xlab = "mean(posteriori)")
hist(df[df$comparing == comparings[1], "minLL"], main = "minimum posteriori distribution", xlab = "minimum(posteriori)")
hist(df[df$comparing == comparings[2], "minLL"], main = "minimum posteriori distribution", xlab = "minimum(posteriori)")
par(mfrow = c(1, 1))


df <- df[order(df$minLL), ]
df[1:8, 1:4]

nrow(df)

for (comparing in comparings[1:2]) {
  cat(sprintf("\n\n>> %s \n", comparing))

  dfAux <- df[df$comparing == comparing, ]
  for (i in 1:nrow(dfAux)) {
    cat(sprintf("\n ## pathway: '%s'\n", dfAux[i, "pathway"]))
    cat(sprintf("    previous found:    '%s'\n", dfAux[i, "genes.found.in.pathway"]))
    cat(sprintf("    common:            '%s'\n", dfAux[i, "common"]))
    cat(sprintf("    only.previous.degs:'%s'\n", dfAux[i, "only.previous.degs"]))
    cat(sprintf("    only.new.degs:     '%s'\n\n\n", dfAux[i, "only.new.degs"]))
  }
}
```

# Compare DBIA as is x clusterProfiler::enricher
```{r}
# DBIA as is
for (i in 1:nrow(enrichment_result)) {
  pathway <- enrichment_result$Description[i] %>% as.character()
  genes.in.pathway <- custom_KEGG_Pathways[[pathway]] %>% sort()
  genes.found.in.pathway <- enrichment_result %>% filter(Description == pathway) %>% dplyr::select(geneID) %>% str_split(pattern = "/") %>% unlist() %>% as.character()
  nGenesFoundInPathway <- genes.found.in.pathway %>% length()



  all.sel.gene_ids <- c(HC_DEGs, Limbo, NO.DEGs) %>% unique()
  nGoodGenes <- all.sel.gene_ids %>% length()
  nAllGenes <- countdata_input$rowname %>% length()
  nBadGenes <- nAllGenes - nGoodGenes

  #------------------------------------------------------------------------------------.
  #--- bootstrap -- bayesian ----------------------------------------------------------
  #------------------------------------------------------------------------------------.

  postDist <- c()
  prioriDist <- c()
  verossDist <- c()
  maxiGenes <- c()

  for (r in 1:nLoop) {
    pinrt(r)
    #-- sample hk --------------------------------
    sample.symbols <- sample(all.sel.gene_ids, nSample, replace = FALSE)
    genes.found <- sample.symbols[sample.symbols %in% genes.in.pathway]
    nFound <- genes.found %>% length()
    # Hypergeometric test
    pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    verossDist <- c(verossDist, pORA)
    log10.pGs <- median_df %>% calculate_pgs_log10(gene_ids = sample.gene_ids, column = pis_2)
    post <- -log10(pORA) - log10.pGs
    prioriDist <- c(prioriDist, log10.pGs)
    postDist <- c(postDist, post)
    maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
  }

  #------------- log -------------------

  # postDist = postDist/sum(postDist)

  postDist2 <- postDist %>% na.omit() %>% sort()
  postDist2 %>% hist(breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")

  L <- postDist2 %>% length()
  up <- L * .95
  postDist2.sig <- postDist2[1:up]

  limSup <- postDist2.sig %>% max()
  n <- which(postDist2 < limSup)

  genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()

  genes.found.hc.degs <- genes %>% .[. %in% HC_DEGs]
  genes.found.limbo <- genes %>% .[. %in% Limbo]
  genes.found.no.degs <- genes %>% .[. %in% NO.DEGs]

  prev.found <- genes.found.in.pathway %>% sort()
  found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo) %>% unique() %>% sort()
  common.symbols <- prev.found %>% .[. %in% found.degs.and.limbo]


  only.prev.found <- prev.found %>% .[!. %in% found.degs.and.limbo]

  only.new.found <- found.degs.and.limbo %>% .[!. %in% prev.found]
  # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?

  sel.genes <- c()
  sel.genes <- c(sel.genes, only.new.found)

  llMu <- postDist2 %>% mean() %>% round(digits = 2)
  minLL <- postDist2 %>% min() %>% round(digits = 2)
  maxLL <- postDist2 %>% max() %>% round(digits = 2)

  listPathways[[pathway]] <- list()
  listPathways[[pathway]][["mu"]] <- llMu
  listPathways[[pathway]][["minLL"]] <- minLL
  listPathways[[pathway]][["maxLL"]] <- maxLL
  listPathways[[pathway]][["n.genes.found.in.pathway"]] <- prev.found %>% length()
  listPathways[[pathway]][["genes.found.in.pathway"]] <- prev.found
  listPathways[[pathway]][["HC-DEGs"]] <- HC_DEGs
  listPathways[[pathway]][["Limbo"]] <- Limbo
  listPathways[[pathway]][["NO-DEGs"]] <- NO.DEGs
  listPathways[[pathway]][["n.genes.found.new.algorithm"]] <- genes %>% length()
  listPathways[[pathway]][["genes.found.new.algorithm"]] <- genes
  listPathways[[pathway]][["n.genes.found.degs.limbo"]] <- found.degs.and.limbo %>% length()
  listPathways[[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
  listPathways[[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
  listPathways[[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
  listPathways[[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
  listPathways[[pathway]][["common"]] <- common.symbols
  listPathways[[pathway]][["only.previous.degs"]] <- only.prev.found
  listPathways[[pathway]][["only.new.degs"]] <- only.new.found
  listPathways[[pathway]][["ll.posteriori"]] <- postDist
}


for (i in 1:nrow(enrichment_result)) {
  pathway <- enrichment_result$Description[i] %>% as.character()
  genes.in.pathway <- custom_KEGG_Pathways[[pathway]] %>% sort()
  genes.found.in.pathway <- enrichment_result %>% filter(Description == pathway) %>% dplyr::select(geneID) %>% str_split(pattern = "/") %>% unlist() %>% as.character()
  nGenesFoundInPathway <- genes.found.in.pathway %>% length()

  all.sel.gene_ids <- c(HC_DEGs, Limbo, NO.DEGs) %>% unique()
  nGoodGenes <- all.sel.gene_ids %>% length()
  nAllGenes <- countdata_input$rowname %>% length()
  nBadGenes <- nAllGenes - nGoodGenes

  #------------------------------------------------------------------------------------.
  #--- bootstrap -- bayesian ----------------------------------------------------------
  #------------------------------------------------------------------------------------.

  postDist <- c()
  prioriDist <- c()
  verossDist <- c()
  maxiGenes <- c()

  for (r in 1:nLoop) {
    pinrt(r)
    #-- sample hk --------------------------------
    sample.symbols <- sample(all.sel.gene_ids, nSample, replace = FALSE)
    
    genes.found <- sample.symbols[sample.symbols %in% genes.in.pathway]
    
    nFound <- genes.found %>% length()
    
    
    # Hypergeometric test
    allgenes <- gmt_list[, "gene"] %>% unique

    pORA <- clusterProfiler::enricher(
      gene = genes.found,
      pvalueCutoff = 1,
      qvalueCutoff = 1,
      universe = allgenes,
      TERM2GENE = gmt_list
    )    
    
    
    pORA <- phyper(nFound, nGoodGenes, nBadGenes, nDrawn, lower.tail = TRUE)
    verossDist <- c(verossDist, pORA)
    log10.pGs <- median_df %>% calculate_pgs_log10(gene_ids = sample.gene_ids, column = pis_2)
    post <- -log10(pORA) - log10.pGs
    prioriDist <- c(prioriDist, log10.pGs)
    postDist <- c(postDist, post)
    maxiGenes <- c(maxiGenes, paste(genes.found, collapse = ";"))
  }

  #------------- log -------------------

  # postDist = postDist/sum(postDist)

  postDist2 <- postDist %>% na.omit() %>% sort()
  postDist2 %>% hist(breaks = 25, main = pathway, xlab = "-log10(posterior distrib)")

  L <- postDist2 %>% length()
  up <- L * .95
  postDist2.sig <- postDist2[1:up]

  limSup <- postDist2.sig %>% max()
  n <- which(postDist2 < limSup)

  genes <- maxiGenes[n] %>% strsplit(";") %>% unlist() %>% na.omit() %>% unique() %>% sort()

  genes.found.hc.degs <- genes %>% .[. %in% HC_DEGs]
  genes.found.limbo <- genes %>% .[. %in% Limbo]
  genes.found.no.degs <- genes %>% .[. %in% NO.DEGs]

  prev.found <- genes.found.in.pathway %>% sort()
  found.degs.and.limbo <- c(genes.found.hc.degs, genes.found.limbo) %>% unique() %>% sort()
  common.symbols <- prev.found %>% .[. %in% found.degs.and.limbo]


  only.prev.found <- prev.found %>% .[!. %in% found.degs.and.limbo]

  only.new.found <- found.degs.and.limbo %>% .[!. %in% prev.found]
  # only.new.found <- only.new.found %>% .[!. %in% countdata_input$rowname] #?

  sel.genes <- c()
  sel.genes <- c(sel.genes, only.new.found)

  llMu <- postDist2 %>% mean() %>% round(digits = 2)
  minLL <- postDist2 %>% min() %>% round(digits = 2)
  maxLL <- postDist2 %>% max() %>% round(digits = 2)

  listPathways[[pathway]] <- list()
  listPathways[[pathway]][["mu"]] <- llMu
  listPathways[[pathway]][["minLL"]] <- minLL
  listPathways[[pathway]][["maxLL"]] <- maxLL
  listPathways[[pathway]][["n.genes.found.in.pathway"]] <- prev.found %>% length()
  listPathways[[pathway]][["genes.found.in.pathway"]] <- prev.found
  listPathways[[pathway]][["HC-DEGs"]] <- HC_DEGs
  listPathways[[pathway]][["Limbo"]] <- Limbo
  listPathways[[pathway]][["NO-DEGs"]] <- NO.DEGs
  listPathways[[pathway]][["n.genes.found.new.algorithm"]] <- genes %>% length()
  listPathways[[pathway]][["genes.found.new.algorithm"]] <- genes
  listPathways[[pathway]][["n.genes.found.degs.limbo"]] <- found.degs.and.limbo %>% length()
  listPathways[[pathway]][["genes.found.degs.limbo"]] <- found.degs.and.limbo
  listPathways[[pathway]][["genes.found.in.hc-degs"]] <- genes.found.hc.degs
  listPathways[[pathway]][["genes.found.in.limbo"]] <- genes.found.limbo
  listPathways[[pathway]][["genes.found.in.no.degs"]] <- genes.found.no.degs
  listPathways[[pathway]][["common"]] <- common.symbols
  listPathways[[pathway]][["only.previous.degs"]] <- only.prev.found
  listPathways[[pathway]][["only.new.degs"]] <- only.new.found
  listPathways[[pathway]][["ll.posteriori"]] <- postDist
}


```
